// @uniswap/smart-order-router/build/main/routers/router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cloneMixedRouteWithNewPools = exports.cloneV4RouteWithNewPools = exports.cloneV3RouteWithNewPools = exports.cloneV2RouteWithNewPools = exports.ISwapToRatio = exports.IRouter = exports.SwapType = exports.SwapToRatioStatus = exports.MixedRoute = exports.V2Route = exports.V3Route = exports.V4Route = void 0;
const router_sdk_1 = require("@uniswap/router-sdk");
const v2_sdk_1 = require("@uniswap/v2-sdk");
const v3_sdk_1 = require("@uniswap/v3-sdk");
const v4_sdk_1 = require("@uniswap/v4-sdk");
class V4Route extends v4_sdk_1.Route {
    constructor() {
        super(...arguments);
        this.protocol = router_sdk_1.Protocol.V4;
    }
}
exports.V4Route = V4Route;
class V3Route extends v3_sdk_1.Route {
    constructor() {
        super(...arguments);
        this.protocol = router_sdk_1.Protocol.V3;
    }
}
exports.V3Route = V3Route;
class V2Route extends v2_sdk_1.Route {
    constructor() {
        super(...arguments);
        this.protocol = router_sdk_1.Protocol.V2;
    }
}
exports.V2Route = V2Route;
class MixedRoute extends router_sdk_1.MixedRouteSDK {
    constructor() {
        super(...arguments);
        this.protocol = router_sdk_1.Protocol.MIXED;
    }
}
exports.MixedRoute = MixedRoute;
var SwapToRatioStatus;
(function (SwapToRatioStatus) {
    SwapToRatioStatus[SwapToRatioStatus["SUCCESS"] = 1] = "SUCCESS";
    SwapToRatioStatus[SwapToRatioStatus["NO_ROUTE_FOUND"] = 2] = "NO_ROUTE_FOUND";
    SwapToRatioStatus[SwapToRatioStatus["NO_SWAP_NEEDED"] = 3] = "NO_SWAP_NEEDED";
})(SwapToRatioStatus = exports.SwapToRatioStatus || (exports.SwapToRatioStatus = {}));
var SwapType;
(function (SwapType) {
    SwapType[SwapType["UNIVERSAL_ROUTER"] = 0] = "UNIVERSAL_ROUTER";
    SwapType[SwapType["SWAP_ROUTER_02"] = 1] = "SWAP_ROUTER_02";
})(SwapType = exports.SwapType || (exports.SwapType = {}));
/**
 * Provides functionality for finding optimal swap routes on the Uniswap protocol.
 *
 * @export
 * @abstract
 * @class IRouter
 */
class IRouter {
}
exports.IRouter = IRouter;
class ISwapToRatio {
}
exports.ISwapToRatio = ISwapToRatio;
function cloneV2RouteWithNewPools(originalRoute, newPools) {
    // Reuse the original input and output currencies
    const input = originalRoute.input;
    const output = originalRoute.output;
    // Construct a new Route instance with the new pairs
    return new V2Route(newPools, input, output);
}
exports.cloneV2RouteWithNewPools = cloneV2RouteWithNewPools;
function cloneV3RouteWithNewPools(originalRoute, newPools) {
    // Reuse the original input and output currencies
    const input = originalRoute.input;
    const output = originalRoute.output;
    // Construct a new Route instance with the new pairs
    return new V3Route(newPools, input, output);
}
exports.cloneV3RouteWithNewPools = cloneV3RouteWithNewPools;
function cloneV4RouteWithNewPools(originalRoute, newPools) {
    // Reuse the original input and output currencies
    const input = originalRoute.input;
    const output = originalRoute.output;
    // Construct a new Route instance with the new pairs
    return new V4Route(newPools, input, output);
}
exports.cloneV4RouteWithNewPools = cloneV4RouteWithNewPools;
function cloneMixedRouteWithNewPools(originalRoute, newPools) {
    // Reuse the original input and output currencies
    const input = originalRoute.input;
    const output = originalRoute.output;
    // Construct a new Route instance with the new pairs
    return new MixedRoute(newPools, input, output, /* retainFakePools = */ true);
}
exports.cloneMixedRouteWithNewPools = cloneMixedRouteWithNewPools;
