// @0glabs/0g-serving-broker/lib.commonjs/broker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZGComputeNetworkReadOnlyBroker = exports.ZGComputeNetworkBroker = exports.getNetworkType = exports.isDevMode = exports.CONTRACT_ADDRESSES = exports.HARDHAT_CHAIN_ID = exports.MAINNET_CHAIN_ID = exports.TESTNET_CHAIN_ID = void 0;
exports.createZGComputeNetworkBroker = createZGComputeNetworkBroker;
exports.createZGComputeNetworkReadOnlyBroker = createZGComputeNetworkReadOnlyBroker;
const ethers_1 = require("ethers");
const ledger_1 = require("./ledger");
const broker_1 = require("./fine-tuning/broker");
const broker_2 = require("./inference/broker/broker");
const read_only_broker_1 = require("./inference/broker/read-only-broker");
const read_only_broker_2 = require("./fine-tuning/broker/read-only-broker");
const constants_1 = require("./constants");
Object.defineProperty(exports, "TESTNET_CHAIN_ID", { enumerable: true, get: function () { return constants_1.TESTNET_CHAIN_ID; } });
Object.defineProperty(exports, "MAINNET_CHAIN_ID", { enumerable: true, get: function () { return constants_1.MAINNET_CHAIN_ID; } });
Object.defineProperty(exports, "HARDHAT_CHAIN_ID", { enumerable: true, get: function () { return constants_1.HARDHAT_CHAIN_ID; } });
Object.defineProperty(exports, "CONTRACT_ADDRESSES", { enumerable: true, get: function () { return constants_1.CONTRACT_ADDRESSES; } });
Object.defineProperty(exports, "isDevMode", { enumerable: true, get: function () { return constants_1.isDevMode; } });
Object.defineProperty(exports, "getNetworkType", { enumerable: true, get: function () { return constants_1.getNetworkType; } });
class ZGComputeNetworkBroker {
    ledger;
    inference;
    fineTuning;
    constructor(ledger, inferenceBroker, fineTuningBroker) {
        this.ledger = ledger;
        this.inference = inferenceBroker;
        this.fineTuning = fineTuningBroker;
    }
}
exports.ZGComputeNetworkBroker = ZGComputeNetworkBroker;
/**
 * createZGComputeNetworkBroker is used to initialize ZGComputeNetworkBroker
 *
 * This function automatically detects the network from the signer's provider and uses
 * appropriate contract addresses. You can override any address by providing it explicitly.
 *
 * @param signer - Signer from ethers.js.
 * @param ledgerCA - 0G Compute Network Ledger Contact address, auto-detected if not provided.
 * @param inferenceCA - 0G Compute Network Inference Serving contract address, auto-detected if not provided.
 * @param fineTuningCA - 0G Compute Network Fine Tuning Serving contract address, auto-detected if not provided.
 * @param gasPrice - Gas price for transactions. If not provided, the gas price will be calculated automatically.
 * @param maxGasPrice - Maximum gas price for transactions.
 * @param step - Step for gas price adjustment.
 *
 * @returns broker instance.
 *
 * @throws An error if the broker cannot be initialized.
 */
async function createZGComputeNetworkBroker(signer, ledgerCA, inferenceCA, fineTuningCA, gasPrice, maxGasPrice, step) {
    try {
        // Auto-detect network from signer's provider
        let defaultAddresses = constants_1.CONTRACT_ADDRESSES.testnet; // Default to testnet
        if (signer.provider) {
            const network = await signer.provider.getNetwork();
            const chainId = network.chainId;
            if (chainId === constants_1.MAINNET_CHAIN_ID) {
                defaultAddresses = constants_1.CONTRACT_ADDRESSES.mainnet;
                console.log(`Detected mainnet (chain ID: ${chainId})`);
            }
            else if (chainId === constants_1.TESTNET_CHAIN_ID) {
                if ((0, constants_1.isDevMode)()) {
                    defaultAddresses = constants_1.CONTRACT_ADDRESSES.testnetDev;
                    console.log(`Detected testnet [DEV MODE] (chain ID: ${chainId})`);
                }
                else {
                    defaultAddresses = constants_1.CONTRACT_ADDRESSES.testnet;
                    console.log(`Detected testnet (chain ID: ${chainId})`);
                }
            }
            else if (chainId === constants_1.HARDHAT_CHAIN_ID) {
                defaultAddresses = constants_1.CONTRACT_ADDRESSES.hardhat;
                console.log(`Detected hardhat (chain ID: ${chainId})`);
            }
            else {
                console.warn(`Unknown chain ID: ${chainId}. Using testnet addresses as default.`);
            }
        }
        else {
            console.warn('No provider found on signer. Using testnet addresses as default.');
        }
        // Use provided addresses or fall back to auto-detected defaults
        const finalLedgerCA = ledgerCA || defaultAddresses.ledger;
        const finalInferenceCA = inferenceCA || defaultAddresses.inference;
        const finalFineTuningCA = fineTuningCA || defaultAddresses.fineTuning;
        const ledger = await (0, ledger_1.createLedgerBroker)(signer, finalLedgerCA, finalInferenceCA, finalFineTuningCA, gasPrice, maxGasPrice, step);
        const inferenceBroker = await (0, broker_2.createInferenceBroker)(signer, finalInferenceCA, ledger);
        let fineTuningBroker;
        if (signer instanceof ethers_1.Wallet) {
            fineTuningBroker = await (0, broker_1.createFineTuningBroker)(signer, finalFineTuningCA, ledger, gasPrice, maxGasPrice, step);
        }
        const broker = new ZGComputeNetworkBroker(ledger, inferenceBroker, fineTuningBroker);
        return broker;
    }
    catch (error) {
        throw error;
    }
}
/**
 * Read-only version of ZGComputeNetworkBroker that doesn't require wallet connection.
 * Provides access to public blockchain data without authentication.
 *
 * Use this broker to:
 * - Browse available AI providers before connecting wallet
 * - Fetch service information and pricing
 * - Get provider health metrics
 *
 * Limitations:
 * - Cannot perform authenticated operations (send requests, manage accounts, etc.)
 * - No ledger services (require authentication)
 * - Read-only operations only
 */
class ZGComputeNetworkReadOnlyBroker {
    inference;
    fineTuning;
    constructor(inferenceBroker, fineTuningBroker) {
        this.inference = inferenceBroker;
        this.fineTuning = fineTuningBroker;
    }
}
exports.ZGComputeNetworkReadOnlyBroker = ZGComputeNetworkReadOnlyBroker;
/**
 * createZGComputeNetworkReadOnlyBroker creates a read-only broker WITHOUT wallet connection
 *
 * This broker provides access to public blockchain data (e.g., list providers) without
 * requiring user authentication. Perfect for browsing services before connecting a wallet.
 *
 * @param rpcUrl - JSON-RPC endpoint URL (e.g., 'https://evmrpc-testnet.0g.ai')
 * @param chainId - Optional chain ID. If not provided, will be detected from RPC endpoint.
 *
 * @returns Read-only broker instance with inference.listService() and inference.listServiceWithDetail()
 *
 * @example
 * ```typescript
 * // Create read-only broker (no wallet needed!)
 * const broker = await createZGComputeNetworkReadOnlyBroker(
 *   'https://evmrpc-testnet.0g.ai'
 * );
 *
 * // List all available providers (no authentication required)
 * const providers = await broker.inference.listService();
 * console.log(`Found ${providers.length} providers`);
 *
 * // Get detailed provider info with health metrics
 * const providersWithHealth = await broker.inference.listServiceWithDetail();
 * providersWithHealth.forEach(p => {
 *   console.log(`${p.provider}: ${p.healthMetrics?.uptime}% uptime`);
 * });
 * ```
 *
 * @throws An error if the broker cannot be initialized.
 */
async function createZGComputeNetworkReadOnlyBroker(rpcUrl, chainId) {
    try {
        // Create provider to detect network if chainId not provided
        let detectedChainId = chainId;
        if (!detectedChainId) {
            const provider = new ethers_1.JsonRpcProvider(rpcUrl);
            const network = await provider.getNetwork();
            detectedChainId = Number(network.chainId);
        }
        // Log detected network for debugging
        const chainIdBigInt = BigInt(detectedChainId);
        if (chainIdBigInt === constants_1.MAINNET_CHAIN_ID) {
            console.log(`Detected mainnet (chain ID: ${detectedChainId})`);
        }
        else if (chainIdBigInt === constants_1.TESTNET_CHAIN_ID) {
            if ((0, constants_1.isDevMode)()) {
                console.log(`Detected testnet [DEV MODE] (chain ID: ${detectedChainId})`);
            }
            else {
                console.log(`Detected testnet (chain ID: ${detectedChainId})`);
            }
        }
        else if (chainIdBigInt === constants_1.HARDHAT_CHAIN_ID) {
            console.log(`Detected hardhat (chain ID: ${detectedChainId})`);
        }
        else {
            console.warn(`Unknown chain ID: ${detectedChainId}. Using testnet addresses as default.`);
        }
        // Create read-only brokers (no authentication!)
        // The brokers will auto-detect contract addresses based on chainId
        const inferenceBroker = await (0, read_only_broker_1.createReadOnlyInferenceBroker)(rpcUrl, detectedChainId);
        const fineTuningBroker = await (0, read_only_broker_2.createReadOnlyFineTuningBroker)(rpcUrl, detectedChainId);
        const broker = new ZGComputeNetworkReadOnlyBroker(inferenceBroker, fineTuningBroker);
        return broker;
    }
    catch (error) {
        throw error;
    }
}
