// @uniswap/smart-order-router/build/main/providers/v4/euler-swap-hooks-subgraph-provider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EulerSwapHooksSubgraphProvider = void 0;
const router_sdk_1 = require("@uniswap/router-sdk");
const async_retry_1 = __importDefault(require("async-retry"));
const await_timeout_1 = __importDefault(require("await-timeout"));
const graphql_request_1 = require("graphql-request");
const lodash_1 = __importDefault(require("lodash"));
const util_1 = require("../../util");
const subgraph_provider_1 = require("../subgraph-provider");
const subgraph_provider_2 = require("./subgraph-provider");
class EulerSwapHooksSubgraphProvider {
    constructor(chainId, retries = 2, timeout = 30000, rollback = true, subgraphUrlOverride = subgraph_provider_2.SUBGRAPH_URL_BY_CHAIN[chainId]) {
        this.chainId = chainId;
        this.retries = retries;
        this.timeout = timeout;
        this.rollback = rollback;
        this.protocol = router_sdk_1.Protocol.V4;
        if (!subgraphUrlOverride) {
            throw new Error(`No subgraph url for chain id: ${chainId}`);
        }
        this.client = new graphql_request_1.GraphQLClient(subgraphUrlOverride);
    }
    async getHooks(providerConfig) {
        const beforeAll = Date.now();
        let blockNumber = (providerConfig === null || providerConfig === void 0 ? void 0 : providerConfig.blockNumber)
            ? await providerConfig.blockNumber
            : undefined;
        const query = (0, graphql_request_1.gql) `
      query getEulerSwapHooks($pageSize: Int!, $id: String) {
        eulerSwapHooks(
          first: $pageSize,
          ${blockNumber ? `block: { number: ${blockNumber} }` : ``}
          where: { id_gt: $id }
        ) {
          id
          hook
          asset0
          asset1
          eulerAccount
        }
      }
    `;
        let hooks = [];
        util_1.log.info(`Getting hooks from the subgraph with page size ${subgraph_provider_1.PAGE_SIZE}${(providerConfig === null || providerConfig === void 0 ? void 0 : providerConfig.blockNumber)
            ? ` as of block ${providerConfig === null || providerConfig === void 0 ? void 0 : providerConfig.blockNumber}`
            : ''}.`);
        let retries = 0;
        await (0, async_retry_1.default)(async () => {
            const timeout = new await_timeout_1.default();
            const getHooks = async () => {
                let lastId = '';
                let hooks = [];
                let hooksPage = [];
                // metrics variables
                let totalPages = 0;
                do {
                    totalPages += 1;
                    const hooksResult = await this.client.request(query, {
                        pageSize: subgraph_provider_1.PAGE_SIZE,
                        id: lastId,
                    });
                    hooksPage = hooksResult.eulerSwapHooks;
                    hooks = hooks.concat(hooksPage);
                    lastId = hooks[hooks.length - 1].id;
                    util_1.metric.putMetric(`SubgraphProvider.chain_${this.chainId}.getHooks.paginate.pageSize`, hooksPage.length);
                } while (hooksPage.length > 0);
                util_1.metric.putMetric(`SubgraphProvider.chain_${this.chainId}.getHooks.paginate`, totalPages);
                util_1.metric.putMetric(`SubgraphProvider.chain_${this.chainId}.getHooks.hooks.length`, hooks.length);
                return hooks;
            };
            try {
                const getHooksPromise = getHooks();
                const timerPromise = timeout.set(this.timeout).then(() => {
                    throw new Error(`Timed out getting hooks from subgraph: ${this.timeout}`);
                });
                hooks = await Promise.race([getHooksPromise, timerPromise]);
                return;
            }
            catch (err) {
                util_1.log.error({ err }, `Error fetching ${this.protocol} Subgraph Hooks.`);
                throw err;
            }
            finally {
                timeout.clear();
            }
        }, {
            retries: this.retries,
            onRetry: (err, retry) => {
                retries += 1;
                if (this.rollback &&
                    blockNumber &&
                    lodash_1.default.includes(err.message, 'indexed up to')) {
                    util_1.metric.putMetric(`SubgraphProvider.chain_${this.chainId}.getHooks.indexError`, 1);
                    blockNumber = blockNumber - 10;
                    util_1.log.info(`Detected subgraph indexing error. Rolled back block number to: ${blockNumber}`);
                }
                util_1.metric.putMetric(`SubgraphProvider.chain_${this.chainId}.getHooks.timeout`, 1);
                hooks = [];
                util_1.log.info({ err }, `Failed to get hooks from subgraph. Retry attempt: ${retry}`);
            },
        });
        util_1.metric.putMetric(`${this.protocol}SubgraphProvider.chain_${this.chainId}.getHooks.retries`, retries);
        util_1.metric.putMetric(`${this.protocol}SubgraphProvider.chain_${this.chainId}.getHooks.latency`, Date.now() - beforeAll);
        return hooks;
    }
    async getPoolByHook(hook, providerConfig) {
        const beforeAll = Date.now();
        const blockNumber = (providerConfig === null || providerConfig === void 0 ? void 0 : providerConfig.blockNumber)
            ? await providerConfig.blockNumber
            : undefined;
        const query = (0, graphql_request_1.gql) `
      query getPools($pageSize: Int!, $hooks: String) {
        pools(
          first: $pageSize,
          ${blockNumber ? `block: { number: ${blockNumber} }` : ``}
          where: {hooks: $hooks}
        ) {
          id
          token0 {
            symbol
            id
            derivedETH
          }
          token1 {
            symbol
            id
            derivedETH
          }
          feeTier
          tick
          tickSpacing
          liquidity
          hooks
          totalValueLockedUSD
          totalValueLockedETH
          totalValueLockedUSDUntracked
          sqrtPrice
        }
      }
    `;
        let pool = undefined;
        util_1.log.info(`Getting pool by hook from the subgraph with page size ${subgraph_provider_1.PAGE_SIZE}${(providerConfig === null || providerConfig === void 0 ? void 0 : providerConfig.blockNumber)
            ? ` as of block ${providerConfig === null || providerConfig === void 0 ? void 0 : providerConfig.blockNumber}`
            : ''}.`);
        const poolResult = await this.client.request(query, {
            pageSize: subgraph_provider_1.PAGE_SIZE,
            hooks: hook.toLowerCase(),
        });
        pool = poolResult.pools[0];
        util_1.metric.putMetric(`SubgraphProvider.chain_${this.chainId}.getPoolByHook.pools.length`, poolResult.pools.length);
        util_1.metric.putMetric(`${this.protocol}SubgraphProvider.chain_${this.chainId}.getPoolByHook.latency`, Date.now() - beforeAll);
        return pool;
    }
}
exports.EulerSwapHooksSubgraphProvider = EulerSwapHooksSubgraphProvider;
