// @uniswap/smart-order-router/build/main/providers/v4/subgraph-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.V4SubgraphProvider = exports.SUBGRAPH_URL_BY_CHAIN = void 0;
const router_sdk_1 = require("@uniswap/router-sdk");
const sdk_core_1 = require("@uniswap/sdk-core");
const subgraph_provider_1 = require("../subgraph-provider");
exports.SUBGRAPH_URL_BY_CHAIN = {
    [sdk_core_1.ChainId.SEPOLIA]: '',
};
class V4SubgraphProvider extends subgraph_provider_1.SubgraphProvider {
    constructor(chainId, retries = 2, timeout = 30000, rollback = true, trackedEthThreshold = 0.01, trackedZoraEthThreshold = 0.001, zoraHooks = new Set(), untrackedUsdThreshold = Number.MAX_VALUE, subgraphUrlOverride, bearerToken) {
        super(router_sdk_1.Protocol.V4, chainId, retries, timeout, rollback, trackedEthThreshold, trackedZoraEthThreshold, zoraHooks, untrackedUsdThreshold, subgraphUrlOverride !== null && subgraphUrlOverride !== void 0 ? subgraphUrlOverride : exports.SUBGRAPH_URL_BY_CHAIN[chainId], bearerToken);
    }
    mapSubgraphPool(rawPool) {
        return {
            id: rawPool.id,
            feeTier: rawPool.feeTier,
            tickSpacing: rawPool.tickSpacing,
            hooks: rawPool.hooks,
            liquidity: rawPool.liquidity,
            token0: {
                symbol: rawPool.token0.symbol,
                id: rawPool.token0.id,
                name: rawPool.token0.name,
                decimals: rawPool.token0.decimals,
            },
            token1: {
                symbol: rawPool.token1.symbol,
                id: rawPool.token1.id,
                name: rawPool.token1.name,
                decimals: rawPool.token1.decimals,
            },
            tvlETH: parseFloat(rawPool.totalValueLockedETH),
            tvlUSD: parseFloat(rawPool.totalValueLockedUSD),
        };
    }
    // Override to include V4-specific fields
    getPoolFields() {
        return `
      id
      token0 {
        symbol
        id
        name
        decimals
      }
      token1 {
        symbol
        id
        name
        decimals
      }
      feeTier
      tickSpacing
      hooks
      liquidity
      totalValueLockedUSD
      totalValueLockedETH
    `;
    }
}
exports.V4SubgraphProvider = V4SubgraphProvider;
