// @uniswap/v3-sdk/dist/cjs/src/utils/sqrtPriceMath.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqrtPriceMath = void 0;
const tslib_1 = require("tslib");
const sdk_core_1 = require("@uniswap/sdk-core");
const jsbi_1 = tslib_1.__importDefault(require("jsbi"));
const tiny_invariant_1 = tslib_1.__importDefault(require("tiny-invariant"));
const internalConstants_1 = require("../internalConstants");
const fullMath_1 = require("./fullMath");
const MaxUint160 = jsbi_1.default.subtract(jsbi_1.default.exponentiate(jsbi_1.default.BigInt(2), jsbi_1.default.BigInt(160)), internalConstants_1.ONE);
function multiplyIn256(x, y) {
    const product = jsbi_1.default.multiply(x, y);
    return jsbi_1.default.bitwiseAnd(product, sdk_core_1.MaxUint256);
}
function addIn256(x, y) {
    const sum = jsbi_1.default.add(x, y);
    return jsbi_1.default.bitwiseAnd(sum, sdk_core_1.MaxUint256);
}
class SqrtPriceMath {
    /**
     * Cannot be constructed.
     */
    constructor() { }
    static getAmount0Delta(sqrtRatioAX96, sqrtRatioBX96, liquidity, roundUp) {
        if (jsbi_1.default.greaterThan(sqrtRatioAX96, sqrtRatioBX96)) {
            ;
            [sqrtRatioAX96, sqrtRatioBX96] = [sqrtRatioBX96, sqrtRatioAX96];
        }
        const numerator1 = jsbi_1.default.leftShift(liquidity, jsbi_1.default.BigInt(96));
        const numerator2 = jsbi_1.default.subtract(sqrtRatioBX96, sqrtRatioAX96);
        return roundUp
            ? fullMath_1.FullMath.mulDivRoundingUp(fullMath_1.FullMath.mulDivRoundingUp(numerator1, numerator2, sqrtRatioBX96), internalConstants_1.ONE, sqrtRatioAX96)
            : jsbi_1.default.divide(jsbi_1.default.divide(jsbi_1.default.multiply(numerator1, numerator2), sqrtRatioBX96), sqrtRatioAX96);
    }
    static getAmount1Delta(sqrtRatioAX96, sqrtRatioBX96, liquidity, roundUp) {
        if (jsbi_1.default.greaterThan(sqrtRatioAX96, sqrtRatioBX96)) {
            ;
            [sqrtRatioAX96, sqrtRatioBX96] = [sqrtRatioBX96, sqrtRatioAX96];
        }
        return roundUp
            ? fullMath_1.FullMath.mulDivRoundingUp(liquidity, jsbi_1.default.subtract(sqrtRatioBX96, sqrtRatioAX96), internalConstants_1.Q96)
            : jsbi_1.default.divide(jsbi_1.default.multiply(liquidity, jsbi_1.default.subtract(sqrtRatioBX96, sqrtRatioAX96)), internalConstants_1.Q96);
    }
    static getNextSqrtPriceFromInput(sqrtPX96, liquidity, amountIn, zeroForOne) {
        (0, tiny_invariant_1.default)(jsbi_1.default.greaterThan(sqrtPX96, internalConstants_1.ZERO));
        (0, tiny_invariant_1.default)(jsbi_1.default.greaterThan(liquidity, internalConstants_1.ZERO));
        return zeroForOne
            ? this.getNextSqrtPriceFromAmount0RoundingUp(sqrtPX96, liquidity, amountIn, true)
            : this.getNextSqrtPriceFromAmount1RoundingDown(sqrtPX96, liquidity, amountIn, true);
    }
    static getNextSqrtPriceFromOutput(sqrtPX96, liquidity, amountOut, zeroForOne) {
        (0, tiny_invariant_1.default)(jsbi_1.default.greaterThan(sqrtPX96, internalConstants_1.ZERO));
        (0, tiny_invariant_1.default)(jsbi_1.default.greaterThan(liquidity, internalConstants_1.ZERO));
        return zeroForOne
            ? this.getNextSqrtPriceFromAmount1RoundingDown(sqrtPX96, liquidity, amountOut, false)
            : this.getNextSqrtPriceFromAmount0RoundingUp(sqrtPX96, liquidity, amountOut, false);
    }
    static getNextSqrtPriceFromAmount0RoundingUp(sqrtPX96, liquidity, amount, add) {
        if (jsbi_1.default.equal(amount, internalConstants_1.ZERO))
            return sqrtPX96;
        const numerator1 = jsbi_1.default.leftShift(liquidity, jsbi_1.default.BigInt(96));
        if (add) {
            let product = multiplyIn256(amount, sqrtPX96);
            if (jsbi_1.default.equal(jsbi_1.default.divide(product, amount), sqrtPX96)) {
                const denominator = addIn256(numerator1, product);
                if (jsbi_1.default.greaterThanOrEqual(denominator, numerator1)) {
                    return fullMath_1.FullMath.mulDivRoundingUp(numerator1, sqrtPX96, denominator);
                }
            }
            return fullMath_1.FullMath.mulDivRoundingUp(numerator1, internalConstants_1.ONE, jsbi_1.default.add(jsbi_1.default.divide(numerator1, sqrtPX96), amount));
        }
        else {
            let product = multiplyIn256(amount, sqrtPX96);
            (0, tiny_invariant_1.default)(jsbi_1.default.equal(jsbi_1.default.divide(product, amount), sqrtPX96));
            (0, tiny_invariant_1.default)(jsbi_1.default.greaterThan(numerator1, product));
            const denominator = jsbi_1.default.subtract(numerator1, product);
            return fullMath_1.FullMath.mulDivRoundingUp(numerator1, sqrtPX96, denominator);
        }
    }
    static getNextSqrtPriceFromAmount1RoundingDown(sqrtPX96, liquidity, amount, add) {
        if (add) {
            const quotient = jsbi_1.default.lessThanOrEqual(amount, MaxUint160)
                ? jsbi_1.default.divide(jsbi_1.default.leftShift(amount, jsbi_1.default.BigInt(96)), liquidity)
                : jsbi_1.default.divide(jsbi_1.default.multiply(amount, internalConstants_1.Q96), liquidity);
            return jsbi_1.default.add(sqrtPX96, quotient);
        }
        else {
            const quotient = fullMath_1.FullMath.mulDivRoundingUp(amount, internalConstants_1.Q96, liquidity);
            (0, tiny_invariant_1.default)(jsbi_1.default.greaterThan(sqrtPX96, quotient));
            return jsbi_1.default.subtract(sqrtPX96, quotient);
        }
    }
}
exports.SqrtPriceMath = SqrtPriceMath;
