// @uniswap/v3-sdk/dist/cjs/src/utils/v3swap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.v3Swap = void 0;
const tslib_1 = require("tslib");
const swapMath_1 = require("./swapMath");
const liquidityMath_1 = require("./liquidityMath");
const jsbi_1 = tslib_1.__importDefault(require("jsbi"));
const tiny_invariant_1 = tslib_1.__importDefault(require("tiny-invariant"));
const tickMath_1 = require("./tickMath");
const internalConstants_1 = require("../internalConstants");
async function v3Swap(fee, sqrtRatioX96, tickCurrent, liquidity, tickSpacing, tickDataProvider, zeroForOne, amountSpecified, sqrtPriceLimitX96) {
    if (!sqrtPriceLimitX96)
        sqrtPriceLimitX96 = zeroForOne
            ? jsbi_1.default.add(tickMath_1.TickMath.MIN_SQRT_RATIO, internalConstants_1.ONE)
            : jsbi_1.default.subtract(tickMath_1.TickMath.MAX_SQRT_RATIO, internalConstants_1.ONE);
    if (zeroForOne) {
        (0, tiny_invariant_1.default)(jsbi_1.default.greaterThan(sqrtPriceLimitX96, tickMath_1.TickMath.MIN_SQRT_RATIO), 'RATIO_MIN');
        (0, tiny_invariant_1.default)(jsbi_1.default.lessThan(sqrtPriceLimitX96, sqrtRatioX96), 'RATIO_CURRENT');
    }
    else {
        (0, tiny_invariant_1.default)(jsbi_1.default.lessThan(sqrtPriceLimitX96, tickMath_1.TickMath.MAX_SQRT_RATIO), 'RATIO_MAX');
        (0, tiny_invariant_1.default)(jsbi_1.default.greaterThan(sqrtPriceLimitX96, sqrtRatioX96), 'RATIO_CURRENT');
    }
    const exactInput = jsbi_1.default.greaterThanOrEqual(amountSpecified, internalConstants_1.ZERO);
    // keep track of swap state
    const state = {
        amountSpecifiedRemaining: amountSpecified,
        amountCalculated: internalConstants_1.ZERO,
        sqrtPriceX96: sqrtRatioX96,
        tick: tickCurrent,
        liquidity: liquidity,
    };
    // start swap while loop
    while (jsbi_1.default.notEqual(state.amountSpecifiedRemaining, internalConstants_1.ZERO) && state.sqrtPriceX96 !== sqrtPriceLimitX96) {
        let step = {};
        step.sqrtPriceStartX96 = state.sqrtPriceX96;
        [step.tickNext, step.initialized] = await tickDataProvider.nextInitializedTickWithinOneWord(state.tick, zeroForOne, tickSpacing);
        if (step.tickNext < tickMath_1.TickMath.MIN_TICK) {
            step.tickNext = tickMath_1.TickMath.MIN_TICK;
        }
        else if (step.tickNext > tickMath_1.TickMath.MAX_TICK) {
            step.tickNext = tickMath_1.TickMath.MAX_TICK;
        }
        step.sqrtPriceNextX96 = tickMath_1.TickMath.getSqrtRatioAtTick(step.tickNext);
        [state.sqrtPriceX96, step.amountIn, step.amountOut, step.feeAmount] = swapMath_1.SwapMath.computeSwapStep(state.sqrtPriceX96, (zeroForOne
            ? jsbi_1.default.lessThan(step.sqrtPriceNextX96, sqrtPriceLimitX96)
            : jsbi_1.default.greaterThan(step.sqrtPriceNextX96, sqrtPriceLimitX96))
            ? sqrtPriceLimitX96
            : step.sqrtPriceNextX96, state.liquidity, state.amountSpecifiedRemaining, fee);
        if (exactInput) {
            state.amountSpecifiedRemaining = jsbi_1.default.subtract(state.amountSpecifiedRemaining, jsbi_1.default.add(step.amountIn, step.feeAmount));
            state.amountCalculated = jsbi_1.default.subtract(state.amountCalculated, step.amountOut);
        }
        else {
            state.amountSpecifiedRemaining = jsbi_1.default.add(state.amountSpecifiedRemaining, step.amountOut);
            state.amountCalculated = jsbi_1.default.add(state.amountCalculated, jsbi_1.default.add(step.amountIn, step.feeAmount));
        }
        // TODO
        if (jsbi_1.default.equal(state.sqrtPriceX96, step.sqrtPriceNextX96)) {
            // if the tick is initialized, run the tick transition
            if (step.initialized) {
                let liquidityNet = jsbi_1.default.BigInt((await tickDataProvider.getTick(step.tickNext)).liquidityNet);
                // if we're moving leftward, we interpret liquidityNet as the opposite sign
                // safe because liquidityNet cannot be type(int128).min
                if (zeroForOne)
                    liquidityNet = jsbi_1.default.multiply(liquidityNet, internalConstants_1.NEGATIVE_ONE);
                state.liquidity = liquidityMath_1.LiquidityMath.addDelta(state.liquidity, liquidityNet);
            }
            state.tick = zeroForOne ? step.tickNext - 1 : step.tickNext;
        }
        else if (jsbi_1.default.notEqual(state.sqrtPriceX96, step.sqrtPriceStartX96)) {
            // updated comparison function
            // recompute unless we're on a lower tick boundary (i.e. already transitioned ticks), and haven't moved
            state.tick = tickMath_1.TickMath.getTickAtSqrtRatio(state.sqrtPriceX96);
        }
    }
    return {
        amountCalculated: state.amountCalculated,
        sqrtRatioX96: state.sqrtPriceX96,
        liquidity: state.liquidity,
        tickCurrent: state.tick,
    };
}
exports.v3Swap = v3Swap;
