// @uniswap/v3-sdk/dist/esm/src/quoter.js
import { Interface } from '@ethersproject/abi';
import { TradeType } from '@uniswap/sdk-core';
import { encodeRouteToPath, toHex } from './utils';
import IQuoter from '@uniswap/v3-periphery/artifacts/contracts/lens/Quoter.sol/Quoter.json';
import IQuoterV2 from '@uniswap/swap-router-contracts/artifacts/contracts/lens/QuoterV2.sol/QuoterV2.json';
import invariant from 'tiny-invariant';
/**
 * Represents the Uniswap V3 QuoterV1 contract with a method for returning the formatted
 * calldata needed to call the quoter contract.
 */
export class SwapQuoter {
    /**
     * Produces the on-chain method name of the appropriate function within QuoterV2,
     * and the relevant hex encoded parameters.
     * @template TInput The input token, either Ether or an ERC-20
     * @template TOutput The output token, either Ether or an ERC-20
     * @param route The swap route, a list of pools through which a swap can occur
     * @param amount The amount of the quote, either an amount in, or an amount out
     * @param tradeType The trade type, either exact input or exact output
     * @param options The optional params including price limit and Quoter contract switch
     * @returns The formatted calldata
     */
    static quoteCallParameters(route, amount, tradeType, options = {}) {
        var _a;
        const singleHop = route.pools.length === 1;
        const quoteAmount = toHex(amount.quotient);
        let calldata;
        const swapInterface = options.useQuoterV2 ? this.V2INTERFACE : this.V1INTERFACE;
        if (singleHop) {
            const baseQuoteParams = {
                tokenIn: route.tokenPath[0].address,
                tokenOut: route.tokenPath[1].address,
                fee: route.pools[0].fee,
                sqrtPriceLimitX96: toHex((_a = options === null || options === void 0 ? void 0 : options.sqrtPriceLimitX96) !== null && _a !== void 0 ? _a : 0),
            };
            const v2QuoteParams = {
                ...baseQuoteParams,
                ...(tradeType === TradeType.EXACT_INPUT ? { amountIn: quoteAmount } : { amount: quoteAmount }),
            };
            const v1QuoteParams = [
                baseQuoteParams.tokenIn,
                baseQuoteParams.tokenOut,
                baseQuoteParams.fee,
                quoteAmount,
                baseQuoteParams.sqrtPriceLimitX96,
            ];
            const tradeTypeFunctionName = tradeType === TradeType.EXACT_INPUT ? 'quoteExactInputSingle' : 'quoteExactOutputSingle';
            calldata = swapInterface.encodeFunctionData(tradeTypeFunctionName, options.useQuoterV2 ? [v2QuoteParams] : v1QuoteParams);
        }
        else {
            invariant((options === null || options === void 0 ? void 0 : options.sqrtPriceLimitX96) === undefined, 'MULTIHOP_PRICE_LIMIT');
            const path = encodeRouteToPath(route, tradeType === TradeType.EXACT_OUTPUT);
            const tradeTypeFunctionName = tradeType === TradeType.EXACT_INPUT ? 'quoteExactInput' : 'quoteExactOutput';
            calldata = swapInterface.encodeFunctionData(tradeTypeFunctionName, [path, quoteAmount]);
        }
        return {
            calldata,
            value: toHex(0),
        };
    }
}
SwapQuoter.V1INTERFACE = new Interface(IQuoter.abi);
SwapQuoter.V2INTERFACE = new Interface(IQuoterV2.abi);
