// @uniswap/v2-sdk/dist/cjs/src/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ONE_HUNDRED_PERCENT = exports.ZERO_PERCENT = exports.BASIS_POINTS = exports._1000 = exports._997 = exports.FIVE = exports.ONE = exports.ZERO = exports.MINIMUM_LIQUIDITY = exports.INIT_CODE_HASH = exports.FACTORY_ADDRESS_MAP = exports.FACTORY_ADDRESS = void 0;
const tslib_1 = require("tslib");
const sdk_core_1 = require("@uniswap/sdk-core");
const jsbi_1 = tslib_1.__importDefault(require("jsbi"));
/**
 * @deprecated use FACTORY_ADDRESS_MAP instead
 */
exports.FACTORY_ADDRESS = '0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f';
exports.FACTORY_ADDRESS_MAP = sdk_core_1.V2_FACTORY_ADDRESSES;
exports.INIT_CODE_HASH = '0x96e8ac4277198ff8b6f785478aa9a39f403cb768dd02cbee326c3e7da348845f';
exports.MINIMUM_LIQUIDITY = jsbi_1.default.BigInt(1000);
// exports for internal consumption
exports.ZERO = jsbi_1.default.BigInt(0);
exports.ONE = jsbi_1.default.BigInt(1);
exports.FIVE = jsbi_1.default.BigInt(5);
exports._997 = jsbi_1.default.BigInt(997);
exports._1000 = jsbi_1.default.BigInt(1000);
exports.BASIS_POINTS = jsbi_1.default.BigInt(10000);
exports.ZERO_PERCENT = new sdk_core_1.Percent(exports.ZERO);
exports.ONE_HUNDRED_PERCENT = new sdk_core_1.Percent(exports.ONE);
