// @uniswap/router-sdk/dist/cjs/src/approveAndCall.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApproveAndCall = exports.isMint = exports.ApprovalTypes = void 0;
const tslib_1 = require("tslib");
const abi_1 = require("@ethersproject/abi");
const tiny_invariant_1 = tslib_1.__importDefault(require("tiny-invariant"));
const IApproveAndCall_json_1 = tslib_1.__importDefault(require("@uniswap/swap-router-contracts/artifacts/contracts/interfaces/IApproveAndCall.sol/IApproveAndCall.json"));
const v3_sdk_1 = require("@uniswap/v3-sdk");
const jsbi_1 = tslib_1.__importDefault(require("jsbi"));
var ApprovalTypes;
(function (ApprovalTypes) {
    ApprovalTypes[ApprovalTypes["NOT_REQUIRED"] = 0] = "NOT_REQUIRED";
    ApprovalTypes[ApprovalTypes["MAX"] = 1] = "MAX";
    ApprovalTypes[ApprovalTypes["MAX_MINUS_ONE"] = 2] = "MAX_MINUS_ONE";
    ApprovalTypes[ApprovalTypes["ZERO_THEN_MAX"] = 3] = "ZERO_THEN_MAX";
    ApprovalTypes[ApprovalTypes["ZERO_THEN_MAX_MINUS_ONE"] = 4] = "ZERO_THEN_MAX_MINUS_ONE";
})(ApprovalTypes = exports.ApprovalTypes || (exports.ApprovalTypes = {}));
// type guard
function isMint(options) {
    return Object.keys(options).some((k) => k === 'recipient');
}
exports.isMint = isMint;
class ApproveAndCall {
    /**
     * Cannot be constructed.
     */
    constructor() { }
    static encodeApproveMax(token) {
        return ApproveAndCall.INTERFACE.encodeFunctionData('approveMax', [token.address]);
    }
    static encodeApproveMaxMinusOne(token) {
        return ApproveAndCall.INTERFACE.encodeFunctionData('approveMaxMinusOne', [token.address]);
    }
    static encodeApproveZeroThenMax(token) {
        return ApproveAndCall.INTERFACE.encodeFunctionData('approveZeroThenMax', [token.address]);
    }
    static encodeApproveZeroThenMaxMinusOne(token) {
        return ApproveAndCall.INTERFACE.encodeFunctionData('approveZeroThenMaxMinusOne', [token.address]);
    }
    static encodeCallPositionManager(calldatas) {
        (0, tiny_invariant_1.default)(calldatas.length > 0, 'NULL_CALLDATA');
        if (calldatas.length === 1) {
            return ApproveAndCall.INTERFACE.encodeFunctionData('callPositionManager', calldatas);
        }
        else {
            const encodedMulticall = v3_sdk_1.NonfungiblePositionManager.INTERFACE.encodeFunctionData('multicall', [calldatas]);
            return ApproveAndCall.INTERFACE.encodeFunctionData('callPositionManager', [encodedMulticall]);
        }
    }
    /**
     * Encode adding liquidity to a position in the nft manager contract
     * @param position Forcasted position with expected amount out from swap
     * @param minimalPosition Forcasted position with custom minimal token amounts
     * @param addLiquidityOptions Options for adding liquidity
     * @param slippageTolerance Defines maximum slippage
     */
    static encodeAddLiquidity(position, minimalPosition, addLiquidityOptions, slippageTolerance) {
        let { amount0: amount0Min, amount1: amount1Min } = position.mintAmountsWithSlippage(slippageTolerance);
        // position.mintAmountsWithSlippage() can create amounts not dependenable in scenarios
        // such as range orders. Allow the option to provide a position with custom minimum amounts
        // for these scenarios
        if (jsbi_1.default.lessThan(minimalPosition.amount0.quotient, amount0Min)) {
            amount0Min = minimalPosition.amount0.quotient;
        }
        if (jsbi_1.default.lessThan(minimalPosition.amount1.quotient, amount1Min)) {
            amount1Min = minimalPosition.amount1.quotient;
        }
        if (isMint(addLiquidityOptions)) {
            return ApproveAndCall.INTERFACE.encodeFunctionData('mint', [
                {
                    token0: position.pool.token0.address,
                    token1: position.pool.token1.address,
                    fee: position.pool.fee,
                    tickLower: position.tickLower,
                    tickUpper: position.tickUpper,
                    amount0Min: (0, v3_sdk_1.toHex)(amount0Min),
                    amount1Min: (0, v3_sdk_1.toHex)(amount1Min),
                    recipient: addLiquidityOptions.recipient,
                },
            ]);
        }
        else {
            return ApproveAndCall.INTERFACE.encodeFunctionData('increaseLiquidity', [
                {
                    token0: position.pool.token0.address,
                    token1: position.pool.token1.address,
                    amount0Min: (0, v3_sdk_1.toHex)(amount0Min),
                    amount1Min: (0, v3_sdk_1.toHex)(amount1Min),
                    tokenId: (0, v3_sdk_1.toHex)(addLiquidityOptions.tokenId),
                },
            ]);
        }
    }
    static encodeApprove(token, approvalType) {
        switch (approvalType) {
            case ApprovalTypes.MAX:
                return ApproveAndCall.encodeApproveMax(token.wrapped);
            case ApprovalTypes.MAX_MINUS_ONE:
                return ApproveAndCall.encodeApproveMaxMinusOne(token.wrapped);
            case ApprovalTypes.ZERO_THEN_MAX:
                return ApproveAndCall.encodeApproveZeroThenMax(token.wrapped);
            case ApprovalTypes.ZERO_THEN_MAX_MINUS_ONE:
                return ApproveAndCall.encodeApproveZeroThenMaxMinusOne(token.wrapped);
            default:
                throw new Error('Error: invalid ApprovalType');
        }
    }
}
exports.ApproveAndCall = ApproveAndCall;
ApproveAndCall.INTERFACE = new abi_1.Interface(IApproveAndCall_json_1.default.abi);
