// @uniswap/router-sdk/dist/cjs/src/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ONE_HUNDRED_PERCENT = exports.ZERO_PERCENT = exports.MIXED_QUOTER_V2_V4_FEE_PATH_PLACEHOLDER = exports.MIXED_QUOTER_V2_V3_FEE_PATH_PLACEHOLDER = exports.MIXED_QUOTER_V2_V2_FEE_PATH_PLACEHOLDER = exports.MIXED_QUOTER_V1_V2_FEE_PATH_PLACEHOLDER = exports.ONE = exports.ZERO = exports.ADDRESS_THIS = exports.MSG_SENDER = exports.ADDRESS_ZERO = void 0;
const tslib_1 = require("tslib");
const sdk_core_1 = require("@uniswap/sdk-core");
const jsbi_1 = tslib_1.__importDefault(require("jsbi"));
exports.ADDRESS_ZERO = '0x0000000000000000000000000000000000000000';
exports.MSG_SENDER = '0x0000000000000000000000000000000000000001';
exports.ADDRESS_THIS = '0x0000000000000000000000000000000000000002';
exports.ZERO = jsbi_1.default.BigInt(0);
exports.ONE = jsbi_1.default.BigInt(1);
// = 1 << 23 or 0b0100000000000000000000000
exports.MIXED_QUOTER_V1_V2_FEE_PATH_PLACEHOLDER = 1 << 23;
// = 10 << 4 or 0b00100000
exports.MIXED_QUOTER_V2_V2_FEE_PATH_PLACEHOLDER = 2 << 4;
// = 11 << 20 or 0b001100000000000000000000
exports.MIXED_QUOTER_V2_V3_FEE_PATH_PLACEHOLDER = 3 << 20;
// = 100 << 20 or 0b010000000000000000000000
exports.MIXED_QUOTER_V2_V4_FEE_PATH_PLACEHOLDER = 4 << 20;
exports.ZERO_PERCENT = new sdk_core_1.Percent(exports.ZERO);
exports.ONE_HUNDRED_PERCENT = new sdk_core_1.Percent(100, 100);
