// @uniswap/router-sdk/dist/cjs/src/paymentsExtended.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentsExtended = void 0;
const tslib_1 = require("tslib");
const abi_1 = require("@ethersproject/abi");
const sdk_core_1 = require("@uniswap/sdk-core");
const IPeripheryPaymentsWithFeeExtended_json_1 = tslib_1.__importDefault(require("@uniswap/swap-router-contracts/artifacts/contracts/interfaces/IPeripheryPaymentsWithFeeExtended.sol/IPeripheryPaymentsWithFeeExtended.json"));
const v3_sdk_1 = require("@uniswap/v3-sdk");
function encodeFeeBips(fee) {
    return (0, v3_sdk_1.toHex)(fee.multiply(10000).quotient);
}
class PaymentsExtended {
    /**
     * Cannot be constructed.
     */
    constructor() { }
    static encodeUnwrapWETH9(amountMinimum, recipient, feeOptions) {
        // if there's a recipient, just pass it along
        if (typeof recipient === 'string') {
            return v3_sdk_1.Payments.encodeUnwrapWETH9(amountMinimum, recipient, feeOptions);
        }
        if (!!feeOptions) {
            const feeBips = encodeFeeBips(feeOptions.fee);
            const feeRecipient = (0, sdk_core_1.validateAndParseAddress)(feeOptions.recipient);
            return PaymentsExtended.INTERFACE.encodeFunctionData('unwrapWETH9WithFee(uint256,uint256,address)', [
                (0, v3_sdk_1.toHex)(amountMinimum),
                feeBips,
                feeRecipient,
            ]);
        }
        else {
            return PaymentsExtended.INTERFACE.encodeFunctionData('unwrapWETH9(uint256)', [(0, v3_sdk_1.toHex)(amountMinimum)]);
        }
    }
    static encodeSweepToken(token, amountMinimum, recipient, feeOptions) {
        // if there's a recipient, just pass it along
        if (typeof recipient === 'string') {
            return v3_sdk_1.Payments.encodeSweepToken(token, amountMinimum, recipient, feeOptions);
        }
        if (!!feeOptions) {
            const feeBips = encodeFeeBips(feeOptions.fee);
            const feeRecipient = (0, sdk_core_1.validateAndParseAddress)(feeOptions.recipient);
            return PaymentsExtended.INTERFACE.encodeFunctionData('sweepTokenWithFee(address,uint256,uint256,address)', [
                token.address,
                (0, v3_sdk_1.toHex)(amountMinimum),
                feeBips,
                feeRecipient,
            ]);
        }
        else {
            return PaymentsExtended.INTERFACE.encodeFunctionData('sweepToken(address,uint256)', [
                token.address,
                (0, v3_sdk_1.toHex)(amountMinimum),
            ]);
        }
    }
    static encodePull(token, amount) {
        return PaymentsExtended.INTERFACE.encodeFunctionData('pull', [token.address, (0, v3_sdk_1.toHex)(amount)]);
    }
    static encodeWrapETH(amount) {
        return PaymentsExtended.INTERFACE.encodeFunctionData('wrapETH', [(0, v3_sdk_1.toHex)(amount)]);
    }
}
exports.PaymentsExtended = PaymentsExtended;
PaymentsExtended.INTERFACE = new abi_1.Interface(IPeripheryPaymentsWithFeeExtended_json_1.default.abi);
