// @uniswap/smart-order-router/build/main/providers/caching-subgraph-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CachingSubgraphProvider = exports.BASES_TO_CHECK_TRADES_AGAINST = void 0;
const sdk_core_1 = require("@uniswap/sdk-core");
const util_1 = require("../util");
const token_provider_1 = require("./token-provider");
exports.BASES_TO_CHECK_TRADES_AGAINST = {
    [sdk_core_1.ChainId.MAINNET]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.MAINNET),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.MAINNET],
        token_provider_1.DAI_MAINNET,
        token_provider_1.USDC_MAINNET,
        token_provider_1.USDT_MAINNET,
        token_provider_1.WBTC_MAINNET,
        token_provider_1.WSTETH_MAINNET,
    ],
    [sdk_core_1.ChainId.GOERLI]: [util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.GOERLI]],
    [sdk_core_1.ChainId.SEPOLIA]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.SEPOLIA),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.SEPOLIA],
    ],
    //v2 not deployed on [arbitrum, polygon, celo, gnosis, moonbeam, bnb, avalanche] and their testnets
    [sdk_core_1.ChainId.OPTIMISM]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.OPTIMISM),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.OPTIMISM],
        token_provider_1.USDC_OPTIMISM,
        token_provider_1.DAI_OPTIMISM,
        token_provider_1.USDT_OPTIMISM,
        token_provider_1.WBTC_OPTIMISM,
        token_provider_1.OP_OPTIMISM,
    ],
    [sdk_core_1.ChainId.ARBITRUM_ONE]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.ARBITRUM_ONE),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.ARBITRUM_ONE],
        token_provider_1.WBTC_ARBITRUM,
        token_provider_1.DAI_ARBITRUM,
        token_provider_1.USDC_ARBITRUM,
        token_provider_1.USDC_NATIVE_ARBITRUM,
        token_provider_1.USDT_ARBITRUM,
        token_provider_1.ARB_ARBITRUM,
    ],
    [sdk_core_1.ChainId.ARBITRUM_GOERLI]: [],
    [sdk_core_1.ChainId.ARBITRUM_SEPOLIA]: [],
    [sdk_core_1.ChainId.OPTIMISM_GOERLI]: [],
    [sdk_core_1.ChainId.OPTIMISM_SEPOLIA]: [],
    [sdk_core_1.ChainId.POLYGON]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.POLYGON),
        token_provider_1.USDC_POLYGON,
        token_provider_1.WETH_POLYGON,
        token_provider_1.WMATIC_POLYGON,
    ],
    [sdk_core_1.ChainId.POLYGON_MUMBAI]: [],
    [sdk_core_1.ChainId.CELO]: [token_provider_1.CELO, token_provider_1.CUSD_CELO, token_provider_1.CEUR_CELO, token_provider_1.DAI_CELO],
    [sdk_core_1.ChainId.CELO_ALFAJORES]: [],
    [sdk_core_1.ChainId.GNOSIS]: [],
    [sdk_core_1.ChainId.MOONBEAM]: [
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.MOONBEAM],
        token_provider_1.DAI_MOONBEAM,
        token_provider_1.USDC_MOONBEAM,
        token_provider_1.WBTC_MOONBEAM,
    ],
    [sdk_core_1.ChainId.BNB]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.BNB),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.BNB],
        token_provider_1.BUSD_BNB,
        token_provider_1.DAI_BNB,
        token_provider_1.USDC_BNB,
        token_provider_1.USDT_BNB,
        token_provider_1.BTC_BNB,
        token_provider_1.ETH_BNB,
    ],
    [sdk_core_1.ChainId.AVALANCHE]: [
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.AVALANCHE],
        token_provider_1.USDC_AVAX,
        token_provider_1.DAI_AVAX,
    ],
    [sdk_core_1.ChainId.BASE_GOERLI]: [],
    [sdk_core_1.ChainId.BASE]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.BASE),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.BASE],
        token_provider_1.USDC_BASE,
    ],
    [sdk_core_1.ChainId.ZORA]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.ZORA),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.ZORA],
    ],
    [sdk_core_1.ChainId.ZORA_SEPOLIA]: [util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.ZORA_SEPOLIA]],
    [sdk_core_1.ChainId.ROOTSTOCK]: [util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.ROOTSTOCK]],
    [sdk_core_1.ChainId.BLAST]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.BLAST),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.BLAST],
        token_provider_1.USDB_BLAST,
    ],
    [sdk_core_1.ChainId.ZKSYNC]: [
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.ZKSYNC],
        token_provider_1.USDCE_ZKSYNC,
        token_provider_1.USDC_ZKSYNC,
    ],
    [sdk_core_1.ChainId.WORLDCHAIN]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.WORLDCHAIN),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.WORLDCHAIN],
        token_provider_1.USDC_WORLDCHAIN,
        token_provider_1.WLD_WORLDCHAIN,
        token_provider_1.WBTC_WORLDCHAIN,
    ],
    [sdk_core_1.ChainId.UNICHAIN_SEPOLIA]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.UNICHAIN_SEPOLIA),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.UNICHAIN_SEPOLIA],
        token_provider_1.USDC_UNICHAIN_SEPOLIA,
    ],
    [sdk_core_1.ChainId.UNICHAIN]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.UNICHAIN),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.UNICHAIN],
        token_provider_1.DAI_UNICHAIN,
        token_provider_1.USDC_UNICHAIN,
    ],
    [sdk_core_1.ChainId.BASE_SEPOLIA]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.BASE_SEPOLIA),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.BASE_SEPOLIA],
        token_provider_1.USDC_BASE_SEPOLIA,
    ],
    [sdk_core_1.ChainId.MONAD_TESTNET]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.MONAD_TESTNET),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.MONAD_TESTNET],
        token_provider_1.USDT_MONAD_TESTNET,
    ],
    [sdk_core_1.ChainId.SONEIUM]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.SONEIUM),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.SONEIUM],
        token_provider_1.USDC_SONEIUM,
    ],
    [sdk_core_1.ChainId.MONAD]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.MONAD),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.MONAD],
        token_provider_1.USDC_MONAD,
    ],
    [sdk_core_1.ChainId.XLAYER]: [
        (0, util_1.nativeOnChain)(sdk_core_1.ChainId.XLAYER),
        util_1.WRAPPED_NATIVE_CURRENCY[sdk_core_1.ChainId.XLAYER],
        token_provider_1.USDC_XLAYER,
    ],
};
class CachingSubgraphProvider {
    /**
     * Creates an instance of CachingV3SubgraphProvider.
     * @param chainId The chain id to use.
     * @param subgraphProvider The provider to use to get the subgraph pools when not in the cache.
     * @param cache Cache instance to hold cached pools.
     * @param protocol Subgraph protocol version
     */
    constructor(chainId, subgraphProvider, cache, protocol) {
        this.chainId = chainId;
        this.subgraphProvider = subgraphProvider;
        this.cache = cache;
        this.protocol = protocol;
        this.SUBGRAPH_KEY = (chainId) => `subgraph-pools-${this.protocol}-${chainId}`;
    }
    async getPools() {
        const cachedPools = await this.cache.get(this.SUBGRAPH_KEY(this.chainId));
        if (cachedPools) {
            return cachedPools;
        }
        const pools = await this.subgraphProvider.getPools();
        await this.cache.set(this.SUBGRAPH_KEY(this.chainId), pools);
        return pools;
    }
}
exports.CachingSubgraphProvider = CachingSubgraphProvider;
