// @uniswap/smart-order-router/build/main/providers/subgraph-provider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubgraphProvider = exports.BASE_V4_PAGE_SIZE = exports.PAGE_SIZE = void 0;
const router_sdk_1 = require("@uniswap/router-sdk");
const sdk_core_1 = require("@uniswap/sdk-core");
const async_retry_1 = __importDefault(require("async-retry"));
const await_timeout_1 = __importDefault(require("await-timeout"));
const graphql_request_1 = require("graphql-request");
const lodash_1 = __importDefault(require("lodash"));
const util_1 = require("../util");
exports.PAGE_SIZE = 1000; // 1k is max possible query size from subgraph.
exports.BASE_V4_PAGE_SIZE = 3500; // TheGraph v4 base max pagesize is 3600.
class SubgraphProvider {
    constructor(protocol, chainId, retries = 2, timeout = 30000, rollback = true, trackedEthThreshold = 0.01, trackedZoraEthThreshold = 0.001, zoraHooks, 
    // @ts-expect-error - kept for backward compatibility
    untrackedUsdThreshold = Number.MAX_VALUE, subgraphUrl, bearerToken) {
        this.protocol = protocol;
        this.chainId = chainId;
        this.retries = retries;
        this.timeout = timeout;
        this.rollback = rollback;
        this.trackedEthThreshold = trackedEthThreshold;
        this.trackedZoraEthThreshold = trackedZoraEthThreshold;
        this.zoraHooks = zoraHooks;
        this.untrackedUsdThreshold = untrackedUsdThreshold;
        this.subgraphUrl = subgraphUrl;
        this.bearerToken = bearerToken;
        this.protocol = protocol;
        if (!this.subgraphUrl) {
            throw new Error(`No subgraph url for chain id: ${this.chainId}`);
        }
        util_1.log.info('bearerToken is', this.bearerToken);
        if (this.bearerToken) {
            this.client = new graphql_request_1.GraphQLClient(this.subgraphUrl, {
                headers: {
                    authorization: `Bearer ${this.bearerToken}`,
                },
            });
        }
        else {
            this.client = new graphql_request_1.GraphQLClient(this.subgraphUrl);
        }
        if (protocol === router_sdk_1.Protocol.V4 && this.zoraHooks.size === 0) {
            throw new Error('Zora hooks param is mandatory for V4');
        }
    }
    async getPools(_currencyIn, _currencyOut, providerConfig) {
        const beforeAll = Date.now();
        let blockNumber = (providerConfig === null || providerConfig === void 0 ? void 0 : providerConfig.blockNumber)
            ? await providerConfig.blockNumber
            : undefined;
        const pageSizeToUse = this.protocol === router_sdk_1.Protocol.V4 && this.chainId == sdk_core_1.ChainId.BASE
            ? exports.BASE_V4_PAGE_SIZE
            : exports.PAGE_SIZE;
        util_1.log.info(`Getting ${this.protocol} pools from the subgraph with page size ${pageSizeToUse}${(providerConfig === null || providerConfig === void 0 ? void 0 : providerConfig.blockNumber)
            ? ` as of block ${providerConfig === null || providerConfig === void 0 ? void 0 : providerConfig.blockNumber}`
            : ''}.`);
        // Define separate queries for each filtering condition
        const queries = [
            // 1. Pools with high tracked ETH (for both V3 and V4)
            {
                name: 'High tracked ETH pools',
                query: (0, graphql_request_1.gql) `
          query getHighTrackedETHPools($pageSize: Int!, $id: String, $threshold: String!) {
            pools(
              first: $pageSize
              ${blockNumber ? `block: { number: ${blockNumber} }` : ``}
              where: {
                id_gt: $id,
                totalValueLockedETH_gt: $threshold
              }
            ) {
              ${this.getPoolFields()}
            }
          }
        `,
                variables: { threshold: this.trackedEthThreshold.toString() },
            },
            // 2. V4: Non-Zora pools with liquidity > 0
            ...(this.protocol === router_sdk_1.Protocol.V4
                ? [
                    {
                        name: 'V4 non-Zora high liquidity pools',
                        query: (0, graphql_request_1.gql) `
          query getV4NonZoraHighLiquidityPools($pageSize: Int!, $id: String, $zoraHooks: [String!]!) {
            pools(
              first: $pageSize
              ${blockNumber ? `block: { number: ${blockNumber} }` : ``}
              where: {
                id_gt: $id,
                liquidity_gt: "0",
                hooks_not_in: $zoraHooks
              }
            ) {
              ${this.getPoolFields()}
            }
          }
        `,
                        variables: { zoraHooks: Array.from(this.zoraHooks) },
                    },
                    // 3. V4: Zora pools with liquidity > 0 AND TVL > trackedZoraEthThreshold
                    {
                        name: 'V4 Zora high liquidity pools',
                        query: (0, graphql_request_1.gql) `
          query getV4ZoraHighLiquidityPools($pageSize: Int!, $id: String, $zoraHooks: [String!]!, $zoraThreshold: String!) {
            pools(
              first: $pageSize
              ${blockNumber ? `block: { number: ${blockNumber} }` : ``}
              where: {
                id_gt: $id,
                liquidity_gt: "0",
                hooks_in: $zoraHooks,
                totalValueLockedETH_gt: $zoraThreshold
              }
            ) {
              ${this.getPoolFields()}
            }
          }
        `,
                        variables: {
                            zoraHooks: Array.from(this.zoraHooks),
                            zoraThreshold: this.trackedZoraEthThreshold.toString(),
                        },
                    },
                ]
                : []),
            // 4. V3: Pools with liquidity > 0 AND totalValueLockedETH = 0 (special V3 condition)
            ...(this.protocol === router_sdk_1.Protocol.V3
                ? [
                    {
                        name: 'V3 zero ETH pools',
                        query: (0, graphql_request_1.gql) `
          query getV3ZeroETHPools($pageSize: Int!, $id: String) {
            pools(
              first: $pageSize
              ${blockNumber ? `block: { number: ${blockNumber} }` : ``}
              where: {
                id_gt: $id,
                liquidity_gt: "0",
                totalValueLockedETH: "0"
              }
            ) {
              ${this.getPoolFields()}
            }
          }
        `,
                        variables: {},
                    },
                ]
                : []),
        ];
        let allPools = [];
        let retries = 0;
        await (0, async_retry_1.default)(async () => {
            const timeout = new await_timeout_1.default();
            const fetchPoolsForQuery = async (queryConfig) => {
                let lastId = '';
                let pools = [];
                let poolsPage = [];
                let totalPages = 0;
                do {
                    totalPages += 1;
                    const start = Date.now();
                    util_1.log.info(`Starting fetching for ${queryConfig.name} page ${totalPages} with page size ${pageSizeToUse}`);
                    const poolsResult = await this.client.request(queryConfig.query, Object.assign({ pageSize: pageSizeToUse, id: lastId }, queryConfig.variables));
                    poolsPage = poolsResult.pools;
                    pools = pools.concat(poolsPage);
                    if (pools.length > 0) {
                        lastId = pools[pools.length - 1].id;
                    }
                    util_1.metric.putMetric(`${this.protocol}SubgraphProvider.chain_${this.chainId}.getPools.${queryConfig.name
                        .replace(/\s+/g, '_')
                        .toLowerCase()}.paginate.pageSize`, poolsPage.length);
                    util_1.log.info(`Fetched ${poolsPage.length} pools for ${queryConfig.name} in ${Date.now() - start}ms`);
                } while (poolsPage.length > 0);
                util_1.metric.putMetric(`${this.protocol}SubgraphProvider.chain_${this.chainId}.getPools.${queryConfig.name
                    .replace(/\s+/g, '_')
                    .toLowerCase()}.paginate`, totalPages);
                util_1.metric.putMetric(`${this.protocol}SubgraphProvider.chain_${this.chainId}.getPools.${queryConfig.name
                    .replace(/\s+/g, '_')
                    .toLowerCase()}.pools.length`, pools.length);
                return pools;
            };
            try {
                // Fetch pools for each query in parallel
                const poolPromises = queries.map((queryConfig) => fetchPoolsForQuery(queryConfig));
                const allPoolsArrays = await Promise.all(poolPromises);
                // Merge all results and deduplicate by pool ID
                const poolMap = new Map();
                allPoolsArrays.forEach((pools) => {
                    pools.forEach((pool) => {
                        poolMap.set(pool.id, pool);
                    });
                });
                allPools = Array.from(poolMap.values());
                const getPoolsPromise = Promise.resolve(allPools);
                const timerPromise = timeout.set(this.timeout).then(() => {
                    throw new Error(`Timed out getting pools from subgraph: ${this.timeout}`);
                });
                allPools = await Promise.race([getPoolsPromise, timerPromise]);
                return;
            }
            catch (err) {
                util_1.log.error({ err }, `Error fetching ${this.protocol} Subgraph Pools.`);
                throw err;
            }
            finally {
                timeout.clear();
            }
        }, {
            retries: this.retries,
            onRetry: (err, retry) => {
                retries += 1;
                if (this.rollback &&
                    blockNumber &&
                    lodash_1.default.includes(err.message, 'indexed up to')) {
                    util_1.metric.putMetric(`${this.protocol}SubgraphProvider.chain_${this.chainId}.getPools.indexError`, 1);
                    blockNumber = blockNumber - 10;
                    util_1.log.info(`Detected subgraph indexing error. Rolled back block number to: ${blockNumber}`);
                }
                util_1.metric.putMetric(`${this.protocol}SubgraphProvider.chain_${this.chainId}.getPools.timeout`, 1);
                allPools = [];
                util_1.log.info({ err }, `Failed to get pools from subgraph. Retry attempt: ${retry}`);
            },
        });
        util_1.metric.putMetric(`${this.protocol}SubgraphProvider.chain_${this.chainId}.getPools.retries`, retries);
        const beforeFilter = Date.now();
        let poolsSanitized = [];
        if (this.protocol === router_sdk_1.Protocol.V3) {
            // Special treatment for all V3 pools in order to reduce latency due to thousands of pools with very low TVL locked
            // - Include "parseFloat(pool.totalValueLockedETH) === 0" as in certain occasions we have no way of calculating derivedETH so this is 0
            poolsSanitized = allPools
                .filter((pool) => (parseInt(pool.liquidity) > 0 &&
                parseFloat(pool.totalValueLockedETH) === 0) ||
                parseFloat(pool.totalValueLockedETH) > this.trackedEthThreshold)
                .map((pool) => {
                return this.mapSubgraphPool(pool);
            });
        }
        else if (this.protocol === router_sdk_1.Protocol.V4) {
            // For V4, apply additional filtering as a safety measure even though queries are optimized
            poolsSanitized = allPools
                .filter((pool) => {
                const liquidity = parseInt(pool.liquidity);
                const tvl = parseFloat(pool.totalValueLockedETH);
                const hooks = pool.hooks;
                const isZora = this.zoraHooks.has(hooks);
                if (isZora) {
                    return tvl > this.trackedZoraEthThreshold;
                }
                return liquidity > 0 || tvl > this.trackedEthThreshold;
            })
                .map((pool) => {
                return this.mapSubgraphPool(pool);
            });
        }
        util_1.metric.putMetric(`${this.protocol}SubgraphProvider.chain_${this.chainId}.getPools.filter.latency`, Date.now() - beforeFilter);
        util_1.metric.putMetric(`${this.protocol}SubgraphProvider.chain_${this.chainId}.getPools.filter.length`, poolsSanitized.length);
        util_1.metric.putMetric(`${this.protocol}SubgraphProvider.chain_${this.chainId}.getPools.filter.percent`, (poolsSanitized.length / allPools.length) * 100);
        util_1.metric.putMetric(`${this.protocol}SubgraphProvider.chain_${this.chainId}.getPools`, 1);
        util_1.metric.putMetric(`${this.protocol}SubgraphProvider.chain_${this.chainId}.getPools.latency`, Date.now() - beforeAll);
        util_1.log.info(`Got ${allPools.length} ${this.protocol} pools from the subgraph (after deduplication). ${poolsSanitized.length} after filtering`);
        return poolsSanitized;
    }
    // Helper method to get the pool fields for GraphQL queries
    getPoolFields() {
        return `
      id
      token0 {
        symbol
        id
      }
      token1 {
        symbol
        id
      }
      feeTier
      liquidity
      totalValueLockedUSD
      totalValueLockedETH
    `;
    }
}
exports.SubgraphProvider = SubgraphProvider;
