// @uniswap/universal-router-sdk/dist/esm/src/swapRouter.js
import invariant from 'tiny-invariant';
import UniversalRouter from '@uniswap/universal-router/artifacts/contracts/UniversalRouter.sol/UniversalRouter.json';
import { Interface } from '@ethersproject/abi';
import { BigNumber } from 'ethers';
import { Multicall, NonfungiblePositionManager as V3PositionManager, } from '@uniswap/v3-sdk';
import { V4PositionManager, Pool as V4Pool, URVersion, } from '@uniswap/v4-sdk';
import { Percent, CHAIN_TO_ADDRESSES_MAP } from '@uniswap/sdk-core';
import { UniswapTrade, TokenTransferMode } from './entities/actions/uniswap';
import { RoutePlanner, CommandType } from './utils/routerCommands';
import { encodePermit, encodeV3PositionPermit } from './utils/inputTokens';
import { UNIVERSAL_ROUTER_ADDRESS, UniversalRouterVersion } from './utils/constants';
import { getUniversalRouterDomain, EXECUTE_SIGNED_TYPES, generateNonce } from './utils/eip712';
function isMint(options) {
    return Object.keys(options).some((k) => k === 'recipient');
}
const UR_VERSION_MAP = {
    [URVersion.V2_0]: UniversalRouterVersion.V2_0,
    [URVersion.V2_1_1]: UniversalRouterVersion.V2_1_1,
};
function toUniversalRouterVersion(urVersion) {
    if (urVersion === undefined)
        return undefined;
    const mapped = UR_VERSION_MAP[urVersion];
    if (!mapped)
        throw new Error(`Unsupported URVersion: ${urVersion}`);
    return mapped;
}
export class SwapRouter {
    static swapCallParameters(trades, options, bridgeOptions // Optional bridge parameters
    ) {
        // TODO: use permit if signature included in swapOptions
        const planner = new RoutePlanner();
        const trade = new UniswapTrade(trades, options);
        const inputCurrency = trade.trade.inputAmount.currency;
        if (options.tokenTransferMode === TokenTransferMode.ApproveProxy) {
            invariant(!inputCurrency.isNative, 'PROXY_NATIVE_INPUT: SwapProxy only supports ERC20 input');
            invariant(!!options.chainId, 'PROXY_MISSING_CHAIN_ID: chainId required when tokenTransferMode is ApproveProxy');
            invariant(!options.inputTokenPermit, 'PROXY_PERMIT_CONFLICT: Permit2 not used with SwapProxy');
        }
        else {
            invariant(!(inputCurrency.isNative && !!options.inputTokenPermit), 'NATIVE_INPUT_PERMIT');
            if (options.inputTokenPermit) {
                encodePermit(planner, options.inputTokenPermit);
            }
        }
        const nativeCurrencyValue = inputCurrency.isNative
            ? BigNumber.from(trade.trade.maximumAmountIn(options.slippageTolerance).quotient.toString())
            : BigNumber.from(0);
        trade.encode(planner, { allowRevert: false });
        // Add bridge commands if provided
        if (bridgeOptions) {
            for (const bridge of bridgeOptions) {
                planner.addAcrossBridge(bridge);
            }
        }
        if (options.tokenTransferMode === TokenTransferMode.ApproveProxy) {
            return SwapRouter.encodeProxyPlan(planner, trade, options);
        }
        return SwapRouter.encodePlan(planner, nativeCurrencyValue, {
            deadline: options.deadlineOrPreviousBlockhash ? BigNumber.from(options.deadlineOrPreviousBlockhash) : undefined,
        });
    }
    /**
     * Generate EIP712 payload for signed execution (no signing performed)
     * Decodes existing execute() calldata and prepares it for signing
     *
     * @param calldata The calldata from swapCallParameters() or similar
     * @param signedOptions Options for signed execution (intent, data, sender, nonce)
     * @param deadline The deadline timestamp
     * @param chainId The chain ID
     * @param routerAddress The Universal Router contract address
     * @returns EIP712 payload ready to be signed externally
     */
    static getExecuteSignedPayload(calldata, signedOptions, deadline, chainId, routerAddress) {
        // Decode the execute() calldata to extract commands and inputs
        // Try to decode with deadline first, then without
        let decoded;
        let commands;
        let inputs;
        try {
            decoded = SwapRouter.INTERFACE.decodeFunctionData('execute(bytes,bytes[],uint256)', calldata);
            commands = decoded.commands;
            inputs = decoded.inputs;
        }
        catch (e) {
            // Try without deadline
            decoded = SwapRouter.INTERFACE.decodeFunctionData('execute(bytes,bytes[])', calldata);
            commands = decoded.commands;
            inputs = decoded.inputs;
        }
        // Use provided nonce or generate random one
        const nonce = signedOptions.nonce || generateNonce();
        // sender is provided directly (address(0) = skip verification)
        const sender = signedOptions.sender;
        const domain = getUniversalRouterDomain(chainId, routerAddress);
        const intent = signedOptions.intent;
        const data = signedOptions.data;
        const deadlineStr = BigNumber.from(deadline).toString();
        const value = {
            commands,
            inputs,
            intent,
            data,
            sender,
            nonce,
            deadline: deadlineStr,
        };
        return {
            domain,
            types: EXECUTE_SIGNED_TYPES,
            value,
        };
    }
    /**
     * Encode executeSigned() call with signature
     *
     * @param calldata The original calldata from swapCallParameters()
     * @param signature The signature obtained from external signing
     * @param signedOptions The same options used in getExecuteSignedPayload()
     * @param deadline The deadline timestamp
     * @param nativeCurrencyValue The native currency value (ETH) to send
     * @returns Method parameters for executeSigned()
     */
    static encodeExecuteSigned(calldata, signature, signedOptions, deadline, nativeCurrencyValue = BigNumber.from(0)) {
        // Decode the execute() calldata to extract commands and inputs
        // Try to decode with deadline first, then without
        let decoded;
        let commands;
        let inputs;
        try {
            decoded = SwapRouter.INTERFACE.decodeFunctionData('execute(bytes,bytes[],uint256)', calldata);
            commands = decoded.commands;
            inputs = decoded.inputs;
        }
        catch (e) {
            // Try without deadline
            decoded = SwapRouter.INTERFACE.decodeFunctionData('execute(bytes,bytes[])', calldata);
            commands = decoded.commands;
            inputs = decoded.inputs;
        }
        // Use provided nonce (must match what was signed)
        // Nonce must match what was signed - require it to be provided
        if (!signedOptions.nonce) {
            throw new Error('Nonce is required for encodeExecuteSigned - use the nonce from getExecuteSignedPayload');
        }
        const nonce = signedOptions.nonce;
        // Determine verifySender based on sender address
        const verifySender = signedOptions.sender !== '0x0000000000000000000000000000000000000000';
        // Encode executeSigned function call using the Universal Router v2.1 ABI
        const signedCalldata = SwapRouter.INTERFACE.encodeFunctionData('executeSigned', [
            commands,
            inputs,
            signedOptions.intent,
            signedOptions.data,
            verifySender,
            nonce,
            signature,
            deadline,
        ]);
        return { calldata: signedCalldata, value: nativeCurrencyValue.toHexString() };
    }
    /**
     * Builds the call parameters for a migration from a V3 position to a V4 position.
     * Some requirements of the parameters:
     *   - v3RemoveLiquidityOptions.collectOptions.recipient must equal v4PositionManager
     *   - v3RemoveLiquidityOptions.liquidityPercentage must be 100%
     *   - input pool and output pool must have the same tokens
     *   - V3 NFT must be approved, or valid inputV3NFTPermit must be provided with UR as spender
     */
    static migrateV3ToV4CallParameters(options, positionManagerOverride) {
        const v4Pool = options.outputPosition.pool;
        const v3Token0 = options.inputPosition.pool.token0;
        const v3Token1 = options.inputPosition.pool.token1;
        const v4PositionManagerAddress = positionManagerOverride !== null && positionManagerOverride !== void 0 ? positionManagerOverride : CHAIN_TO_ADDRESSES_MAP[v4Pool.chainId].v4PositionManagerAddress;
        // owner of the v3 nft must be the receiver of the v4 nft
        // validate the parameters
        if (v4Pool.currency0.isNative) {
            invariant((v4Pool.currency0.wrapped.equals(v3Token0) && v4Pool.currency1.equals(v3Token1)) ||
                (v4Pool.currency0.wrapped.equals(v3Token1) && v4Pool.currency1.equals(v3Token0)), 'TOKEN_MISMATCH');
        }
        else {
            invariant(v3Token0 === v4Pool.token0, 'TOKEN0_MISMATCH');
            invariant(v3Token1 === v4Pool.token1, 'TOKEN1_MISMATCH');
        }
        invariant(options.v3RemoveLiquidityOptions.liquidityPercentage.equalTo(new Percent(100, 100)), 'FULL_REMOVAL_REQUIRED');
        invariant(options.v3RemoveLiquidityOptions.burnToken == true, 'BURN_TOKEN_REQUIRED');
        invariant(options.v3RemoveLiquidityOptions.collectOptions.recipient === v4PositionManagerAddress, 'RECIPIENT_NOT_POSITION_MANAGER');
        invariant(isMint(options.v4AddLiquidityOptions), 'MINT_REQUIRED');
        invariant(options.v4AddLiquidityOptions.migrate, 'MIGRATE_REQUIRED');
        const planner = new RoutePlanner();
        // to prevent reentrancy by the pool hook, we initialize the v4 pool before moving funds
        if (options.v4AddLiquidityOptions.createPool) {
            const poolKey = V4Pool.getPoolKey(v4Pool.currency0, v4Pool.currency1, v4Pool.fee, v4Pool.tickSpacing, v4Pool.hooks);
            planner.addCommand(CommandType.V4_INITIALIZE_POOL, [poolKey, v4Pool.sqrtRatioX96.toString()]);
            // remove createPool setting, so that it doesnt get encoded again later
            delete options.v4AddLiquidityOptions.createPool;
        }
        // add position permit to the universal router planner
        if (options.v3RemoveLiquidityOptions.permit) {
            // permit spender should be UR
            const universalRouterAddress = UNIVERSAL_ROUTER_ADDRESS(UniversalRouterVersion.V2_0, options.inputPosition.pool.chainId);
            invariant(universalRouterAddress == options.v3RemoveLiquidityOptions.permit.spender, 'INVALID_SPENDER');
            // don't need to transfer it because v3posm uses isApprovedOrOwner()
            encodeV3PositionPermit(planner, options.v3RemoveLiquidityOptions.permit, options.v3RemoveLiquidityOptions.tokenId);
            // remove permit so that multicall doesnt add it again
            delete options.v3RemoveLiquidityOptions.permit;
        }
        // encode v3 withdraw
        const v3RemoveParams = V3PositionManager.removeCallParameters(options.inputPosition, options.v3RemoveLiquidityOptions);
        const v3Calls = Multicall.decodeMulticall(v3RemoveParams.calldata);
        for (const v3Call of v3Calls) {
            // slice selector - 0x + 4 bytes = 10 characters
            const selector = v3Call.slice(0, 10);
            invariant(selector == V3PositionManager.INTERFACE.getSighash('collect') ||
                selector == V3PositionManager.INTERFACE.getSighash('decreaseLiquidity') ||
                selector == V3PositionManager.INTERFACE.getSighash('burn'), 'INVALID_V3_CALL: ' + selector);
            planner.addCommand(CommandType.V3_POSITION_MANAGER_CALL, [v3Call]);
        }
        // encode v4 mint
        const v4AddParams = V4PositionManager.addCallParameters(options.outputPosition, options.v4AddLiquidityOptions);
        // only modifyLiquidities can be called by the UniversalRouter
        const selector = v4AddParams.calldata.slice(0, 10);
        invariant(selector == V4PositionManager.INTERFACE.getSighash('modifyLiquidities'), 'INVALID_V4_CALL: ' + selector);
        planner.addCommand(CommandType.V4_POSITION_MANAGER_CALL, [v4AddParams.calldata]);
        return SwapRouter.encodePlan(planner, BigNumber.from(0), {
            deadline: BigNumber.from(options.v4AddLiquidityOptions.deadline),
        });
    }
    /**
     * Encodes a planned route into a method name and parameters for the Router contract.
     * @param planner the planned route
     * @param nativeCurrencyValue the native currency value of the planned route
     * @param config the router config
     */
    static encodePlan(planner, nativeCurrencyValue, config = {}) {
        const { commands, inputs } = planner;
        const functionSignature = !!config.deadline ? 'execute(bytes,bytes[],uint256)' : 'execute(bytes,bytes[])';
        const parameters = !!config.deadline ? [commands, inputs, config.deadline] : [commands, inputs];
        const calldata = SwapRouter.INTERFACE.encodeFunctionData(functionSignature, parameters);
        return { calldata, value: nativeCurrencyValue.toHexString() };
    }
    /**
     * Encodes a planned route into calldata targeting the SwapProxy contract.
     * The proxy pulls ERC20 tokens from the user into the UR, then executes commands.
     */
    static encodeProxyPlan(planner, trade, options) {
        var _a;
        const { commands, inputs } = planner;
        const urVersion = (_a = toUniversalRouterVersion(options.urVersion)) !== null && _a !== void 0 ? _a : UniversalRouterVersion.V2_0;
        const routerAddress = UNIVERSAL_ROUTER_ADDRESS(urVersion, options.chainId);
        const inputToken = trade.trade.inputAmount.currency.address;
        const inputAmount = BigNumber.from(trade.trade.maximumAmountIn(options.slippageTolerance).quotient.toString());
        const deadline = options.deadlineOrPreviousBlockhash
            ? BigNumber.from(options.deadlineOrPreviousBlockhash)
            : BigNumber.from(Math.floor(Date.now() / 1000) + 1800); // 30 min default
        const calldata = SwapRouter.PROXY_INTERFACE.encodeFunctionData('execute', [
            routerAddress,
            inputToken,
            inputAmount,
            commands,
            inputs,
            deadline,
        ]);
        return { calldata, value: BigNumber.from(0).toHexString() };
    }
}
SwapRouter.INTERFACE = new Interface(UniversalRouter.abi);
SwapRouter.PROXY_INTERFACE = new Interface([
    'function execute(address router, address token, uint256 amount, bytes calldata commands, bytes[] calldata inputs, uint256 deadline) external',
]);
