// @0glabs/0g-serving-broker/lib.commonjs/inference/broker/lora.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoRAProcessor = void 0;
const tslib_1 = require("tslib");
const axios_1 = tslib_1.__importDefault(require("axios"));
const adapter_name_1 = require("../../common/utils/adapter-name");
const ADAPTER_POLL_INTERVAL_MS = 3000;
const DEPLOY_POLL_INTERVAL_MS = 2000;
const DEFAULT_DEPLOY_TIMEOUT_SECONDS = 120;
class LoRAProcessor {
    deps;
    constructor(deps) {
        this.deps = deps;
    }
    async getBrokerBaseUrl(providerAddress) {
        const endpoint = await this.deps.getEndpoint(providerAddress);
        return endpoint.replace(/\/v1\/proxy$/, '');
    }
    async resolveAdapterName(providerAddress, taskId, baseModel) {
        const localName = (0, adapter_name_1.makeAdapterName)(baseModel, taskId);
        const baseUrl = await this.getBrokerBaseUrl(providerAddress);
        try {
            const headers = await this.deps.getHeaders(providerAddress);
            const resp = await axios_1.default.get(`${baseUrl}/v1/lora/adapters`, {
                headers,
            });
            const adapters = resp.data?.adapters || [];
            const match = adapters.find((a) => a.taskId === taskId);
            if (match?.adapterName) {
                return match.adapterName;
            }
        }
        catch {
            // Fall back to locally generated name
        }
        return localName;
    }
    async listAdapters(providerAddress) {
        const baseUrl = await this.getBrokerBaseUrl(providerAddress);
        const headers = await this.deps.getHeaders(providerAddress);
        const resp = await axios_1.default.get(`${baseUrl}/v1/lora/adapters`, {
            headers,
        });
        return resp.data?.adapters || [];
    }
    async getAdapterStatus(providerAddress, adapterName) {
        const baseUrl = await this.getBrokerBaseUrl(providerAddress);
        const headers = await this.deps.getHeaders(providerAddress);
        const resp = await axios_1.default.get(`${baseUrl}/v1/lora/adapters/${adapterName}`, { headers });
        return resp.data;
    }
    async deployAdapter(providerAddress, baseModel, taskId, options = {}) {
        const { wait = false, timeoutSeconds = DEFAULT_DEPLOY_TIMEOUT_SECONDS, onProgress, } = options;
        const baseUrl = await this.getBrokerBaseUrl(providerAddress);
        const deadline = wait ? Date.now() + timeoutSeconds * 1000 : 0;
        let adapterName = await this.resolveAdapterName(providerAddress, taskId, baseModel);
        const localName = (0, adapter_name_1.makeAdapterName)(baseModel, taskId);
        let nameResolved = adapterName !== localName;
        if (wait) {
            let lastState = '';
            while (Date.now() < deadline) {
                if (!nameResolved) {
                    try {
                        const resolved = await this.resolveAdapterName(providerAddress, taskId, baseModel);
                        if (resolved !== localName) {
                            adapterName = resolved;
                            nameResolved = true;
                        }
                    }
                    catch {
                        // Broker may not have processed event yet
                    }
                }
                try {
                    const status = await this.getAdapterStatus(providerAddress, adapterName);
                    const state = status.state;
                    if (state && state !== lastState) {
                        lastState = state;
                        onProgress?.(state);
                    }
                    if (state === 'ready' || state === 'active')
                        break;
                    if (state === 'failed')
                        break;
                }
                catch {
                    // Not found yet
                }
                await new Promise((r) => setTimeout(r, ADAPTER_POLL_INTERVAL_MS));
            }
            if (lastState === 'failed') {
                throw new Error('Adapter preparation failed. Check broker logs for details.');
            }
            if (Date.now() >= deadline &&
                lastState !== 'ready' &&
                lastState !== 'active') {
                throw new Error(`Timed out after ${timeoutSeconds}s waiting for adapter to be ready (last state: ${lastState || 'not found'})`);
            }
        }
        // If already active, skip deploy
        try {
            const status = await this.getAdapterStatus(providerAddress, adapterName);
            if (status.state === 'active') {
                return { message: 'Adapter is already deployed and active!' };
            }
        }
        catch {
            // Not found, proceed
        }
        // Call deploy API
        const authHeaders = await this.deps.getHeaders(providerAddress);
        const deployResp = await axios_1.default.post(`${baseUrl}/v1/lora/adapters/deploy`, { taskId, baseModel }, { headers: authHeaders });
        const result = {
            message: deployResp.data?.message || 'Deploy request sent',
            adapterName,
        };
        if (wait) {
            while (Date.now() < deadline) {
                try {
                    const status = await this.getAdapterStatus(providerAddress, adapterName);
                    if (status.state === 'active') {
                        result.message =
                            'Adapter deployed successfully! You can now chat with it.';
                        return result;
                    }
                    if (status.state === 'failed') {
                        throw new Error('Adapter deployment failed.');
                    }
                    onProgress?.(status.state);
                }
                catch (err) {
                    if (err instanceof Error &&
                        err.message.includes('failed')) {
                        throw err;
                    }
                }
                await new Promise((r) => setTimeout(r, DEPLOY_POLL_INTERVAL_MS));
            }
            throw new Error(`Timed out after ${timeoutSeconds}s waiting for deployment to complete.`);
        }
        return result;
    }
    async deployAdapterByName(providerAddress, adapterName, options = {}) {
        const { wait = false, timeoutSeconds = DEFAULT_DEPLOY_TIMEOUT_SECONDS, onProgress, } = options;
        const baseUrl = await this.getBrokerBaseUrl(providerAddress);
        const deadline = wait ? Date.now() + timeoutSeconds * 1000 : 0;
        // If already active, skip deploy
        try {
            const status = await this.getAdapterStatus(providerAddress, adapterName);
            if (status.state === 'active') {
                return { message: 'Adapter is already deployed and active!' };
            }
        }
        catch {
            // Not found, proceed
        }
        const authHeaders = await this.deps.getHeaders(providerAddress);
        const deployResp = await axios_1.default.post(`${baseUrl}/v1/lora/adapters/deploy`, { adapterName }, { headers: authHeaders });
        const result = {
            message: deployResp.data?.message || 'Deploy request sent',
            adapterName,
        };
        if (wait) {
            while (Date.now() < deadline) {
                try {
                    const status = await this.getAdapterStatus(providerAddress, adapterName);
                    if (status.state === 'active') {
                        result.message =
                            'Adapter deployed successfully! You can now chat with it.';
                        return result;
                    }
                    if (status.state === 'failed') {
                        throw new Error('Adapter deployment failed.');
                    }
                    onProgress?.(status.state);
                }
                catch (err) {
                    if (err instanceof Error &&
                        err.message.includes('failed')) {
                        throw err;
                    }
                }
                await new Promise((r) => setTimeout(r, DEPLOY_POLL_INTERVAL_MS));
            }
            throw new Error(`Timed out after ${timeoutSeconds}s waiting for deployment to complete.`);
        }
        return result;
    }
    async chat(providerAddress, adapterName, message, options = {}) {
        const { systemPrompt = 'You are a helpful assistant.' } = options;
        const endpoint = await this.deps.getEndpoint(providerAddress);
        const messages = [
            { role: 'system', content: systemPrompt },
            { role: 'user', content: message },
        ];
        const headers = await this.deps.getHeaders(providerAddress, JSON.stringify({ model: adapterName, messages }));
        const resp = await axios_1.default.post(`${endpoint}/chat/completions`, { model: adapterName, messages }, {
            headers: {
                'Content-Type': 'application/json',
                ...headers,
            },
        });
        return resp.data;
    }
}
exports.LoRAProcessor = LoRAProcessor;
