// @0glabs/0g-serving-broker/lib.commonjs/inference/broker/model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelProcessor = exports.isVerifiability = exports.VerifiabilityEnum = void 0;
const utils_1 = require("../../common/utils");
const read_only_model_1 = require("./read-only-model");
const read_only_model_2 = require("./read-only-model");
Object.defineProperty(exports, "VerifiabilityEnum", { enumerable: true, get: function () { return read_only_model_2.VerifiabilityEnum; } });
Object.defineProperty(exports, "isVerifiability", { enumerable: true, get: function () { return read_only_model_2.isVerifiability; } });
/**
 * Authenticated model processor extending read-only functionality
 * Adds provider-only operations like removeService and updateService
 *
 * Inherits from ReadOnlyModelProcessor:
 * - listService() - List all services
 * - listServiceWithDetail() - List services with health metrics
 */
class ModelProcessor extends read_only_model_1.ReadOnlyModelProcessor {
    ledger;
    metadata;
    cache;
    constructor(contract, ledger, metadata, cache) {
        super(contract);
        this.contract = contract;
        this.ledger = ledger;
        this.metadata = metadata;
        this.cache = cache;
    }
    // Note: listService() and listServiceWithDetail() are inherited from ReadOnlyModelProcessor
    /**
     * Remove service (Provider owner only)
     *
     * This function allows the provider owner to remove their service from the contract.
     *
     * @param {number} gasPrice - Optional gas price for the transaction.
     * @throws Will throw an error if the caller is not the service owner or if removal fails.
     */
    async removeService(gasPrice) {
        try {
            const txOptions = {};
            if (gasPrice) {
                txOptions.gasPrice = gasPrice;
            }
            await this.contract.sendTx('removeService', [], txOptions);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    /**
     * Update service (Provider owner only)
     *
     * This function allows the provider owner to update their existing service.
     * All parameters are optional - if not provided, the current value is preserved.
     *
     * @param options - Update options
     * @param options.url - New service URL
     * @param options.model - New model name
     * @param options.inputPrice - New input price (in neuron, the smallest unit)
     * @param options.outputPrice - New output price (in neuron, the smallest unit)
     * @param options.gasPrice - Optional gas price for the transaction
     * @throws Will throw an error if the caller is not the service owner or if update fails.
     */
    async updateService(options) {
        try {
            // Get current service to preserve unchanged fields
            const userAddress = this.contract.getUserAddress();
            const currentService = await this.contract.getService(userAddress);
            if (!currentService || !currentService.provider) {
                throw new Error('Service not found for the current provider');
            }
            // Build ServiceParams with updated values (use new value if provided, otherwise keep current)
            const params = {
                serviceType: currentService.serviceType,
                url: options.url ?? currentService.url,
                model: options.model ?? currentService.model,
                verifiability: currentService.verifiability,
                inputPrice: options.inputPrice ?? currentService.inputPrice,
                outputPrice: options.outputPrice ?? currentService.outputPrice,
                additionalInfo: currentService.additionalInfo,
                teeSignerAddress: currentService.teeSignerAddress,
            };
            const txOptions = {};
            if (options.gasPrice) {
                txOptions.gasPrice = options.gasPrice;
            }
            await this.contract.sendTx('addOrUpdateService', [params], txOptions);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
}
exports.ModelProcessor = ModelProcessor;
