// @0glabs/0g-serving-broker/lib.commonjs/inference/broker/read-only-broker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReadOnlyInferenceBroker = void 0;
exports.createReadOnlyInferenceBroker = createReadOnlyInferenceBroker;
const ethers_1 = require("ethers");
const contract_1 = require("../contract");
const read_only_model_1 = require("./read-only-model");
const constants_1 = require("../../constants");
/**
 * Read-only inference broker with operations that don't require authentication
 * Can be used with a Provider (read-only) without wallet connection
 *
 * Use this broker to:
 * - List available AI providers before connecting wallet
 * - Get service information and health metrics
 * - Browse services without authentication
 */
class ReadOnlyInferenceBroker {
    modelProcessor;
    contractAddress;
    constructor(provider, contractAddress) {
        this.contractAddress = contractAddress;
        const contract = new contract_1.ReadOnlyInferenceServingContract(provider, contractAddress);
        this.modelProcessor = new read_only_model_1.ReadOnlyModelProcessor(contract);
    }
    /**
     * Retrieves a list of services from the contract.
     * This is a read-only operation that doesn't require authentication.
     *
     * @param {number} offset - The offset for pagination (default: 0).
     * @param {number} limit - The limit for pagination (default: 50).
     * @param {boolean} includeUnacknowledged - Whether to include providers whose TEE signer is not acknowledged (default: false).
     * @returns {Promise<ServiceStructOutput[]>} A promise that resolves to an array of ServiceStructOutput objects.
     * @throws An error if the service list cannot be retrieved.
     */
    async listService(offset = 0, limit = 50, includeUnacknowledged = false) {
        return this.modelProcessor.listService(offset, limit, includeUnacknowledged);
    }
    /**
     * Retrieves a list of services with detailed health metrics from the monitoring API.
     * This is a read-only operation that doesn't require authentication.
     *
     * This method combines on-chain service data with real-time health metrics including
     * uptime percentage and average response time (latency) for each service provider.
     *
     * @param {number} offset - The offset for pagination (default: 0).
     * @param {number} limit - The limit for pagination (default: 50).
     * @param {boolean} includeUnacknowledged - Whether to include providers whose TEE signer is not acknowledged (default: false).
     * @returns {Promise<ServiceWithDetail[]>} A promise that resolves to an array of ServiceWithDetail objects containing both blockchain and health data.
     * @throws An error if the service list cannot be retrieved.
     *
     * @example
     * ```typescript
     * const servicesWithHealth = await broker.listServiceWithDetail();
     * servicesWithHealth.forEach(service => {
     *   console.log(`Provider: ${service.provider}`);
     *   if (service.healthMetrics) {
     *     console.log(`  Uptime: ${service.healthMetrics.uptime}%`);
     *     console.log(`  Latency: ${service.healthMetrics.avgResponseTime}ms`);
     *   }
     * });
     * ```
     */
    async listServiceWithDetail(offset = 0, limit = 50, includeUnacknowledged = false) {
        return this.modelProcessor.listServiceWithDetail(offset, limit, includeUnacknowledged);
    }
}
exports.ReadOnlyInferenceBroker = ReadOnlyInferenceBroker;
/**
 * Factory function to create a read-only inference broker
 * No authentication required - perfect for listing providers without wallet connection
 *
 * @param rpcUrl - JSON-RPC endpoint URL (e.g., https://rpc-testnet.0g.ai)
 * @param chainId - Optional chain ID (auto-detected if not provided)
 * @returns Read-only broker instance
 *
 * @example
 * ```typescript
 * // Create read-only broker (no wallet needed)
 * const broker = await createReadOnlyInferenceBroker('https://rpc-testnet.0g.ai')
 *
 * // List all services
 * const services = await broker.listService()
 *
 * // List services with health details
 * const detailedServices = await broker.listServiceWithDetail()
 * ```
 */
async function createReadOnlyInferenceBroker(rpcUrl, chainId) {
    const provider = new ethers_1.JsonRpcProvider(rpcUrl);
    // Auto-detect chain ID if not provided
    const detectedChainId = chainId ?? Number((await provider.getNetwork()).chainId);
    // Get contract address based on chain ID
    let inferenceCA;
    if (detectedChainId === Number(constants_1.MAINNET_CHAIN_ID)) {
        inferenceCA = constants_1.CONTRACT_ADDRESSES.mainnet.inference;
        console.log(`Using mainnet inference contract (chain ID: ${detectedChainId})`);
    }
    else if (detectedChainId === Number(constants_1.TESTNET_CHAIN_ID)) {
        inferenceCA = (0, constants_1.isDevMode)()
            ? constants_1.CONTRACT_ADDRESSES.testnetDev.inference
            : constants_1.CONTRACT_ADDRESSES.testnet.inference;
        console.log(`Using testnet inference contract${(0, constants_1.isDevMode)() ? ' [DEV MODE]' : ''} (chain ID: ${detectedChainId})`);
    }
    else if (detectedChainId === Number(constants_1.HARDHAT_CHAIN_ID)) {
        inferenceCA = constants_1.CONTRACT_ADDRESSES.hardhat.inference;
        console.log(`Using hardhat inference contract (chain ID: ${detectedChainId})`);
    }
    else {
        // Default to testnet
        inferenceCA = constants_1.CONTRACT_ADDRESSES.testnet.inference;
        console.warn(`Unknown chain ID ${detectedChainId}, defaulting to testnet contract`);
    }
    // Create read-only broker
    return new ReadOnlyInferenceBroker(provider, inferenceCA);
}
