// @0glabs/0g-serving-broker/lib.commonjs/inference/broker/read-only-model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReadOnlyModelProcessor = exports.VerifiabilityEnum = void 0;
exports.parseTieredPricing = parseTieredPricing;
exports.parseTieredPricingFromModelInfo = parseTieredPricingFromModelInfo;
exports.parseCacheTokenBilling = parseCacheTokenBilling;
exports.parseCacheTokenBillingFromModelInfo = parseCacheTokenBillingFromModelInfo;
exports.isVerifiability = isVerifiability;
const tslib_1 = require("tslib");
const utils_1 = require("../../common/utils");
const axios_1 = tslib_1.__importDefault(require("axios"));
var VerifiabilityEnum;
(function (VerifiabilityEnum) {
    VerifiabilityEnum["OpML"] = "OpML";
    VerifiabilityEnum["TeeML"] = "TeeML";
    VerifiabilityEnum["ZKML"] = "ZKML";
})(VerifiabilityEnum || (exports.VerifiabilityEnum = VerifiabilityEnum = {}));
/**
 * Parse tiered pricing info from service additionalInfo JSON string.
 * Returns undefined if not present or invalid.
 */
function parseTieredPricing(additionalInfo) {
    try {
        const parsed = JSON.parse(additionalInfo);
        if (parsed?.tieredPricing?.tiers &&
            Array.isArray(parsed.tieredPricing.tiers)) {
            const tiers = parsed.tieredPricing.tiers
                .filter((t) => typeof t.maxInputTokens === 'number' &&
                typeof t.inputMultiplier === 'number' &&
                typeof t.outputMultiplier === 'number')
                .map((t) => ({
                maxInputTokens: t.maxInputTokens,
                inputMultiplier: t.inputMultiplier,
                outputMultiplier: t.outputMultiplier,
            }));
            if (tiers.length > 0) {
                return { tiers };
            }
        }
    }
    catch {
        // additionalInfo is not valid JSON or doesn't contain tieredPricing
    }
    return undefined;
}
/**
 * Parse tiered pricing from ProviderModelInfo (returned by /v1/models endpoint).
 * Returns undefined if not present.
 */
function parseTieredPricingFromModelInfo(modelInfo) {
    const raw = modelInfo?.pricing?.tiered_pricing;
    if (!raw || !Array.isArray(raw) || raw.length === 0) {
        return undefined;
    }
    const tiers = raw
        .filter((t) => typeof t.max_input_tokens === 'number' &&
        typeof t.input_multiplier === 'number' &&
        typeof t.output_multiplier === 'number')
        .map((t) => ({
        maxInputTokens: t.max_input_tokens,
        inputMultiplier: t.input_multiplier,
        outputMultiplier: t.output_multiplier,
    }));
    if (tiers.length > 0) {
        return { tiers };
    }
    return undefined;
}
/**
 * Parse cache token billing info from service additionalInfo JSON string.
 * Returns undefined if not present or invalid.
 */
function parseCacheTokenBilling(additionalInfo) {
    try {
        const parsed = JSON.parse(additionalInfo);
        if (parsed?.cacheTokenBilling &&
            typeof parsed.cacheTokenBilling.divisor === 'number' &&
            parsed.cacheTokenBilling.divisor > 0) {
            return { divisor: parsed.cacheTokenBilling.divisor };
        }
    }
    catch {
        // additionalInfo is not valid JSON or doesn't contain cacheTokenBilling
    }
    return undefined;
}
/**
 * Parse cache token billing from ProviderModelInfo (returned by /v1/models endpoint).
 * Returns undefined if not present.
 */
function parseCacheTokenBillingFromModelInfo(modelInfo) {
    const raw = modelInfo?.pricing?.cache_token_billing;
    if (raw && typeof raw.divisor === 'number' && raw.divisor > 0) {
        return { divisor: raw.divisor };
    }
    return undefined;
}
/**
 * Read-only model processor for listing services and fetching health metrics
 * Works without authentication - only requires a read-only contract
 */
class ReadOnlyModelProcessor {
    contract;
    constructor(contract) {
        this.contract = contract;
    }
    /**
     * List services from the blockchain
     *
     * @param offset - Pagination offset (default: 0)
     * @param limit - Pagination limit (default: 50)
     * @param includeUnacknowledged - Include unacknowledged services (default: false)
     * @returns Array of service struct outputs
     */
    async listService(offset = 0, limit = 50, includeUnacknowledged = false) {
        return this.contract.listService(offset, limit, includeUnacknowledged);
    }
    /**
     * Retrieves a list of services enriched with health metrics and model info from the status API.
     *
     * @param offset - The offset for pagination (default: 0)
     * @param limit - The limit for pagination (default: 50)
     * @param includeUnacknowledged - Whether to include providers whose TEE signer is not acknowledged (default: false)
     * @returns Promise that resolves to an array of ServiceWithDetail objects, each containing:
     *   - Blockchain service data (provider, model, pricing, verifiability, etc.)
     *   - `healthMetrics` — uptime, avg response time, and status (omitted if unavailable)
     *   - `modelInfo` — rich model metadata: context length, max completion tokens, tokenizer,
     *     TEE attestation details, supported parameters, pricing, and more (omitted if unavailable)
     * @throws An error if the service list cannot be retrieved
     *
     * @example
     * ```typescript
     * const services = await processor.listServiceWithDetail();
     * services.forEach(service => {
     *   console.log(`Provider: ${service.provider}`);
     *   if (service.healthMetrics) {
     *     console.log(`  Uptime: ${service.healthMetrics.uptime}%`);
     *     console.log(`  Latency: ${service.healthMetrics.avgResponseTime}ms`);
     *   }
     *   if (service.modelInfo) {
     *     console.log(`  Model: ${service.modelInfo.name}`);
     *     console.log(`  Context: ${service.modelInfo.context_length} tokens`);
     *     console.log(`  TEE Attested: ${service.modelInfo.tee_attested}`);
     *   }
     * });
     * ```
     */
    async listServiceWithDetail(offset = 0, limit = 50, includeUnacknowledged = false) {
        try {
            // Get services from blockchain
            const services = await this.listService(offset, limit, includeUnacknowledged);
            // Determine status API endpoint based on chain ID
            const chainId = await this.contract.getChainId();
            const statusApiEndpoint = this.getStatusApiEndpoint(chainId);
            // Fetch health metrics and aggregated model info from status API in parallel
            let healthMetrics = [];
            let allModels = [];
            await Promise.all([
                axios_1.default
                    .get(`${statusApiEndpoint}/health`, { timeout: 10000 })
                    .then((r) => {
                    healthMetrics = Array.isArray(r.data?.services)
                        ? r.data.services
                        : [];
                })
                    .catch(() => { }),
                axios_1.default
                    .get(`${statusApiEndpoint}/models`, { timeout: 10000 })
                    .then((r) => {
                    allModels = Array.isArray(r.data?.data)
                        ? r.data.data
                        : [];
                })
                    .catch(() => { }),
            ]);
            // Create a map of health metrics by provider address
            const healthMap = new Map();
            for (const metric of healthMetrics) {
                healthMap.set(metric.provider.toLowerCase(), metric);
            }
            // Create a map of model info by provider address
            const providerModelsMap = new Map();
            for (const model of allModels) {
                if (!model.provider)
                    continue;
                const key = model.provider.toLowerCase();
                const list = providerModelsMap.get(key) ?? [];
                list.push(model);
                providerModelsMap.set(key, list);
            }
            // Merge health metrics and model info with services
            // Note: Explicitly construct clean objects to avoid numeric indices from ethers Result type
            const servicesWithDetail = services.map((service) => {
                const health = healthMap.get(service.provider.toLowerCase());
                const providerModels = providerModelsMap.get(service.provider.toLowerCase()) ??
                    [];
                const modelInfo = providerModels.find((m) => m.id === service.model);
                return {
                    provider: service.provider,
                    serviceType: service.serviceType,
                    url: service.url,
                    inputPrice: service.inputPrice,
                    outputPrice: service.outputPrice,
                    updatedAt: service.updatedAt,
                    model: service.model,
                    verifiability: service.verifiability,
                    additionalInfo: service.additionalInfo,
                    teeSignerAddress: service.teeSignerAddress,
                    teeSignerAcknowledged: service.teeSignerAcknowledged,
                    healthMetrics: health
                        ? {
                            status: health.status,
                            uptime: health.checks.uptime,
                            avgResponseTime: health.performance.response_time?.avg ?? 0,
                            lastCheck: health.lastCheck,
                        }
                        : undefined,
                    modelInfo,
                    tieredPricing: parseTieredPricing(service.additionalInfo) ??
                        parseTieredPricingFromModelInfo(modelInfo),
                    cacheTokenBilling: parseCacheTokenBilling(service.additionalInfo) ??
                        parseCacheTokenBillingFromModelInfo(modelInfo),
                };
            });
            return servicesWithDetail;
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    /**
     * Get status API endpoint based on chain ID
     * @param chainId - The chain ID
     * @returns The status API endpoint URL
     */
    getStatusApiEndpoint(chainId) {
        // Mainnet: 16661n, Testnet: 16602n
        if (chainId === 16661n) {
            return 'https://compute-status.0g.ai';
        }
        else {
            // Default to testnet
            return 'https://compute-status-testnet.0g.ai';
        }
    }
}
exports.ReadOnlyModelProcessor = ReadOnlyModelProcessor;
function isVerifiability(value) {
    return Object.values(VerifiabilityEnum).includes(value);
}
