// @uniswap/universal-router-sdk/dist/cjs/src/entities/actions/uniswap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniswapTrade = exports.TokenTransferMode = void 0;
const routerCommands_1 = require("../../utils/routerCommands");
const v2_sdk_1 = require("@uniswap/v2-sdk");
const v3_sdk_1 = require("@uniswap/v3-sdk");
const v4_sdk_1 = require("@uniswap/v4-sdk");
const router_sdk_1 = require("@uniswap/router-sdk");
const pathCurrency_1 = require("../../utils/pathCurrency");
const sdk_core_1 = require("@uniswap/sdk-core");
const Command_1 = require("../Command");
const constants_1 = require("../../utils/constants");
const getCurrencyAddress_1 = require("../../utils/getCurrencyAddress");
const numbers_1 = require("../../utils/numbers");
const ethers_1 = require("ethers");
// the existing router permit object doesn't include enough data for permit2
// so we extend swap options with the permit2 permit
// when safe mode is enabled, the SDK will add an extra ETH sweep for security
// when useRouterBalance is enabled the SDK will use the balance in the router for the swap
var TokenTransferMode;
(function (TokenTransferMode) {
    TokenTransferMode["Permit2"] = "Permit2";
    TokenTransferMode["ApproveProxy"] = "ApproveProxy";
})(TokenTransferMode = exports.TokenTransferMode || (exports.TokenTransferMode = {}));
const REFUND_ETH_PRICE_IMPACT_THRESHOLD = new sdk_core_1.Percent(50, 100);
// Wrapper for uniswap router-sdk trade entity to encode swaps for Universal Router
// also translates trade objects from previous (v2, v3) SDKs
class UniswapTrade {
    constructor(trade, options) {
        this.trade = trade;
        this.options = options;
        this.tradeType = Command_1.RouterActionType.UniswapTrade;
        if (!!options.fee && !!options.flatFee)
            throw new Error('Only one fee option permitted');
        if (options.tokenTransferMode === TokenTransferMode.ApproveProxy) {
            if (!options.recipient || options.recipient === constants_1.SENDER_AS_RECIPIENT) {
                throw new Error('Explicit recipient address required when using SwapProxy (SENDER_AS_RECIPIENT resolves to proxy)');
            }
            this.payerIsUser = false;
        }
        else if (this.inputRequiresWrap || this.inputRequiresUnwrap || this.options.useRouterBalance) {
            this.payerIsUser = false;
        }
        else {
            this.payerIsUser = true;
        }
    }
    get isAllV4() {
        let result = true;
        for (const swap of this.trade.swaps) {
            result = result && swap.route.protocol == router_sdk_1.Protocol.V4;
        }
        return result;
    }
    // this.trade.swaps is an array of swaps / trades.
    // we are iterating over one swap (trade) at a time so length is 1
    // route is either v2, v3, v4, or mixed
    // pathInput and pathOutput are the currencies of the input and output of the route
    // this.trade.inputAmount is the input currency of the trade (could be different from pathInput)
    // this.trade.outputAmount is the output currency of the trade (could be different from pathOutput)
    // each route can have multiple pools
    get inputRequiresWrap() {
        const swap = this.trade.swaps[0];
        const route = swap.route;
        const firstPool = route.pools[0];
        if (firstPool instanceof v4_sdk_1.Pool) {
            // If first pool is a v4 pool and input currency is native and the path input currency in the route is not native, we need to wrap.
            return (this.trade.inputAmount.currency.isNative &&
                !this.trade.swaps[0].route.pathInput.isNative);
        }
        // If first pool is not a v4 pool and input currency is native, we need to wrap
        return this.trade.inputAmount.currency.isNative;
    }
    get inputRequiresUnwrap() {
        const swap = this.trade.swaps[0];
        const route = swap.route;
        const firstPool = route.pools[0];
        if (firstPool instanceof v4_sdk_1.Pool) {
            // If the first pool is a v4 pool and input currency is not native and the path input currency is native, we need to unwrap
            return (!this.trade.inputAmount.currency.isNative &&
                this.trade.swaps[0].route.pathInput.isNative);
        }
        // If the first pool is not a v4 pool, we don't need to unwrap.
        return false;
    }
    get outputRequiresWrap() {
        const swap = this.trade.swaps[0];
        const lastRoute = swap.route;
        const lastPool = lastRoute.pools[lastRoute.pools.length - 1];
        // if last pool is v4:
        if (lastPool instanceof v4_sdk_1.Pool) {
            // If output currency is not native but path currency output is native, we need to wrap
            if (!this.trade.outputAmount.currency.isNative) {
                if (lastRoute.pathOutput.isNative) {
                    // this means path output is native and we need to wrap
                    return true;
                }
                else if (lastPool.currency1.equals(lastPool.currency0.wrapped) && lastRoute.pools.length > 1) {
                    let poolBefore = lastRoute.pools[lastRoute.pools.length - 2];
                    // this means last pool is eth-weth and pool before contains weth
                    if (poolBefore instanceof v4_sdk_1.Pool &&
                        (poolBefore.currency0.equals(lastPool.currency1) || poolBefore.currency1.equals(lastPool.currency1))) {
                        return true;
                    }
                    else if (poolBefore.token0.equals(lastPool.currency1) || poolBefore.token1.equals(lastPool.currency1)) {
                        // same for v2 and v3 pools
                        return true;
                    }
                }
            }
        }
        // if last pool is not v4:
        // we do not need to wrap because v2 and v3 pools already require wrapped tokens
        return false;
    }
    get outputRequiresUnwrap() {
        const swap = this.trade.swaps[0];
        const lastRoute = swap.route;
        const lastPool = lastRoute.pools[lastRoute.pools.length - 1];
        // if last pool is v4:
        if (lastPool instanceof v4_sdk_1.Pool) {
            // If output currency is native and path currency output is not native, we need to unwrap
            if (this.trade.outputAmount.currency.isNative) {
                if (!this.trade.swaps[0].route.pathOutput.isNative) {
                    // this means path output is weth and we need to unwrap
                    return true;
                }
                else if (lastRoute.pools.length > 1 &&
                    lastRoute.pools[lastRoute.pools.length - 2] instanceof v4_sdk_1.Pool &&
                    lastRoute.pools[lastRoute.pools.length - 2].currency0.isNative &&
                    lastPool.currency1.equals(lastPool.currency0.wrapped)) {
                    // this means last pool is eth-weth and we need to unwrap
                    return true;
                }
                else {
                    return false;
                }
            }
        }
        // else: if path output currency is native, we need to unwrap because v2 and v3 pools already require wrapped tokens
        return this.trade.outputAmount.currency.isNative;
    }
    get outputRequiresTransition() {
        return this.outputRequiresWrap || this.outputRequiresUnwrap;
    }
    encode(planner, _config) {
        var _a;
        // If the input currency is the native currency, we need to wrap it with the router as the recipient
        if (this.inputRequiresWrap) {
            // TODO: optimize if only one v2 pool we can directly send this to the pool
            planner.addCommand(routerCommands_1.CommandType.WRAP_ETH, [
                constants_1.ROUTER_AS_RECIPIENT,
                this.trade.maximumAmountIn(this.options.slippageTolerance).quotient.toString(),
            ]);
        }
        else if (this.inputRequiresUnwrap) {
            if (this.options.tokenTransferMode !== TokenTransferMode.ApproveProxy) {
                // send wrapped token to router to unwrap via Permit2
                planner.addCommand(routerCommands_1.CommandType.PERMIT2_TRANSFER_FROM, [
                    this.trade.inputAmount.currency.address,
                    constants_1.ROUTER_AS_RECIPIENT,
                    this.trade.maximumAmountIn(this.options.slippageTolerance).quotient.toString(),
                ]);
            }
            // In proxy mode, the proxy already transferred tokens to the UR; just unwrap
            planner.addCommand(routerCommands_1.CommandType.UNWRAP_WETH, [constants_1.ROUTER_AS_RECIPIENT, 0]);
        }
        // The overall recipient at the end of the trade, SENDER_AS_RECIPIENT uses the msg.sender
        this.options.recipient = (_a = this.options.recipient) !== null && _a !== void 0 ? _a : constants_1.SENDER_AS_RECIPIENT;
        // flag for whether we want to perform slippage check on aggregate output of multiple routes
        //   1. when there are >2 exact input trades. this is only a heuristic,
        //      as it's still more gas-expensive even in this case, but has benefits
        //      in that the reversion probability is lower
        const performAggregatedSlippageCheck = this.trade.tradeType === sdk_core_1.TradeType.EXACT_INPUT && this.trade.routes.length > 2;
        const routerMustCustody = performAggregatedSlippageCheck || this.outputRequiresTransition || hasFeeOption(this.options);
        for (const swap of this.trade.swaps) {
            switch (swap.route.protocol) {
                case router_sdk_1.Protocol.V2:
                    addV2Swap(planner, swap, this.trade.tradeType, this.options, this.payerIsUser, routerMustCustody);
                    break;
                case router_sdk_1.Protocol.V3:
                    addV3Swap(planner, swap, this.trade.tradeType, this.options, this.payerIsUser, routerMustCustody);
                    break;
                case router_sdk_1.Protocol.V4:
                    addV4Swap(planner, swap, this.trade.tradeType, this.options, this.payerIsUser, routerMustCustody);
                    break;
                case router_sdk_1.Protocol.MIXED:
                    addMixedSwap(planner, swap, this.trade.tradeType, this.options, this.payerIsUser, routerMustCustody);
                    break;
                default:
                    throw new Error('UNSUPPORTED_TRADE_PROTOCOL');
            }
        }
        let minimumAmountOut = ethers_1.BigNumber.from(this.trade.minimumAmountOut(this.options.slippageTolerance).quotient.toString());
        // The router custodies for 3 reasons: to unwrap, to take a fee, and/or to do a slippage check
        if (routerMustCustody) {
            const pools = this.trade.swaps[0].route.pools;
            const pathOutputCurrencyAddress = (0, getCurrencyAddress_1.getCurrencyAddress)((0, pathCurrency_1.getPathCurrency)(this.trade.outputAmount.currency, pools[pools.length - 1]));
            let feeDeduction = ethers_1.BigNumber.from(0);
            // If there is a fee, that percentage is sent to the fee recipient
            // In the case where ETH is the output currency, the fee is taken in WETH (for gas reasons)
            if (!!this.options.fee) {
                // UR >= V2_1_1 supports PAY_PORTION_FULL_PRECISION (1e18 precision),
                // older versions only support PAY_PORTION (bips)
                const useFullPrecision = this.options.urVersion !== undefined && this.options.urVersion >= v4_sdk_1.URVersion.V2_1_1;
                // Reject fractional bips fees on older UR versions to prevent silent precision loss
                if (!useFullPrecision && !this.options.fee.fee.multiply(10000).remainder.equalTo(0)) {
                    throw new Error('Fractional fee bips require Universal Router version V2_1_1 or higher');
                }
                if (useFullPrecision) {
                    const fee1e18 = (0, numbers_1.encodeFee1e18)(this.options.fee.fee);
                    planner.addCommand(routerCommands_1.CommandType.PAY_PORTION_FULL_PRECISION, [pathOutputCurrencyAddress, this.options.fee.recipient, fee1e18], false, this.options.urVersion);
                    feeDeduction = minimumAmountOut.mul(fee1e18).div(ethers_1.BigNumber.from(10).pow(18));
                }
                else {
                    const feeBips = (0, numbers_1.encodeFeeBips)(this.options.fee.fee);
                    planner.addCommand(routerCommands_1.CommandType.PAY_PORTION, [pathOutputCurrencyAddress, this.options.fee.recipient, feeBips]);
                    feeDeduction = minimumAmountOut.mul(feeBips).div(10000);
                }
            }
            // If there is a flat fee, that absolute amount is sent to the fee recipient
            // In the case where ETH is the output currency, the fee is taken in WETH (for gas reasons)
            if (!!this.options.flatFee) {
                const feeAmount = this.options.flatFee.amount;
                if (minimumAmountOut.lt(feeAmount))
                    throw new Error('Flat fee amount greater than minimumAmountOut');
                planner.addCommand(routerCommands_1.CommandType.TRANSFER, [pathOutputCurrencyAddress, this.options.flatFee.recipient, feeAmount]);
                feeDeduction = ethers_1.BigNumber.from(feeAmount);
            }
            // If the trade is exact output, and a fee was taken, we must adjust the amount out to be the amount after the fee
            // Otherwise we continue as expected with the trade's normal expected output
            if (this.trade.tradeType === sdk_core_1.TradeType.EXACT_OUTPUT) {
                minimumAmountOut = minimumAmountOut.sub(feeDeduction);
            }
            // The remaining tokens that need to be sent to the user after the fee is taken will be caught
            // by this if-else clause.
            if (this.outputRequiresUnwrap) {
                planner.addCommand(routerCommands_1.CommandType.UNWRAP_WETH, [this.options.recipient, minimumAmountOut]);
            }
            else if (this.outputRequiresWrap) {
                planner.addCommand(routerCommands_1.CommandType.WRAP_ETH, [this.options.recipient, constants_1.CONTRACT_BALANCE]);
            }
            else {
                planner.addCommand(routerCommands_1.CommandType.SWEEP, [
                    (0, getCurrencyAddress_1.getCurrencyAddress)(this.trade.outputAmount.currency),
                    this.options.recipient,
                    minimumAmountOut,
                ]);
            }
        }
        // for exactOutput swaps with native input or that perform an inputToken transition (wrap or unwrap)
        // we need to send back the change to the user
        if (this.trade.tradeType === sdk_core_1.TradeType.EXACT_OUTPUT || riskOfPartialFill(this.trade)) {
            if (this.inputRequiresWrap) {
                planner.addCommand(routerCommands_1.CommandType.UNWRAP_WETH, [this.options.recipient, 0]);
            }
            else if (this.inputRequiresUnwrap) {
                planner.addCommand(routerCommands_1.CommandType.WRAP_ETH, [this.options.recipient, constants_1.CONTRACT_BALANCE]);
            }
            else if (this.options.tokenTransferMode === TokenTransferMode.ApproveProxy) {
                // Proxy pulled maximumAmountIn into the UR; sweep any unused input back to the user
                planner.addCommand(routerCommands_1.CommandType.SWEEP, [
                    (0, getCurrencyAddress_1.getCurrencyAddress)(this.trade.inputAmount.currency),
                    this.options.recipient,
                    0,
                ]);
            }
            else if (this.trade.inputAmount.currency.isNative) {
                // must refund extra native currency sent along for native v4 trades (no input transition)
                planner.addCommand(routerCommands_1.CommandType.SWEEP, [constants_1.ETH_ADDRESS, this.options.recipient, 0]);
            }
        }
        if (this.options.safeMode)
            planner.addCommand(routerCommands_1.CommandType.SWEEP, [constants_1.ETH_ADDRESS, this.options.recipient, 0]);
    }
}
exports.UniswapTrade = UniswapTrade;
// encode a uniswap v2 swap
function addV2Swap(planner, { route, inputAmount, outputAmount, maxHopSlippage }, tradeType, options, payerIsUser, routerMustCustody) {
    if ((maxHopSlippage === null || maxHopSlippage === void 0 ? void 0 : maxHopSlippage.length) && maxHopSlippage.length !== route.pools.length) {
        throw new Error(`maxHopSlippage length (${maxHopSlippage.length}) must equal route.pools.length (${route.pools.length})`);
    }
    const trade = new v2_sdk_1.Trade(route, tradeType == sdk_core_1.TradeType.EXACT_INPUT ? inputAmount : outputAmount, tradeType);
    const useV2_1_1 = (0, v4_sdk_1.isAtLeastV2_1_1)(options.urVersion);
    if (tradeType == sdk_core_1.TradeType.EXACT_INPUT) {
        const params = [
            // if native, we have to unwrap so keep in the router for now
            routerMustCustody ? constants_1.ROUTER_AS_RECIPIENT : options.recipient,
            trade.maximumAmountIn(options.slippageTolerance).quotient.toString(),
            // if router will custody funds, we do aggregated slippage check from router
            routerMustCustody ? 0 : trade.minimumAmountOut(options.slippageTolerance).quotient.toString(),
            route.path.map((token) => token.wrapped.address),
            payerIsUser,
        ];
        if (useV2_1_1)
            params.push(maxHopSlippage !== null && maxHopSlippage !== void 0 ? maxHopSlippage : []);
        planner.addCommand(routerCommands_1.CommandType.V2_SWAP_EXACT_IN, params, false, options.urVersion);
    }
    else if (tradeType == sdk_core_1.TradeType.EXACT_OUTPUT) {
        const params = [
            routerMustCustody ? constants_1.ROUTER_AS_RECIPIENT : options.recipient,
            trade.minimumAmountOut(options.slippageTolerance).quotient.toString(),
            trade.maximumAmountIn(options.slippageTolerance).quotient.toString(),
            route.path.map((token) => token.wrapped.address),
            payerIsUser,
        ];
        if (useV2_1_1)
            params.push(maxHopSlippage !== null && maxHopSlippage !== void 0 ? maxHopSlippage : []);
        planner.addCommand(routerCommands_1.CommandType.V2_SWAP_EXACT_OUT, params, false, options.urVersion);
    }
}
// encode a uniswap v3 swap
function addV3Swap(planner, { route, inputAmount, outputAmount, maxHopSlippage }, tradeType, options, payerIsUser, routerMustCustody) {
    if ((maxHopSlippage === null || maxHopSlippage === void 0 ? void 0 : maxHopSlippage.length) && maxHopSlippage.length !== route.pools.length) {
        throw new Error(`maxHopSlippage length (${maxHopSlippage.length}) must equal route.pools.length (${route.pools.length})`);
    }
    const trade = v3_sdk_1.Trade.createUncheckedTrade({
        route: route,
        inputAmount,
        outputAmount,
        tradeType,
    });
    const useV2_1_1 = (0, v4_sdk_1.isAtLeastV2_1_1)(options.urVersion);
    const path = (0, v3_sdk_1.encodeRouteToPath)(route, trade.tradeType === sdk_core_1.TradeType.EXACT_OUTPUT);
    if (tradeType == sdk_core_1.TradeType.EXACT_INPUT) {
        const params = [
            routerMustCustody ? constants_1.ROUTER_AS_RECIPIENT : options.recipient,
            trade.maximumAmountIn(options.slippageTolerance).quotient.toString(),
            routerMustCustody ? 0 : trade.minimumAmountOut(options.slippageTolerance).quotient.toString(),
            path,
            payerIsUser,
        ];
        if (useV2_1_1)
            params.push(maxHopSlippage !== null && maxHopSlippage !== void 0 ? maxHopSlippage : []);
        planner.addCommand(routerCommands_1.CommandType.V3_SWAP_EXACT_IN, params, false, options.urVersion);
    }
    else if (tradeType == sdk_core_1.TradeType.EXACT_OUTPUT) {
        const params = [
            routerMustCustody ? constants_1.ROUTER_AS_RECIPIENT : options.recipient,
            trade.minimumAmountOut(options.slippageTolerance).quotient.toString(),
            trade.maximumAmountIn(options.slippageTolerance).quotient.toString(),
            path,
            payerIsUser,
        ];
        if (useV2_1_1)
            params.push(maxHopSlippage !== null && maxHopSlippage !== void 0 ? maxHopSlippage : []);
        planner.addCommand(routerCommands_1.CommandType.V3_SWAP_EXACT_OUT, params, false, options.urVersion);
    }
}
function addV4Swap(planner, { inputAmount, outputAmount, route, maxHopSlippage }, tradeType, options, payerIsUser, routerMustCustody) {
    var _a, _b;
    if ((maxHopSlippage === null || maxHopSlippage === void 0 ? void 0 : maxHopSlippage.length) && maxHopSlippage.length !== route.pools.length) {
        throw new Error(`maxHopSlippage length (${maxHopSlippage.length}) must equal route.pools.length (${route.pools.length})`);
    }
    // create a deep copy of pools since v4Planner encoding tampers with array
    const pools = route.pools.map((p) => p);
    const v4Route = new v4_sdk_1.Route(pools, inputAmount.currency, outputAmount.currency);
    const trade = v4_sdk_1.Trade.createUncheckedTrade({
        route: v4Route,
        inputAmount,
        outputAmount,
        tradeType,
    });
    const slippageToleranceOnSwap = routerMustCustody && tradeType == sdk_core_1.TradeType.EXACT_INPUT ? undefined : options.slippageTolerance;
    const perHopSlippage = (_a = maxHopSlippage === null || maxHopSlippage === void 0 ? void 0 : maxHopSlippage.map((s) => ethers_1.BigNumber.from(s))) !== null && _a !== void 0 ? _a : [];
    const v4Planner = new v4_sdk_1.V4Planner();
    v4Planner.addTrade(trade, slippageToleranceOnSwap, perHopSlippage, options.urVersion);
    v4Planner.addSettle(trade.route.pathInput, payerIsUser);
    // Handle split route output consistency:
    // - If output is ETH and some routes output WETH: force all to output WETH, then unwrap
    // - If output is WETH and some routes output ETH: force all to output ETH, then wrap
    let pathOutputForTake = trade.route.pathOutput;
    let lastPool = v4Route.pools[v4Route.pools.length - 1];
    let ethWethPool = lastPool.currency1.equals(lastPool.currency0.wrapped);
    if (ethWethPool && v4Route.pools.length > 1) {
        let poolBefore = v4Route.pools[v4Route.pools.length - 2];
        if (pathOutputForTake.isNative && poolBefore.currency0.isNative) {
            pathOutputForTake = pathOutputForTake.wrapped;
        }
        else if (!pathOutputForTake.isNative &&
            (poolBefore.currency0.equals(lastPool.currency1) || poolBefore.currency1.equals(lastPool.currency1))) {
            pathOutputForTake = lastPool.currency0;
        }
    }
    v4Planner.addTake(pathOutputForTake, routerMustCustody ? constants_1.ROUTER_AS_RECIPIENT : (_b = options.recipient) !== null && _b !== void 0 ? _b : constants_1.SENDER_AS_RECIPIENT);
    planner.addCommand(routerCommands_1.CommandType.V4_SWAP, [v4Planner.finalize()]);
}
// encode a mixed route swap, i.e. including both v2 and v3 pools
function addMixedSwap(planner, swap, tradeType, options, payerIsUser, routerMustCustody) {
    var _a, _b, _c, _d;
    const route = swap.route;
    if (((_a = swap.maxHopSlippage) === null || _a === void 0 ? void 0 : _a.length) && swap.maxHopSlippage.length !== route.pools.length) {
        throw new Error(`maxHopSlippage length (${swap.maxHopSlippage.length}) must equal route.pools.length (${route.pools.length})`);
    }
    const inputAmount = swap.inputAmount;
    const outputAmount = swap.outputAmount;
    const tradeRecipient = routerMustCustody ? constants_1.ROUTER_AS_RECIPIENT : (_b = options.recipient) !== null && _b !== void 0 ? _b : constants_1.SENDER_AS_RECIPIENT;
    // single hop, so it can be reduced to plain swap logic for one protocol version
    if (route.pools.length === 1) {
        if (route.pools[0] instanceof v4_sdk_1.Pool) {
            return addV4Swap(planner, swap, tradeType, options, payerIsUser, routerMustCustody);
        }
        else if (route.pools[0] instanceof v3_sdk_1.Pool) {
            return addV3Swap(planner, swap, tradeType, options, payerIsUser, routerMustCustody);
        }
        else if (route.pools[0] instanceof v2_sdk_1.Pair) {
            return addV2Swap(planner, swap, tradeType, options, payerIsUser, routerMustCustody);
        }
        else {
            throw new Error('Invalid route type');
        }
    }
    const trade = router_sdk_1.MixedRouteTrade.createUncheckedTrade({
        route: route,
        inputAmount,
        outputAmount,
        tradeType,
    });
    const amountIn = trade.maximumAmountIn(options.slippageTolerance, inputAmount).quotient.toString();
    const amountOut = routerMustCustody
        ? 0
        : trade.minimumAmountOut(options.slippageTolerance, outputAmount).quotient.toString();
    // logic from
    // https://github.com/Uniswap/router-sdk/blob/d8eed164e6c79519983844ca8b6a3fc24ebcb8f8/src/swapRouter.ts#L276
    const sections = (0, router_sdk_1.partitionMixedRouteByProtocol)(route);
    const isLastSectionInRoute = (i) => {
        return i === sections.length - 1;
    };
    const useV2_1_1 = (0, v4_sdk_1.isAtLeastV2_1_1)(options.urVersion);
    let inputToken = route.pathInput;
    let hopOffset = 0;
    for (let i = 0; i < sections.length; i++) {
        const section = sections[i];
        const routePool = section[0];
        const outputToken = (0, router_sdk_1.getOutputOfPools)(section, inputToken);
        const subRoute = new router_sdk_1.MixedRoute(new router_sdk_1.MixedRouteSDK([...section], inputToken, outputToken));
        // Slice this section's portion of maxHopSlippage from the flat array
        const sectionHopSlippage = (_c = swap.maxHopSlippage) === null || _c === void 0 ? void 0 : _c.slice(hopOffset, hopOffset + section.length);
        let nextInputToken;
        let swapRecipient;
        if (isLastSectionInRoute(i)) {
            nextInputToken = outputToken;
            swapRecipient = tradeRecipient;
        }
        else {
            const nextPool = sections[i + 1][0];
            nextInputToken = (0, pathCurrency_1.getPathCurrency)(outputToken, nextPool);
            const v2PoolIsSwapRecipient = nextPool instanceof v2_sdk_1.Pair && outputToken.equals(nextInputToken);
            swapRecipient = v2PoolIsSwapRecipient ? nextPool.liquidityToken.address : constants_1.ROUTER_AS_RECIPIENT;
        }
        if (routePool instanceof v4_sdk_1.Pool) {
            const v4Planner = new v4_sdk_1.V4Planner();
            const v4SubRoute = new v4_sdk_1.Route(section, subRoute.input, subRoute.output);
            const v4SectionSlippage = (_d = sectionHopSlippage === null || sectionHopSlippage === void 0 ? void 0 : sectionHopSlippage.map((s) => ethers_1.BigNumber.from(s))) !== null && _d !== void 0 ? _d : [];
            v4Planner.addSettle(inputToken, payerIsUser && i === 0, (i == 0 ? amountIn : constants_1.CONTRACT_BALANCE));
            v4Planner.addAction(v4_sdk_1.Actions.SWAP_EXACT_IN, [
                {
                    currencyIn: inputToken.isNative ? constants_1.ETH_ADDRESS : inputToken.address,
                    path: (0, v4_sdk_1.encodeRouteToPath)(v4SubRoute),
                    maxHopSlippage: v4SectionSlippage,
                    amountIn: 0,
                    amountOutMinimum: !isLastSectionInRoute(i) ? 0 : amountOut,
                },
            ], options.urVersion);
            // Handle split route output consistency for V4 sections in mixed routes
            let outputTokenForTake = outputToken;
            if (isLastSectionInRoute(i)) {
                let lastPool = route.pools[route.pools.length - 1];
                let v4Pool = lastPool instanceof v4_sdk_1.Pool;
                let ethWethPool = v4Pool ? lastPool.currency1.equals(lastPool.currency0.wrapped) : false;
                let poolBefore = route.pools[route.pools.length - 2];
                if (ethWethPool) {
                    if (outputToken.isNative && poolBefore.token0.isNative) {
                        outputTokenForTake = outputToken.wrapped;
                    }
                    else if (!outputToken.isNative &&
                        (poolBefore.token0.equals(lastPool.token1) || poolBefore.token1.equals(lastPool.token1))) {
                        outputTokenForTake = lastPool.token0;
                    }
                }
            }
            v4Planner.addTake(outputTokenForTake, swapRecipient);
            planner.addCommand(routerCommands_1.CommandType.V4_SWAP, [v4Planner.finalize()]);
        }
        else if (routePool instanceof v3_sdk_1.Pool) {
            const v3Params = [
                swapRecipient,
                i == 0 ? amountIn : constants_1.CONTRACT_BALANCE,
                !isLastSectionInRoute(i) ? 0 : amountOut,
                (0, router_sdk_1.encodeMixedRouteToPath)(subRoute),
                payerIsUser && i === 0, // payerIsUser
            ];
            if (useV2_1_1)
                v3Params.push(sectionHopSlippage !== null && sectionHopSlippage !== void 0 ? sectionHopSlippage : []);
            planner.addCommand(routerCommands_1.CommandType.V3_SWAP_EXACT_IN, v3Params, false, options.urVersion);
        }
        else if (routePool instanceof v2_sdk_1.Pair) {
            const v2Params = [
                swapRecipient,
                i === 0 ? amountIn : constants_1.CONTRACT_BALANCE,
                !isLastSectionInRoute(i) ? 0 : amountOut,
                subRoute.path.map((token) => token.wrapped.address),
                payerIsUser && i === 0,
            ];
            if (useV2_1_1)
                v2Params.push(sectionHopSlippage !== null && sectionHopSlippage !== void 0 ? sectionHopSlippage : []);
            planner.addCommand(routerCommands_1.CommandType.V2_SWAP_EXACT_IN, v2Params, false, options.urVersion);
        }
        else {
            throw new Error('Unexpected Pool Type');
        }
        // perform a token transition (wrap/unwrap if necessary)
        if (!isLastSectionInRoute(i)) {
            if (outputToken.isNative && !nextInputToken.isNative) {
                planner.addCommand(routerCommands_1.CommandType.WRAP_ETH, [constants_1.ROUTER_AS_RECIPIENT, constants_1.CONTRACT_BALANCE]);
            }
            else if (!outputToken.isNative && nextInputToken.isNative) {
                planner.addCommand(routerCommands_1.CommandType.UNWRAP_WETH, [constants_1.ROUTER_AS_RECIPIENT, 0]);
            }
        }
        hopOffset += section.length;
        inputToken = nextInputToken;
    }
}
// if price impact is very high, there's a chance of hitting max/min prices resulting in a partial fill of the swap
function riskOfPartialFill(trade) {
    return trade.priceImpact.greaterThan(REFUND_ETH_PRICE_IMPACT_THRESHOLD);
}
function hasFeeOption(swapOptions) {
    return !!swapOptions.fee || !!swapOptions.flatFee;
}
