// @uniswap/universal-router-sdk/dist/cjs/src/entities/actions/unwrapWETH.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnwrapWETH = void 0;
const tslib_1 = require("tslib");
const tiny_invariant_1 = tslib_1.__importDefault(require("tiny-invariant"));
const routerCommands_1 = require("../../utils/routerCommands");
const inputTokens_1 = require("../../utils/inputTokens");
const Command_1 = require("../Command");
const constants_1 = require("../../utils/constants");
class UnwrapWETH {
    constructor(amount, chainId, permit2) {
        this.tradeType = Command_1.RouterActionType.UnwrapWETH;
        this.wethAddress = (0, constants_1.WETH_ADDRESS)(chainId);
        this.amount = amount;
        if (!!permit2) {
            (0, tiny_invariant_1.default)(permit2.details.token.toLowerCase() === this.wethAddress.toLowerCase(), `must be permitting WETH address: ${this.wethAddress}`);
            (0, tiny_invariant_1.default)(permit2.details.amount >= amount, `Did not permit enough WETH for unwrapWETH transaction`);
            this.permit2Data = permit2;
        }
    }
    encode(planner, _) {
        (0, inputTokens_1.encodeInputTokenOptions)(planner, {
            permit2Permit: this.permit2Data,
            permit2TransferFrom: {
                token: this.wethAddress,
                amount: this.amount.toString(),
            },
        });
        planner.addCommand(routerCommands_1.CommandType.UNWRAP_WETH, [constants_1.ROUTER_AS_RECIPIENT, this.amount]);
    }
}
exports.UnwrapWETH = UnwrapWETH;
