// @uniswap/router-sdk/dist/cjs/src/entities/route.js
"use strict";
// entities/route.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.MixedRoute = exports.RouteV4 = exports.RouteV3 = exports.RouteV2 = exports.getPathToken = void 0;
const v2_sdk_1 = require("@uniswap/v2-sdk");
const v3_sdk_1 = require("@uniswap/v3-sdk");
const v4_sdk_1 = require("@uniswap/v4-sdk");
const protocol_1 = require("./protocol");
const route_1 = require("./mixedRoute/route");
// Helper function to get the pathInput and pathOutput for a V2 / V3 route
// currency could be native so we check against the wrapped version as they don't support native ETH in path
function getPathToken(currency, pool) {
    if (pool.token0.wrapped.equals(currency.wrapped)) {
        return pool.token0;
    }
    else if (pool.token1.wrapped.equals(currency.wrapped)) {
        return pool.token1;
    }
    else {
        throw new Error(`Expected token ${currency.symbol} to be either ${pool.token0.symbol} or ${pool.token1.symbol}`);
    }
}
exports.getPathToken = getPathToken;
// V2 route wrapper
class RouteV2 extends v2_sdk_1.Route {
    constructor(v2Route) {
        super(v2Route.pairs, v2Route.input, v2Route.output);
        this.protocol = protocol_1.Protocol.V2;
        this.pools = this.pairs;
        this.pathInput = getPathToken(v2Route.input, this.pairs[0]);
        this.pathOutput = getPathToken(v2Route.output, this.pairs[this.pairs.length - 1]);
    }
}
exports.RouteV2 = RouteV2;
// V3 route wrapper
class RouteV3 extends v3_sdk_1.Route {
    constructor(v3Route) {
        super(v3Route.pools, v3Route.input, v3Route.output);
        this.protocol = protocol_1.Protocol.V3;
        this.path = v3Route.tokenPath;
        this.pathInput = getPathToken(v3Route.input, this.pools[0]);
        this.pathOutput = getPathToken(v3Route.output, this.pools[this.pools.length - 1]);
    }
}
exports.RouteV3 = RouteV3;
// V4 route wrapper
class RouteV4 extends v4_sdk_1.Route {
    constructor(v4Route) {
        super(v4Route.pools, v4Route.input, v4Route.output);
        this.protocol = protocol_1.Protocol.V4;
        this.path = v4Route.currencyPath;
    }
}
exports.RouteV4 = RouteV4;
// Mixed route wrapper
class MixedRoute extends route_1.MixedRouteSDK {
    constructor(mixedRoute) {
        super(mixedRoute.pools, mixedRoute.input, mixedRoute.output);
        this.protocol = protocol_1.Protocol.MIXED;
    }
}
exports.MixedRoute = MixedRoute;
