// @uniswap/universal-router/typechain/factories/BeforeSwapDeltaLibrary__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import { Contract, ContractFactory, Overrides } from "@ethersproject/contracts";

import type { BeforeSwapDeltaLibrary } from "../BeforeSwapDeltaLibrary";

export class BeforeSwapDeltaLibrary__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: Overrides): Promise<BeforeSwapDeltaLibrary> {
    return super.deploy(overrides || {}) as Promise<BeforeSwapDeltaLibrary>;
  }
  getDeployTransaction(overrides?: Overrides): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): BeforeSwapDeltaLibrary {
    return super.attach(address) as BeforeSwapDeltaLibrary;
  }
  connect(signer: Signer): BeforeSwapDeltaLibrary__factory {
    return super.connect(signer) as BeforeSwapDeltaLibrary__factory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): BeforeSwapDeltaLibrary {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as BeforeSwapDeltaLibrary;
  }
}

const _abi = [
  {
    inputs: [],
    name: "ZERO_DELTA",
    outputs: [
      {
        internalType: "BeforeSwapDelta",
        name: "",
        type: "int256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
];

const _bytecode =
  "0x6080806040523460175760469081601c823930815050f35b5f80fdfe60808060405260043610156011575f80fd5b5f3560e01c6391cb4669146023575f80fd5b5f366003190112603557805f60209252f35b5f80fdfea164736f6c634300081a000a";
