// @uniswap/universal-router/typechain/factories/BipsLibrary__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import { Contract, ContractFactory, Overrides } from "@ethersproject/contracts";

import type { BipsLibrary } from "../BipsLibrary";

export class BipsLibrary__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: Overrides): Promise<BipsLibrary> {
    return super.deploy(overrides || {}) as Promise<BipsLibrary>;
  }
  getDeployTransaction(overrides?: Overrides): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): BipsLibrary {
    return super.attach(address) as BipsLibrary;
  }
  connect(signer: Signer): BipsLibrary__factory {
    return super.connect(signer) as BipsLibrary__factory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): BipsLibrary {
    return new Contract(address, _abi, signerOrProvider) as BipsLibrary;
  }
}

const _abi = [
  {
    inputs: [],
    name: "InvalidBips",
    type: "error",
  },
];

const _bytecode =
  "0x6080806040523460175760109081601c823930815050f35b5f80fdfe5f80fdfea164736f6c634300081a000a";
