// @uniswap/universal-router/typechain/factories/CustomRevert__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import { Contract, ContractFactory, Overrides } from "@ethersproject/contracts";

import type { CustomRevert } from "../CustomRevert";

export class CustomRevert__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: Overrides): Promise<CustomRevert> {
    return super.deploy(overrides || {}) as Promise<CustomRevert>;
  }
  getDeployTransaction(overrides?: Overrides): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): CustomRevert {
    return super.attach(address) as CustomRevert;
  }
  connect(signer: Signer): CustomRevert__factory {
    return super.connect(signer) as CustomRevert__factory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): CustomRevert {
    return new Contract(address, _abi, signerOrProvider) as CustomRevert;
  }
}

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "target",
        type: "address",
      },
      {
        internalType: "bytes4",
        name: "selector",
        type: "bytes4",
      },
      {
        internalType: "bytes",
        name: "reason",
        type: "bytes",
      },
      {
        internalType: "bytes",
        name: "details",
        type: "bytes",
      },
    ],
    name: "WrappedError",
    type: "error",
  },
];

const _bytecode =
  "0x6080806040523460175760109081601c823930815050f35b5f80fdfe5f80fdfea164736f6c634300081a000a";
