// @uniswap/universal-router/typechain/factories/DeltaResolver__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer } from "ethers";
import { Provider } from "@ethersproject/providers";

import type { DeltaResolver } from "../DeltaResolver";

export class DeltaResolver__factory {
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): DeltaResolver {
    return new Contract(address, _abi, signerOrProvider) as DeltaResolver;
  }
}

const _abi = [
  {
    inputs: [
      {
        internalType: "Currency",
        name: "currency",
        type: "address",
      },
    ],
    name: "DeltaNotNegative",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "Currency",
        name: "currency",
        type: "address",
      },
    ],
    name: "DeltaNotPositive",
    type: "error",
  },
  {
    inputs: [],
    name: "InsufficientBalance",
    type: "error",
  },
  {
    inputs: [],
    name: "NotPoolManager",
    type: "error",
  },
  {
    inputs: [],
    name: "poolManager",
    outputs: [
      {
        internalType: "contract IPoolManager",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
];
