// @uniswap/universal-router/typechain/factories/ExampleModule__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import { Contract, ContractFactory, Overrides } from "@ethersproject/contracts";

import type { ExampleModule } from "../ExampleModule";

export class ExampleModule__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: Overrides): Promise<ExampleModule> {
    return super.deploy(overrides || {}) as Promise<ExampleModule>;
  }
  getDeployTransaction(overrides?: Overrides): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): ExampleModule {
    return super.attach(address) as ExampleModule;
  }
  connect(signer: Signer): ExampleModule__factory {
    return super.connect(signer) as ExampleModule__factory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ExampleModule {
    return new Contract(address, _abi, signerOrProvider) as ExampleModule;
  }
}

const _abi = [
  {
    inputs: [],
    name: "CauseRevert",
    type: "error",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: false,
        internalType: "string",
        name: "message",
        type: "string",
      },
    ],
    name: "ExampleModuleEvent",
    type: "event",
  },
  {
    inputs: [],
    name: "causeRevert",
    outputs: [],
    stateMutability: "pure",
    type: "function",
  },
  {
    inputs: [],
    name: "logEvent",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
];

const _bytecode =
  "0x6080806040523460135760ae908160188239f35b5f80fdfe608060405260043610156010575f80fd5b5f3560e01c806367192b631460845763a5fe087214602c575f80fd5b346080575f3660031901126080577ff173516ab337f2e7729ba9acd5771deab6be31f2ad8d5dd42dab5269e61701b960606040516020815260096020820152681d195cdd115d995b9d60ba1b6040820152a1005b5f80fd5b346080575f366003190112608057630e524bcd60e21b5f5260045ffdfea164736f6c634300081a000a";
