// @uniswap/universal-router/typechain/factories/IExtsload__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer } from "ethers";
import { Provider } from "@ethersproject/providers";

import type { IExtsload } from "../IExtsload";

export class IExtsload__factory {
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IExtsload {
    return new Contract(address, _abi, signerOrProvider) as IExtsload;
  }
}

const _abi = [
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "slot",
        type: "bytes32",
      },
    ],
    name: "extsload",
    outputs: [
      {
        internalType: "bytes32",
        name: "value",
        type: "bytes32",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "startSlot",
        type: "bytes32",
      },
      {
        internalType: "uint256",
        name: "nSlots",
        type: "uint256",
      },
    ],
    name: "extsload",
    outputs: [
      {
        internalType: "bytes32[]",
        name: "values",
        type: "bytes32[]",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bytes32[]",
        name: "slots",
        type: "bytes32[]",
      },
    ],
    name: "extsload",
    outputs: [
      {
        internalType: "bytes32[]",
        name: "values",
        type: "bytes32[]",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
];
