// @uniswap/universal-router/typechain/factories/IV4Router__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer } from "ethers";
import { Provider } from "@ethersproject/providers";

import type { IV4Router } from "../IV4Router";

export class IV4Router__factory {
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IV4Router {
    return new Contract(address, _abi, signerOrProvider) as IV4Router;
  }
}

const _abi = [
  {
    inputs: [],
    name: "InvalidHopSlippageLength",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "minAmountOutReceived",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "amountReceived",
        type: "uint256",
      },
    ],
    name: "V4TooLittleReceived",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "hopIndex",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "maxPrice",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "price",
        type: "uint256",
      },
    ],
    name: "V4TooLittleReceivedPerHop",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "maxAmountInRequested",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "amountRequested",
        type: "uint256",
      },
    ],
    name: "V4TooMuchRequested",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "hopIndex",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "maxPrice",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "price",
        type: "uint256",
      },
    ],
    name: "V4TooMuchRequestedPerHop",
    type: "error",
  },
  {
    inputs: [],
    name: "poolManager",
    outputs: [
      {
        internalType: "contract IPoolManager",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
];
