// @uniswap/universal-router/typechain/factories/Lock__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import { Contract, ContractFactory, Overrides } from "@ethersproject/contracts";

import type { Lock } from "../Lock";

export class Lock__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: Overrides): Promise<Lock> {
    return super.deploy(overrides || {}) as Promise<Lock>;
  }
  getDeployTransaction(overrides?: Overrides): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): Lock {
    return super.attach(address) as Lock;
  }
  connect(signer: Signer): Lock__factory {
    return super.connect(signer) as Lock__factory;
  }
  static connect(address: string, signerOrProvider: Signer | Provider): Lock {
    return new Contract(address, _abi, signerOrProvider) as Lock;
  }
}

const _abi = [
  {
    inputs: [],
    name: "ContractLocked",
    type: "error",
  },
];

const _bytecode =
  "0x608080604052346013576010908160188239f35b5f80fdfe5f80fdfea164736f6c634300081a000a";
