// @uniswap/universal-router/typechain/factories/NativeWrapper__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer } from "ethers";
import { Provider } from "@ethersproject/providers";

import type { NativeWrapper } from "../NativeWrapper";

export class NativeWrapper__factory {
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): NativeWrapper {
    return new Contract(address, _abi, signerOrProvider) as NativeWrapper;
  }
}

const _abi = [
  {
    inputs: [],
    name: "InvalidEthSender",
    type: "error",
  },
  {
    inputs: [],
    name: "NotPoolManager",
    type: "error",
  },
  {
    inputs: [],
    name: "WETH9",
    outputs: [
      {
        internalType: "contract IWETH9",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "poolManager",
    outputs: [
      {
        internalType: "contract IPoolManager",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    stateMutability: "payable",
    type: "receive",
  },
];
