// @uniswap/universal-router/typechain/factories/ProtocolFeeLibrary__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import { Contract, ContractFactory, Overrides } from "@ethersproject/contracts";

import type { ProtocolFeeLibrary } from "../ProtocolFeeLibrary";

export class ProtocolFeeLibrary__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: Overrides): Promise<ProtocolFeeLibrary> {
    return super.deploy(overrides || {}) as Promise<ProtocolFeeLibrary>;
  }
  getDeployTransaction(overrides?: Overrides): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): ProtocolFeeLibrary {
    return super.attach(address) as ProtocolFeeLibrary;
  }
  connect(signer: Signer): ProtocolFeeLibrary__factory {
    return super.connect(signer) as ProtocolFeeLibrary__factory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ProtocolFeeLibrary {
    return new Contract(address, _abi, signerOrProvider) as ProtocolFeeLibrary;
  }
}

const _abi = [
  {
    inputs: [],
    name: "MAX_PROTOCOL_FEE",
    outputs: [
      {
        internalType: "uint16",
        name: "",
        type: "uint16",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
];

const _bytecode =
  "0x6080806040523460175760489081601c823930815050f35b5f80fdfe60808060405260043610156011575f80fd5b5f3560e01c63b8ca3b83146023575f80fd5b5f366003190112603757806103e860209252f35b5f80fdfea164736f6c634300081a000a";
