// @uniswap/universal-router/typechain/factories/SafeCast__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import { Contract, ContractFactory, Overrides } from "@ethersproject/contracts";

import type { SafeCast } from "../SafeCast";

export class SafeCast__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: Overrides): Promise<SafeCast> {
    return super.deploy(overrides || {}) as Promise<SafeCast>;
  }
  getDeployTransaction(overrides?: Overrides): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): SafeCast {
    return super.attach(address) as SafeCast;
  }
  connect(signer: Signer): SafeCast__factory {
    return super.connect(signer) as SafeCast__factory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): SafeCast {
    return new Contract(address, _abi, signerOrProvider) as SafeCast;
  }
}

const _abi = [
  {
    inputs: [],
    name: "SafeCastOverflow",
    type: "error",
  },
];

const _bytecode =
  "0x6080806040523460175760109081601c823930815050f35b5f80fdfe5f80fdfea164736f6c634300081a000a";
