// @uniswap/universal-router/typechain/factories/SqrtPriceMath__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import { Contract, ContractFactory, Overrides } from "@ethersproject/contracts";

import type { SqrtPriceMath } from "../SqrtPriceMath";

export class SqrtPriceMath__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: Overrides): Promise<SqrtPriceMath> {
    return super.deploy(overrides || {}) as Promise<SqrtPriceMath>;
  }
  getDeployTransaction(overrides?: Overrides): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): SqrtPriceMath {
    return super.attach(address) as SqrtPriceMath;
  }
  connect(signer: Signer): SqrtPriceMath__factory {
    return super.connect(signer) as SqrtPriceMath__factory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): SqrtPriceMath {
    return new Contract(address, _abi, signerOrProvider) as SqrtPriceMath;
  }
}

const _abi = [
  {
    inputs: [],
    name: "InvalidPrice",
    type: "error",
  },
  {
    inputs: [],
    name: "InvalidPriceOrLiquidity",
    type: "error",
  },
  {
    inputs: [],
    name: "NotEnoughLiquidity",
    type: "error",
  },
  {
    inputs: [],
    name: "PriceOverflow",
    type: "error",
  },
];

const _bytecode =
  "0x6080806040523460175760109081601c823930815050f35b5f80fdfe5f80fdfea164736f6c634300081a000a";
