// @uniswap/universal-router/typechain/factories/TickBitmap__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import { Contract, ContractFactory, Overrides } from "@ethersproject/contracts";

import type { TickBitmap } from "../TickBitmap";

export class TickBitmap__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: Overrides): Promise<TickBitmap> {
    return super.deploy(overrides || {}) as Promise<TickBitmap>;
  }
  getDeployTransaction(overrides?: Overrides): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): TickBitmap {
    return super.attach(address) as TickBitmap;
  }
  connect(signer: Signer): TickBitmap__factory {
    return super.connect(signer) as TickBitmap__factory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): TickBitmap {
    return new Contract(address, _abi, signerOrProvider) as TickBitmap;
  }
}

const _abi = [
  {
    inputs: [
      {
        internalType: "int24",
        name: "tick",
        type: "int24",
      },
      {
        internalType: "int24",
        name: "tickSpacing",
        type: "int24",
      },
    ],
    name: "TickMisaligned",
    type: "error",
  },
];

const _bytecode =
  "0x6080806040523460175760109081601c823930815050f35b5f80fdfe5f80fdfea164736f6c634300081a000a";
