// @uniswap/universal-router/typechain/factories/V3ToV4Migrator__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer } from "ethers";
import { Provider } from "@ethersproject/providers";

import type { V3ToV4Migrator } from "../V3ToV4Migrator";

export class V3ToV4Migrator__factory {
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): V3ToV4Migrator {
    return new Contract(address, _abi, signerOrProvider) as V3ToV4Migrator;
  }
}

const _abi = [
  {
    inputs: [
      {
        internalType: "bytes4",
        name: "action",
        type: "bytes4",
      },
    ],
    name: "InvalidAction",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "tokenId",
        type: "uint256",
      },
    ],
    name: "NotAuthorizedForToken",
    type: "error",
  },
  {
    inputs: [],
    name: "OnlyMintAllowed",
    type: "error",
  },
  {
    inputs: [],
    name: "V3_POSITION_MANAGER",
    outputs: [
      {
        internalType: "contract INonfungiblePositionManager",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "V4_POSITION_MANAGER",
    outputs: [
      {
        internalType: "contract IPositionManager",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
];
