// @uniswap/universal-router/typechain/factories/V4SwapRouter__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer } from "ethers";
import { Provider } from "@ethersproject/providers";

import type { V4SwapRouter } from "../V4SwapRouter";

export class V4SwapRouter__factory {
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): V4SwapRouter {
    return new Contract(address, _abi, signerOrProvider) as V4SwapRouter;
  }
}

const _abi = [
  {
    inputs: [
      {
        internalType: "Currency",
        name: "currency",
        type: "address",
      },
    ],
    name: "DeltaNotNegative",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "Currency",
        name: "currency",
        type: "address",
      },
    ],
    name: "DeltaNotPositive",
    type: "error",
  },
  {
    inputs: [],
    name: "FromAddressIsNotOwner",
    type: "error",
  },
  {
    inputs: [],
    name: "InputLengthMismatch",
    type: "error",
  },
  {
    inputs: [],
    name: "InsufficientBalance",
    type: "error",
  },
  {
    inputs: [],
    name: "InsufficientETH",
    type: "error",
  },
  {
    inputs: [],
    name: "InsufficientToken",
    type: "error",
  },
  {
    inputs: [],
    name: "InvalidBips",
    type: "error",
  },
  {
    inputs: [],
    name: "InvalidHopSlippageLength",
    type: "error",
  },
  {
    inputs: [],
    name: "NotPoolManager",
    type: "error",
  },
  {
    inputs: [],
    name: "UnsafeCast",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "action",
        type: "uint256",
      },
    ],
    name: "UnsupportedAction",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "minAmountOutReceived",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "amountReceived",
        type: "uint256",
      },
    ],
    name: "V4TooLittleReceived",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "hopIndex",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "maxPrice",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "price",
        type: "uint256",
      },
    ],
    name: "V4TooLittleReceivedPerHop",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "maxAmountInRequested",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "amountRequested",
        type: "uint256",
      },
    ],
    name: "V4TooMuchRequested",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "hopIndex",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "maxPrice",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "price",
        type: "uint256",
      },
    ],
    name: "V4TooMuchRequestedPerHop",
    type: "error",
  },
  {
    inputs: [],
    name: "msgSender",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "poolManager",
    outputs: [
      {
        internalType: "contract IPoolManager",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bytes",
        name: "data",
        type: "bytes",
      },
    ],
    name: "unlockCallback",
    outputs: [
      {
        internalType: "bytes",
        name: "",
        type: "bytes",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
];
