// @uniswap/universal-router-sdk/dist/esm/src/entities/actions/unwrapWETH.js
import invariant from 'tiny-invariant';
import { CommandType } from '../../utils/routerCommands';
import { encodeInputTokenOptions } from '../../utils/inputTokens';
import { RouterActionType } from '../Command';
import { ROUTER_AS_RECIPIENT, WETH_ADDRESS } from '../../utils/constants';
export class UnwrapWETH {
    constructor(amount, chainId, permit2) {
        this.tradeType = RouterActionType.UnwrapWETH;
        this.wethAddress = WETH_ADDRESS(chainId);
        this.amount = amount;
        if (!!permit2) {
            invariant(permit2.details.token.toLowerCase() === this.wethAddress.toLowerCase(), `must be permitting WETH address: ${this.wethAddress}`);
            invariant(permit2.details.amount >= amount, `Did not permit enough WETH for unwrapWETH transaction`);
            this.permit2Data = permit2;
        }
    }
    encode(planner, _) {
        encodeInputTokenOptions(planner, {
            permit2Permit: this.permit2Data,
            permit2TransferFrom: {
                token: this.wethAddress,
                amount: this.amount.toString(),
            },
        });
        planner.addCommand(CommandType.UNWRAP_WETH, [ROUTER_AS_RECIPIENT, this.amount]);
    }
}
