// @0glabs/0g-serving-broker/lib.commonjs/inference/contract/read-only-inference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReadOnlyInferenceServingContract = void 0;
const typechain_1 = require("./typechain");
const utils_1 = require("../../common/utils");
class ReadOnlyInferenceServingContract {
    serving;
    constructor(provider, contractAddress) {
        this.serving = typechain_1.InferenceServing__factory.connect(contractAddress, provider);
    }
    lockTime() {
        return this.serving.lockTime();
    }
    async listService(offset = 0, limit = 50, includeUnacknowledged = false) {
        try {
            const result = await this.serving.getAllServices(offset, limit);
            if (includeUnacknowledged) {
                return result.services;
            }
            return result.services.filter((service) => service.teeSignerAcknowledged);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    async getService(providerAddress) {
        try {
            return this.serving.getService(providerAddress);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    async getChainId() {
        const network = await this.serving.runner?.provider?.getNetwork();
        return network?.chainId;
    }
}
exports.ReadOnlyInferenceServingContract = ReadOnlyInferenceServingContract;
