// @0glabs/0g-serving-broker/lib.commonjs/fine-tuning/contract/read-only-fine-tuning.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReadOnlyFineTuningServingContract = void 0;
const typechain_1 = require("./typechain");
const utils_1 = require("../../common/utils");
class ReadOnlyFineTuningServingContract {
    serving;
    constructor(provider, contractAddress) {
        this.serving = typechain_1.FineTuningServing__factory.connect(contractAddress, provider);
    }
    lockTime() {
        return this.serving.lockTime();
    }
    async listService(includeUnacknowledged = false) {
        try {
            const services = await this.serving.getAllServices();
            if (includeUnacknowledged) {
                return services;
            }
            return services.filter((service) => service.teeSignerAcknowledged);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    async getService(providerAddress) {
        try {
            return this.serving.getService(providerAddress);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    async getChainId() {
        const network = await this.serving.runner?.provider?.getNetwork();
        return network?.chainId;
    }
}
exports.ReadOnlyFineTuningServingContract = ReadOnlyFineTuningServingContract;
