// @uniswap/v2-sdk/dist/cjs/src/entities/pair.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pair = exports.computePairAddress = void 0;
const tslib_1 = require("tslib");
const address_1 = require("@ethersproject/address");
const bignumber_1 = require("@ethersproject/bignumber");
const solidity_1 = require("@ethersproject/solidity");
const sdk_core_1 = require("@uniswap/sdk-core");
const jsbi_1 = tslib_1.__importDefault(require("jsbi"));
const tiny_invariant_1 = tslib_1.__importDefault(require("tiny-invariant"));
const constants_1 = require("../constants");
const errors_1 = require("../errors");
const computePairAddress = ({ factoryAddress, tokenA, tokenB, }) => {
    const [token0, token1] = tokenA.sortsBefore(tokenB) ? [tokenA, tokenB] : [tokenB, tokenA]; // does safety checks
    return (0, address_1.getCreate2Address)(factoryAddress, (0, solidity_1.keccak256)(['bytes'], [(0, solidity_1.pack)(['address', 'address'], [token0.address, token1.address])]), constants_1.INIT_CODE_HASH);
};
exports.computePairAddress = computePairAddress;
class Pair {
    static getAddress(tokenA, tokenB) {
        var _a;
        const factoryAddress = (_a = constants_1.FACTORY_ADDRESS_MAP[tokenA.chainId]) !== null && _a !== void 0 ? _a : constants_1.FACTORY_ADDRESS;
        return (0, exports.computePairAddress)({ factoryAddress, tokenA, tokenB });
    }
    constructor(currencyAmountA, tokenAmountB) {
        const tokenAmounts = currencyAmountA.currency.sortsBefore(tokenAmountB.currency) // does safety checks
            ? [currencyAmountA, tokenAmountB]
            : [tokenAmountB, currencyAmountA];
        this.liquidityToken = new sdk_core_1.Token(tokenAmounts[0].currency.chainId, Pair.getAddress(tokenAmounts[0].currency, tokenAmounts[1].currency), 18, 'UNI-V2', 'Uniswap V2');
        this.tokenAmounts = tokenAmounts;
    }
    /**
     * Returns true if the token is either token0 or token1
     * @param token to check
     */
    involvesToken(token) {
        return token.equals(this.token0) || token.equals(this.token1);
    }
    /**
     * Returns the current mid price of the pair in terms of token0, i.e. the ratio of reserve1 to reserve0
     */
    get token0Price() {
        const result = this.tokenAmounts[1].divide(this.tokenAmounts[0]);
        return new sdk_core_1.Price(this.token0, this.token1, result.denominator, result.numerator);
    }
    /**
     * Returns the current mid price of the pair in terms of token1, i.e. the ratio of reserve0 to reserve1
     */
    get token1Price() {
        const result = this.tokenAmounts[0].divide(this.tokenAmounts[1]);
        return new sdk_core_1.Price(this.token1, this.token0, result.denominator, result.numerator);
    }
    /**
     * Return the price of the given token in terms of the other token in the pair.
     * @param token token to return price of
     */
    priceOf(token) {
        (0, tiny_invariant_1.default)(this.involvesToken(token), 'TOKEN');
        return token.equals(this.token0) ? this.token0Price : this.token1Price;
    }
    /**
     * Returns the chain ID of the tokens in the pair.
     */
    get chainId() {
        return this.token0.chainId;
    }
    get token0() {
        return this.tokenAmounts[0].currency;
    }
    get token1() {
        return this.tokenAmounts[1].currency;
    }
    get reserve0() {
        return this.tokenAmounts[0];
    }
    get reserve1() {
        return this.tokenAmounts[1];
    }
    reserveOf(token) {
        (0, tiny_invariant_1.default)(this.involvesToken(token), 'TOKEN');
        return token.equals(this.token0) ? this.reserve0 : this.reserve1;
    }
    /**
     * getAmountOut is the linear algebra of reserve ratio against amountIn:amountOut.
     * https://ethereum.stackexchange.com/questions/101629/what-is-math-for-uniswap-calculates-the-amountout-and-amountin-why-997-and-1000
     * has the math deduction for the reserve calculation without fee-on-transfer fees.
     *
     * With fee-on-transfer tax, intuitively it's just:
     * inputAmountWithFeeAndTax = 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn
     *                          = (1 - amountIn.sellFeesBips / 10000) * amountInWithFee
     * where amountInWithFee is the amountIn after taking out the LP fees
     * outputAmountWithTax = amountOut * (1 - amountOut.buyFeesBips / 10000)
     *
     * But we are illustrating the math deduction below to ensure that's the case.
     *
     * before swap A * B = K where A = reserveIn B = reserveOut
     *
     * after swap A' * B' = K where only k is a constant value
     *
     * getAmountOut
     *
     * A' = A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn # here 0.3% is deducted
     * B' = B - amountOut * (1 - amountOut.buyFeesBips / 10000)
     * amountOut = (B - B') / (1 - amountOut.buyFeesBips / 10000) # where A' * B' still is k
     *           = (B - K/(A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn))
     *             /
     *             (1 - amountOut.buyFeesBips / 10000)
     *           = (B - AB/(A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn))
     *             /
     *             (1 - amountOut.buyFeesBips / 10000)
     *           = ((BA + B * 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn - AB)/(A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn))
     *             /
     *             (1 - amountOut.buyFeesBips / 10000)
     *           = (B * 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn / (A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn)
     *             /
     *             (1 - amountOut.buyFeesBips / 10000)
     * amountOut * (1 - amountOut.buyFeesBips / 10000) = (B * 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn
     *                                                    /
     *                                                    (A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn)
     *
     * outputAmountWithTax = (B * 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn
     *                       /
     *                       (A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn)
     *                       = (B * 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn * 1000
     *                       /
     *                       ((A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn) * 1000)
     *                     = (B * (1 - amountIn.sellFeesBips / 10000) 997 * * amountIn
     *                       /
     *                       (1000 * A + (1 - amountIn.sellFeesBips / 10000) * 997 * amountIn)
     *                     = (B * (1 - amountIn.sellFeesBips / 10000) * inputAmountWithFee)
     *                       /
     *                       (1000 * A + (1 - amountIn.sellFeesBips / 10000) * inputAmountWithFee)
     *                     = (B * inputAmountWithFeeAndTax)
     *                       /
     *                       (1000 * A + inputAmountWithFeeAndTax)
     *
     * inputAmountWithFeeAndTax = (1 - amountIn.sellFeesBips / 10000) * inputAmountWithFee
     * outputAmountWithTax = amountOut * (1 - amountOut.buyFeesBips / 10000)
     *
     * @param inputAmount
     * @param calculateFotFees
     */
    getOutputAmount(inputAmount, calculateFotFees = true) {
        (0, tiny_invariant_1.default)(this.involvesToken(inputAmount.currency), 'TOKEN');
        if (jsbi_1.default.equal(this.reserve0.quotient, constants_1.ZERO) || jsbi_1.default.equal(this.reserve1.quotient, constants_1.ZERO)) {
            throw new errors_1.InsufficientReservesError();
        }
        const inputReserve = this.reserveOf(inputAmount.currency);
        const outputReserve = this.reserveOf(inputAmount.currency.equals(this.token0) ? this.token1 : this.token0);
        const percentAfterSellFees = calculateFotFees ? this.derivePercentAfterSellFees(inputAmount) : constants_1.ZERO_PERCENT;
        const inputAmountAfterTax = percentAfterSellFees.greaterThan(constants_1.ZERO_PERCENT)
            ? sdk_core_1.CurrencyAmount.fromRawAmount(inputAmount.currency, percentAfterSellFees.multiply(inputAmount).quotient // fraction.quotient will round down by itself, which is desired
            )
            : inputAmount;
        const inputAmountWithFeeAndAfterTax = jsbi_1.default.multiply(inputAmountAfterTax.quotient, constants_1._997);
        const numerator = jsbi_1.default.multiply(inputAmountWithFeeAndAfterTax, outputReserve.quotient);
        const denominator = jsbi_1.default.add(jsbi_1.default.multiply(inputReserve.quotient, constants_1._1000), inputAmountWithFeeAndAfterTax);
        const outputAmount = sdk_core_1.CurrencyAmount.fromRawAmount(inputAmount.currency.equals(this.token0) ? this.token1 : this.token0, jsbi_1.default.divide(numerator, denominator) // JSBI.divide will round down by itself, which is desired
        );
        if (jsbi_1.default.equal(outputAmount.quotient, constants_1.ZERO)) {
            throw new errors_1.InsufficientInputAmountError();
        }
        const percentAfterBuyFees = calculateFotFees ? this.derivePercentAfterBuyFees(outputAmount) : constants_1.ZERO_PERCENT;
        const outputAmountAfterTax = percentAfterBuyFees.greaterThan(constants_1.ZERO_PERCENT)
            ? sdk_core_1.CurrencyAmount.fromRawAmount(outputAmount.currency, outputAmount.multiply(percentAfterBuyFees).quotient // fraction.quotient will round down by itself, which is desired
            )
            : outputAmount;
        if (jsbi_1.default.equal(outputAmountAfterTax.quotient, constants_1.ZERO)) {
            throw new errors_1.InsufficientInputAmountError();
        }
        return [
            outputAmountAfterTax,
            new Pair(inputReserve.add(inputAmountAfterTax), outputReserve.subtract(outputAmountAfterTax)),
        ];
    }
    /**
     * getAmountIn is the linear algebra of reserve ratio against amountIn:amountOut.
     * https://ethereum.stackexchange.com/questions/101629/what-is-math-for-uniswap-calculates-the-amountout-and-amountin-why-997-and-1000
     * has the math deduction for the reserve calculation without fee-on-transfer fees.
     *
     * With fee-on-transfer fees, intuitively it's just:
     * outputAmountWithTax = amountOut / (1 - amountOut.buyFeesBips / 10000)
     * inputAmountWithTax = amountIn / (1 - amountIn.sellFeesBips / 10000) / 0.997
     *
     * But we are illustrating the math deduction below to ensure that's the case.
     *
     * before swap A * B = K where A = reserveIn B = reserveOut
     *
     * after swap A' * B' = K where only k is a constant value
     *
     * getAmountIn
     *
     * B' = B - amountOut * (1 - amountOut.buyFeesBips / 10000)
     * A' = A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn # here 0.3% is deducted
     * amountIn = (A' - A) / (0.997 * (1 - amountIn.sellFeesBips / 10000))
     *          = (K / (B - amountOut / (1 - amountOut.buyFeesBips / 10000)) - A)
     *            /
     *            (0.997 * (1 - amountIn.sellFeesBips / 10000))
     *          = (AB / (B - amountOut / (1 - amountOut.buyFeesBips / 10000)) - A)
     *            /
     *            (0.997 * (1 - amountIn.sellFeesBips / 10000))
     *          = ((AB - AB + A * amountOut / (1 - amountOut.buyFeesBips / 10000)) / (B - amountOut / (1 - amountOut.buyFeesBips / 10000)))
     *            /
     *            (0.997 * (1 - amountIn.sellFeesBips / 10000))
     *          = ((A * amountOut / (1 - amountOut.buyFeesBips / 10000)) / (B - amountOut / (1 - amountOut.buyFeesBips / 10000)))
     *            /
     *            (0.997 * (1 - amountIn.sellFeesBips / 10000))
     *          = ((A * 1000 * amountOut / (1 - amountOut.buyFeesBips / 10000)) / (B - amountOut / (1 - amountOut.buyFeesBips / 10000)))
     *            /
     *            (997 * (1 - amountIn.sellFeesBips / 10000))
     *
     * outputAmountWithTax = amountOut / (1 - amountOut.buyFeesBips / 10000)
     * inputAmountWithTax = amountIn / (997 * (1 - amountIn.sellFeesBips / 10000))
     *                    = (A * outputAmountWithTax * 1000) / ((B - outputAmountWithTax) * 997)
     *
     * @param outputAmount
     */
    getInputAmount(outputAmount, calculateFotFees = true) {
        (0, tiny_invariant_1.default)(this.involvesToken(outputAmount.currency), 'TOKEN');
        const percentAfterBuyFees = calculateFotFees ? this.derivePercentAfterBuyFees(outputAmount) : constants_1.ZERO_PERCENT;
        const outputAmountBeforeTax = percentAfterBuyFees.greaterThan(constants_1.ZERO_PERCENT)
            ? sdk_core_1.CurrencyAmount.fromRawAmount(outputAmount.currency, jsbi_1.default.add(outputAmount.divide(percentAfterBuyFees).quotient, constants_1.ONE) // add 1 for rounding up
            )
            : outputAmount;
        if (jsbi_1.default.equal(this.reserve0.quotient, constants_1.ZERO) ||
            jsbi_1.default.equal(this.reserve1.quotient, constants_1.ZERO) ||
            jsbi_1.default.greaterThanOrEqual(outputAmount.quotient, this.reserveOf(outputAmount.currency).quotient) ||
            jsbi_1.default.greaterThanOrEqual(outputAmountBeforeTax.quotient, this.reserveOf(outputAmount.currency).quotient)) {
            throw new errors_1.InsufficientReservesError();
        }
        const outputReserve = this.reserveOf(outputAmount.currency);
        const inputReserve = this.reserveOf(outputAmount.currency.equals(this.token0) ? this.token1 : this.token0);
        const numerator = jsbi_1.default.multiply(jsbi_1.default.multiply(inputReserve.quotient, outputAmountBeforeTax.quotient), constants_1._1000);
        const denominator = jsbi_1.default.multiply(jsbi_1.default.subtract(outputReserve.quotient, outputAmountBeforeTax.quotient), constants_1._997);
        const inputAmount = sdk_core_1.CurrencyAmount.fromRawAmount(outputAmount.currency.equals(this.token0) ? this.token1 : this.token0, jsbi_1.default.add(jsbi_1.default.divide(numerator, denominator), constants_1.ONE) // add 1 here is part of the formula, no rounding needed here, since there will not be decimal at this point
        );
        const percentAfterSellFees = calculateFotFees ? this.derivePercentAfterSellFees(inputAmount) : constants_1.ZERO_PERCENT;
        const inputAmountBeforeTax = percentAfterSellFees.greaterThan(constants_1.ZERO_PERCENT)
            ? sdk_core_1.CurrencyAmount.fromRawAmount(inputAmount.currency, jsbi_1.default.add(inputAmount.divide(percentAfterSellFees).quotient, constants_1.ONE) // add 1 for rounding up
            )
            : inputAmount;
        return [inputAmountBeforeTax, new Pair(inputReserve.add(inputAmount), outputReserve.subtract(outputAmount))];
    }
    getLiquidityMinted(totalSupply, tokenAmountA, tokenAmountB) {
        (0, tiny_invariant_1.default)(totalSupply.currency.equals(this.liquidityToken), 'LIQUIDITY');
        const tokenAmounts = tokenAmountA.currency.sortsBefore(tokenAmountB.currency) // does safety checks
            ? [tokenAmountA, tokenAmountB]
            : [tokenAmountB, tokenAmountA];
        (0, tiny_invariant_1.default)(tokenAmounts[0].currency.equals(this.token0) && tokenAmounts[1].currency.equals(this.token1), 'TOKEN');
        let liquidity;
        if (jsbi_1.default.equal(totalSupply.quotient, constants_1.ZERO)) {
            liquidity = jsbi_1.default.subtract((0, sdk_core_1.sqrt)(jsbi_1.default.multiply(tokenAmounts[0].quotient, tokenAmounts[1].quotient)), constants_1.MINIMUM_LIQUIDITY);
        }
        else {
            const amount0 = jsbi_1.default.divide(jsbi_1.default.multiply(tokenAmounts[0].quotient, totalSupply.quotient), this.reserve0.quotient);
            const amount1 = jsbi_1.default.divide(jsbi_1.default.multiply(tokenAmounts[1].quotient, totalSupply.quotient), this.reserve1.quotient);
            liquidity = jsbi_1.default.lessThanOrEqual(amount0, amount1) ? amount0 : amount1;
        }
        if (!jsbi_1.default.greaterThan(liquidity, constants_1.ZERO)) {
            throw new errors_1.InsufficientInputAmountError();
        }
        return sdk_core_1.CurrencyAmount.fromRawAmount(this.liquidityToken, liquidity);
    }
    getLiquidityValue(token, totalSupply, liquidity, feeOn = false, kLast) {
        (0, tiny_invariant_1.default)(this.involvesToken(token), 'TOKEN');
        (0, tiny_invariant_1.default)(totalSupply.currency.equals(this.liquidityToken), 'TOTAL_SUPPLY');
        (0, tiny_invariant_1.default)(liquidity.currency.equals(this.liquidityToken), 'LIQUIDITY');
        (0, tiny_invariant_1.default)(jsbi_1.default.lessThanOrEqual(liquidity.quotient, totalSupply.quotient), 'LIQUIDITY');
        let totalSupplyAdjusted;
        if (!feeOn) {
            totalSupplyAdjusted = totalSupply;
        }
        else {
            (0, tiny_invariant_1.default)(!!kLast, 'K_LAST');
            const kLastParsed = jsbi_1.default.BigInt(kLast);
            if (!jsbi_1.default.equal(kLastParsed, constants_1.ZERO)) {
                const rootK = (0, sdk_core_1.sqrt)(jsbi_1.default.multiply(this.reserve0.quotient, this.reserve1.quotient));
                const rootKLast = (0, sdk_core_1.sqrt)(kLastParsed);
                if (jsbi_1.default.greaterThan(rootK, rootKLast)) {
                    const numerator = jsbi_1.default.multiply(totalSupply.quotient, jsbi_1.default.subtract(rootK, rootKLast));
                    const denominator = jsbi_1.default.add(jsbi_1.default.multiply(rootK, constants_1.FIVE), rootKLast);
                    const feeLiquidity = jsbi_1.default.divide(numerator, denominator);
                    totalSupplyAdjusted = totalSupply.add(sdk_core_1.CurrencyAmount.fromRawAmount(this.liquidityToken, feeLiquidity));
                }
                else {
                    totalSupplyAdjusted = totalSupply;
                }
            }
            else {
                totalSupplyAdjusted = totalSupply;
            }
        }
        return sdk_core_1.CurrencyAmount.fromRawAmount(token, jsbi_1.default.divide(jsbi_1.default.multiply(liquidity.quotient, this.reserveOf(token).quotient), totalSupplyAdjusted.quotient));
    }
    derivePercentAfterSellFees(inputAmount) {
        const sellFeeBips = this.token0.wrapped.equals(inputAmount.wrapped.currency)
            ? this.token0.wrapped.sellFeeBps
            : this.token1.wrapped.sellFeeBps;
        if (sellFeeBips === null || sellFeeBips === void 0 ? void 0 : sellFeeBips.gt(bignumber_1.BigNumber.from(0))) {
            return constants_1.ONE_HUNDRED_PERCENT.subtract(new sdk_core_1.Percent(jsbi_1.default.BigInt(sellFeeBips)).divide(constants_1.BASIS_POINTS));
        }
        else {
            return constants_1.ZERO_PERCENT;
        }
    }
    derivePercentAfterBuyFees(outputAmount) {
        const buyFeeBps = this.token0.wrapped.equals(outputAmount.wrapped.currency)
            ? this.token0.wrapped.buyFeeBps
            : this.token1.wrapped.buyFeeBps;
        if (buyFeeBps === null || buyFeeBps === void 0 ? void 0 : buyFeeBps.gt(bignumber_1.BigNumber.from(0))) {
            return constants_1.ONE_HUNDRED_PERCENT.subtract(new sdk_core_1.Percent(jsbi_1.default.BigInt(buyFeeBps)).divide(constants_1.BASIS_POINTS));
        }
        else {
            return constants_1.ZERO_PERCENT;
        }
    }
}
exports.Pair = Pair;
