// @uniswap/router-sdk/dist/cjs/src/entities/mixedRoute/trade.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MixedRouteTrade = exports.tradeComparator = void 0;
const tslib_1 = require("tslib");
const sdk_core_1 = require("@uniswap/sdk-core");
const v2_sdk_1 = require("@uniswap/v2-sdk");
const v3_sdk_1 = require("@uniswap/v3-sdk");
const v4_sdk_1 = require("@uniswap/v4-sdk");
const tiny_invariant_1 = tslib_1.__importDefault(require("tiny-invariant"));
const constants_1 = require("../../constants");
const route_1 = require("./route");
const pathCurrency_1 = require("../../utils/pathCurrency");
/**
 * Trades comparator, an extension of the input output comparator that also considers other dimensions of the trade in ranking them
 * @template TInput The input token, either Ether or an ERC-20
 * @template TOutput The output token, either Ether or an ERC-20
 * @template TTradeType The trade type, either exact input or exact output
 * @param a The first trade to compare
 * @param b The second trade to compare
 * @returns A sorted ordering for two neighboring elements in a trade array
 */
function tradeComparator(a, b) {
    // must have same input and output token for comparison
    (0, tiny_invariant_1.default)(a.inputAmount.currency.equals(b.inputAmount.currency), 'INPUT_CURRENCY');
    (0, tiny_invariant_1.default)(a.outputAmount.currency.equals(b.outputAmount.currency), 'OUTPUT_CURRENCY');
    if (a.outputAmount.equalTo(b.outputAmount)) {
        if (a.inputAmount.equalTo(b.inputAmount)) {
            // consider the number of hops since each hop costs gas
            const aHops = a.swaps.reduce((total, cur) => total + cur.route.path.length, 0);
            const bHops = b.swaps.reduce((total, cur) => total + cur.route.path.length, 0);
            return aHops - bHops;
        }
        // trade A requires less input than trade B, so A should come first
        if (a.inputAmount.lessThan(b.inputAmount)) {
            return -1;
        }
        else {
            return 1;
        }
    }
    else {
        // tradeA has less output than trade B, so should come second
        if (a.outputAmount.lessThan(b.outputAmount)) {
            return 1;
        }
        else {
            return -1;
        }
    }
}
exports.tradeComparator = tradeComparator;
/**
 * Represents a trade executed against a set of routes where some percentage of the input is
 * split across each route.
 *
 * Each route has its own set of pools. Pools can not be re-used across routes.
 *
 * Does not account for slippage, i.e., changes in price environment that can occur between
 * the time the trade is submitted and when it is executed.
 * @notice This class is functionally the same as the `Trade` class in the `@uniswap/v3-sdk` package, aside from typing and some input validation.
 * @template TInput The input token, either Ether or an ERC-20
 * @template TOutput The output token, either Ether or an ERC-20
 * @template TTradeType The trade type, either exact input or exact output
 */
class MixedRouteTrade {
    /**
     * @deprecated Deprecated in favor of 'swaps' property. If the trade consists of multiple routes
     * this will return an error.
     *
     * When the trade consists of just a single route, this returns the route of the trade,
     * i.e. which pools the trade goes through.
     */
    get route() {
        (0, tiny_invariant_1.default)(this.swaps.length === 1, 'MULTIPLE_ROUTES');
        return this.swaps[0].route;
    }
    /**
     * The input amount for the trade assuming no slippage.
     */
    get inputAmount() {
        if (this._inputAmount) {
            return this._inputAmount;
        }
        const inputCurrency = this.swaps[0].inputAmount.currency;
        const totalInputFromRoutes = this.swaps
            .map(({ inputAmount }) => inputAmount)
            .reduce((total, cur) => total.add(cur), sdk_core_1.CurrencyAmount.fromRawAmount(inputCurrency, 0));
        this._inputAmount = totalInputFromRoutes;
        return this._inputAmount;
    }
    /**
     * The output amount for the trade assuming no slippage.
     */
    get outputAmount() {
        if (this._outputAmount) {
            return this._outputAmount;
        }
        const outputCurrency = this.swaps[0].outputAmount.currency;
        const totalOutputFromRoutes = this.swaps
            .map(({ outputAmount }) => outputAmount)
            .reduce((total, cur) => total.add(cur), sdk_core_1.CurrencyAmount.fromRawAmount(outputCurrency, 0));
        this._outputAmount = totalOutputFromRoutes;
        return this._outputAmount;
    }
    /**
     * The price expressed in terms of output amount/input amount.
     */
    get executionPrice() {
        var _a;
        return ((_a = this._executionPrice) !== null && _a !== void 0 ? _a : (this._executionPrice = new sdk_core_1.Price(this.inputAmount.currency, this.outputAmount.currency, this.inputAmount.quotient, this.outputAmount.quotient)));
    }
    /**
     * Returns the percent difference between the route's mid price and the price impact
     */
    get priceImpact() {
        if (this._priceImpact) {
            return this._priceImpact;
        }
        let spotOutputAmount = sdk_core_1.CurrencyAmount.fromRawAmount(this.outputAmount.currency, 0);
        for (const { route, inputAmount } of this.swaps) {
            const midPrice = route.midPrice;
            spotOutputAmount = spotOutputAmount.add(midPrice.quote(inputAmount));
        }
        const priceImpact = spotOutputAmount.subtract(this.outputAmount).divide(spotOutputAmount);
        this._priceImpact = new sdk_core_1.Percent(priceImpact.numerator, priceImpact.denominator);
        return this._priceImpact;
    }
    /**
     * Constructs a trade by simulating swaps through the given route
     * @template TInput The input token, either Ether or an ERC-20.
     * @template TOutput The output token, either Ether or an ERC-20.
     * @template TTradeType The type of the trade, either exact in or exact out.
     * @param route route to swap through
     * @param amount the amount specified, either input or output, depending on tradeType
     * @param tradeType whether the trade is an exact input or exact output swap
     * @returns The route
     */
    static async fromRoute(route, amount, tradeType) {
        const amounts = new Array(route.path.length);
        let inputAmount;
        let outputAmount;
        (0, tiny_invariant_1.default)(tradeType === sdk_core_1.TradeType.EXACT_INPUT, 'TRADE_TYPE');
        (0, tiny_invariant_1.default)(amount.currency.equals(route.input), 'INPUT');
        amounts[0] = (0, pathCurrency_1.amountWithPathCurrency)(amount, route.pools[0]);
        for (let i = 0; i < route.path.length - 1; i++) {
            const pool = route.pools[i];
            const [outputAmount] = await pool.getOutputAmount((0, pathCurrency_1.amountWithPathCurrency)(amounts[i], pool));
            amounts[i + 1] = outputAmount;
        }
        inputAmount = sdk_core_1.CurrencyAmount.fromFractionalAmount(route.input, amount.numerator, amount.denominator);
        outputAmount = sdk_core_1.CurrencyAmount.fromFractionalAmount(route.output, amounts[amounts.length - 1].numerator, amounts[amounts.length - 1].denominator);
        return new MixedRouteTrade({
            routes: [{ inputAmount, outputAmount, route }],
            tradeType,
        });
    }
    /**
     * Constructs a trade from routes by simulating swaps
     *
     * @template TInput The input token, either Ether or an ERC-20.
     * @template TOutput The output token, either Ether or an ERC-20.
     * @template TTradeType The type of the trade, either exact in or exact out.
     * @param routes the routes to swap through and how much of the amount should be routed through each
     * @param tradeType whether the trade is an exact input or exact output swap
     * @returns The trade
     */
    static async fromRoutes(routes, tradeType) {
        const populatedRoutes = [];
        (0, tiny_invariant_1.default)(tradeType === sdk_core_1.TradeType.EXACT_INPUT, 'TRADE_TYPE');
        for (const { route, amount } of routes) {
            const amounts = new Array(route.path.length);
            let inputAmount;
            let outputAmount;
            (0, tiny_invariant_1.default)(amount.currency.equals(route.input), 'INPUT');
            inputAmount = sdk_core_1.CurrencyAmount.fromFractionalAmount(route.input, amount.numerator, amount.denominator);
            amounts[0] = sdk_core_1.CurrencyAmount.fromFractionalAmount(route.pathInput, amount.numerator, amount.denominator);
            for (let i = 0; i < route.path.length - 1; i++) {
                const pool = route.pools[i];
                const [outputAmount] = await pool.getOutputAmount((0, pathCurrency_1.amountWithPathCurrency)(amounts[i], pool));
                amounts[i + 1] = outputAmount;
            }
            outputAmount = sdk_core_1.CurrencyAmount.fromFractionalAmount(route.output, amounts[amounts.length - 1].numerator, amounts[amounts.length - 1].denominator);
            populatedRoutes.push({ route, inputAmount, outputAmount });
        }
        return new MixedRouteTrade({
            routes: populatedRoutes,
            tradeType,
        });
    }
    /**
     * Creates a trade without computing the result of swapping through the route. Useful when you have simulated the trade
     * elsewhere and do not have any tick data
     * @template TInput The input token, either Ether or an ERC-20
     * @template TOutput The output token, either Ether or an ERC-20
     * @template TTradeType The type of the trade, either exact in or exact out
     * @param constructorArguments The arguments passed to the trade constructor
     * @returns The unchecked trade
     */
    static createUncheckedTrade(constructorArguments) {
        return new MixedRouteTrade({
            ...constructorArguments,
            routes: [
                {
                    inputAmount: constructorArguments.inputAmount,
                    outputAmount: constructorArguments.outputAmount,
                    route: constructorArguments.route,
                },
            ],
        });
    }
    /**
     * Creates a trade without computing the result of swapping through the routes. Useful when you have simulated the trade
     * elsewhere and do not have any tick data
     * @template TInput The input token, either Ether or an ERC-20
     * @template TOutput The output token, either Ether or an ERC-20
     * @template TTradeType The type of the trade, either exact in or exact out
     * @param constructorArguments The arguments passed to the trade constructor
     * @returns The unchecked trade
     */
    static createUncheckedTradeWithMultipleRoutes(constructorArguments) {
        return new MixedRouteTrade(constructorArguments);
    }
    /**
     * Construct a trade by passing in the pre-computed property values
     * @param routes The routes through which the trade occurs
     * @param tradeType The type of trade, exact input or exact output
     */
    constructor({ routes, tradeType, }) {
        const inputCurrency = routes[0].inputAmount.currency;
        const outputCurrency = routes[0].outputAmount.currency;
        (0, tiny_invariant_1.default)(routes.every(({ route }) => inputCurrency.wrapped.equals(route.input.wrapped)), 'INPUT_CURRENCY_MATCH');
        (0, tiny_invariant_1.default)(routes.every(({ route }) => outputCurrency.wrapped.equals(route.output.wrapped)), 'OUTPUT_CURRENCY_MATCH');
        const numPools = routes.map(({ route }) => route.pools.length).reduce((total, cur) => total + cur, 0);
        const poolIdentifierSet = new Set();
        for (const { route } of routes) {
            for (const pool of route.pools) {
                if (pool instanceof v4_sdk_1.Pool) {
                    poolIdentifierSet.add(pool.poolId);
                }
                else if (pool instanceof v3_sdk_1.Pool) {
                    poolIdentifierSet.add(v3_sdk_1.Pool.getAddress(pool.token0, pool.token1, pool.fee));
                }
                else if (pool instanceof v2_sdk_1.Pair) {
                    const pair = pool;
                    poolIdentifierSet.add(v2_sdk_1.Pair.getAddress(pair.token0, pair.token1));
                }
                else {
                    throw new Error('Unexpected pool type in route when constructing trade object');
                }
            }
        }
        (0, tiny_invariant_1.default)(numPools === poolIdentifierSet.size, 'POOLS_DUPLICATED');
        (0, tiny_invariant_1.default)(tradeType === sdk_core_1.TradeType.EXACT_INPUT, 'TRADE_TYPE');
        this.swaps = routes;
        this.tradeType = tradeType;
    }
    /**
     * Get the minimum amount that must be received from this trade for the given slippage tolerance
     * @param slippageTolerance The tolerance of unfavorable slippage from the execution price of this trade
     * @returns The amount out
     */
    minimumAmountOut(slippageTolerance, amountOut = this.outputAmount) {
        (0, tiny_invariant_1.default)(!slippageTolerance.lessThan(constants_1.ZERO), 'SLIPPAGE_TOLERANCE');
        /// does not support exactOutput, as enforced in the constructor
        const slippageAdjustedAmountOut = new sdk_core_1.Fraction(constants_1.ONE)
            .add(slippageTolerance)
            .invert()
            .multiply(amountOut.quotient).quotient;
        return sdk_core_1.CurrencyAmount.fromRawAmount(amountOut.currency, slippageAdjustedAmountOut);
    }
    /**
     * Get the maximum amount in that can be spent via this trade for the given slippage tolerance
     * @param slippageTolerance The tolerance of unfavorable slippage from the execution price of this trade
     * @returns The amount in
     */
    maximumAmountIn(slippageTolerance, amountIn = this.inputAmount) {
        (0, tiny_invariant_1.default)(!slippageTolerance.lessThan(constants_1.ZERO), 'SLIPPAGE_TOLERANCE');
        return amountIn;
        /// does not support exactOutput
    }
    /**
     * Return the execution price after accounting for slippage tolerance
     * @param slippageTolerance the allowed tolerated slippage
     * @returns The execution price
     */
    worstExecutionPrice(slippageTolerance) {
        return new sdk_core_1.Price(this.inputAmount.currency, this.outputAmount.currency, this.maximumAmountIn(slippageTolerance).quotient, this.minimumAmountOut(slippageTolerance).quotient);
    }
    /**
     * Given a list of pools, and a fixed amount in, returns the top `maxNumResults` trades that go from an input token
     * amount to an output token, making at most `maxHops` hops.
     * Note this does not consider aggregation, as routes are linear. It's possible a better route exists by splitting
     * the amount in among multiple routes.
     * @param pools the pools to consider in finding the best trade
     * @param nextAmountIn exact amount of input currency to spend
     * @param currencyOut the desired currency out
     * @param maxNumResults maximum number of results to return
     * @param maxHops maximum number of hops a returned trade can make, e.g. 1 hop goes through a single pool
     * @param currentPools used in recursion; the current list of pools
     * @param currencyAmountIn used in recursion; the original value of the currencyAmountIn parameter
     * @param bestTrades used in recursion; the current list of best trades
     * @returns The exact in trade
     */
    static async bestTradeExactIn(pools, currencyAmountIn, currencyOut, { maxNumResults = 3, maxHops = 3 } = {}, 
    // used in recursion.
    currentPools = [], nextAmountIn = currencyAmountIn, bestTrades = []) {
        (0, tiny_invariant_1.default)(pools.length > 0, 'POOLS');
        (0, tiny_invariant_1.default)(maxHops > 0, 'MAX_HOPS');
        (0, tiny_invariant_1.default)(currencyAmountIn === nextAmountIn || currentPools.length > 0, 'INVALID_RECURSION');
        const amountIn = nextAmountIn;
        for (let i = 0; i < pools.length; i++) {
            const pool = pools[i];
            const amountInAdjusted = pool instanceof v4_sdk_1.Pool ? amountIn : amountIn.wrapped;
            // pool irrelevant
            if (!pool.token0.equals(amountInAdjusted.currency) && !pool.token1.equals(amountInAdjusted.currency))
                continue;
            if (pool instanceof v2_sdk_1.Pair) {
                if (pool.reserve0.equalTo(constants_1.ZERO) || pool.reserve1.equalTo(constants_1.ZERO))
                    continue;
            }
            let amountOut;
            try {
                ;
                [amountOut] =
                    pool instanceof v4_sdk_1.Pool
                        ? await pool.getOutputAmount(amountInAdjusted)
                        : await pool.getOutputAmount(amountInAdjusted.wrapped);
            }
            catch (error) {
                // input too low
                // @ts-ignore[2571] error is unknown
                if (error.isInsufficientInputAmountError) {
                    continue;
                }
                throw error;
            }
            // we have arrived at the output token, so this is the final trade of one of the paths
            if (amountOut.currency.wrapped.equals(currencyOut.wrapped)) {
                (0, sdk_core_1.sortedInsert)(bestTrades, await MixedRouteTrade.fromRoute(new route_1.MixedRouteSDK([...currentPools, pool], currencyAmountIn.currency, currencyOut), currencyAmountIn, sdk_core_1.TradeType.EXACT_INPUT), maxNumResults, tradeComparator);
            }
            else if (maxHops > 1 && pools.length > 1) {
                const poolsExcludingThisPool = pools.slice(0, i).concat(pools.slice(i + 1, pools.length));
                // otherwise, consider all the other paths that lead from this token as long as we have not exceeded maxHops
                await MixedRouteTrade.bestTradeExactIn(poolsExcludingThisPool, currencyAmountIn, currencyOut, {
                    maxNumResults,
                    maxHops: maxHops - 1,
                }, [...currentPools, pool], amountOut, bestTrades);
            }
        }
        return bestTrades;
    }
}
exports.MixedRouteTrade = MixedRouteTrade;
