// @0glabs/0g-serving-broker/cli.commonjs/cli/config.js
#!/usr/bin/env ts-node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadConfig = loadConfig;
exports.saveConfig = saveConfig;
exports.getConfiguredRpcEndpoint = getConfiguredRpcEndpoint;
exports.setConfiguredRpcEndpoint = setConfiguredRpcEndpoint;
exports.getConfiguredPrivateKey = getConfiguredPrivateKey;
exports.setConfiguredPrivateKey = setConfiguredPrivateKey;
exports.clearConfig = clearConfig;
exports.getConfiguredControllerEndpoint = getConfiguredControllerEndpoint;
exports.setConfiguredControllerEndpoint = setConfiguredControllerEndpoint;
exports.clearControllerEndpoint = clearControllerEndpoint;
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs"));
const path = tslib_1.__importStar(require("path"));
const os = tslib_1.__importStar(require("os"));
/**
 * Gets the path to the CLI configuration file
 */
function getConfigPath() {
    const homeDir = os.homedir();
    const configDir = path.join(homeDir, '.0g-compute-cli');
    return path.join(configDir, 'config.json');
}
/**
 * Loads the CLI configuration from file
 */
function loadConfig() {
    try {
        const configPath = getConfigPath();
        const configData = fs.readFileSync(configPath, 'utf8');
        return JSON.parse(configData);
    }
    catch {
        // Config file doesn't exist or is invalid, return empty config
        return {};
    }
}
/**
 * Saves the CLI configuration to file
 */
function saveConfig(config) {
    try {
        const configPath = getConfigPath();
        const configDir = path.dirname(configPath);
        // Ensure config directory exists
        fs.mkdirSync(configDir, { recursive: true });
        // Add timestamp
        config.lastUpdated = new Date().toISOString();
        // Write config file
        fs.writeFileSync(configPath, JSON.stringify(config, null, 2));
    }
    catch (error) {
        console.warn('Warning: Failed to save configuration:', error);
    }
}
/**
 * Gets the RPC endpoint from config file, environment variable, or user input
 */
function getConfiguredRpcEndpoint() {
    // Priority: Environment variable > Config file
    const envRpc = process.env['ZG_RPC_ENDPOINT'] || process.env.RPC_ENDPOINT;
    if (envRpc) {
        return envRpc;
    }
    // Check config file
    const config = loadConfig();
    return config.rpcEndpoint;
}
/**
 * Sets the RPC endpoint in config file
 */
function setConfiguredRpcEndpoint(rpcEndpoint, network) {
    const config = loadConfig();
    config.rpcEndpoint = rpcEndpoint;
    config.network = network;
    saveConfig(config);
}
/**
 * Gets the private key from config file only
 */
function getConfiguredPrivateKey() {
    // Only use config file, ignore environment variables
    const config = loadConfig();
    return config.privateKey;
}
/**
 * Sets the private key in config file
 */
function setConfiguredPrivateKey(privateKey) {
    const config = loadConfig();
    config.privateKey = privateKey;
    saveConfig(config);
}
/**
 * Clears the configuration
 */
function clearConfig() {
    try {
        const configPath = getConfigPath();
        fs.unlinkSync(configPath);
    }
    catch {
        // Config file doesn't exist, which is fine
    }
}
/**
 * Gets the controller endpoint from config file
 */
function getConfiguredControllerEndpoint() {
    const config = loadConfig();
    return {
        endpoint: config.controllerEndpoint,
        source: config.controllerEndpointSource,
    };
}
/**
 * Sets the controller endpoint in config file
 */
function setConfiguredControllerEndpoint(endpoint, source) {
    const config = loadConfig();
    config.controllerEndpoint = endpoint;
    config.controllerEndpointSource = source;
    saveConfig(config);
}
/**
 * Clears the controller endpoint from config
 */
function clearControllerEndpoint() {
    const config = loadConfig();
    delete config.controllerEndpoint;
    delete config.controllerEndpointSource;
    saveConfig(config);
}
