// @0glabs/0g-serving-broker/cli.commonjs/sdk/fine-tuning/broker/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrokerBase = void 0;
const broker_1 = require("../../broker");
const const_1 = require("../const");
class BrokerBase {
    contract;
    ledger;
    servingProvider;
    constructor(contract, ledger, servingProvider) {
        this.contract = contract;
        this.ledger = ledger;
        this.servingProvider = servingProvider;
    }
    /**
     * Get storage configuration based on current network
     */
    async getStorageConfig() {
        try {
            const chainId = await this.contract.signer.provider?.getNetwork().then(n => n.chainId);
            const networkType = chainId ? (0, broker_1.getNetworkType)(chainId) : 'unknown';
            if (networkType === 'mainnet') {
                return {
                    rpcUrl: const_1.ZG_RPC_ENDPOINT_MAINNET,
                    indexerUrl: const_1.INDEXER_URL_MAINNET_TURBO,
                };
            }
            else {
                // Default to testnet for testnet, hardhat, and unknown networks
                return {
                    rpcUrl: const_1.ZG_RPC_ENDPOINT_TESTNET,
                    indexerUrl: const_1.INDEXER_URL_TESTNET_TURBO,
                };
            }
        }
        catch {
            // Fallback to testnet if network detection fails
            return {
                rpcUrl: const_1.ZG_RPC_ENDPOINT_TESTNET,
                indexerUrl: const_1.INDEXER_URL_TESTNET_TURBO,
            };
        }
    }
}
exports.BrokerBase = BrokerBase;
