// @0glabs/0g-serving-broker/cli.commonjs/sdk/fine-tuning/broker/broker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FineTuningBroker = void 0;
exports.createFineTuningBroker = createFineTuningBroker;
const contract_1 = require("../contract");
const model_1 = require("./model");
const dataset_1 = require("./dataset");
const service_1 = require("./service");
const provider_1 = require("../provider/provider");
const utils_1 = require("../../common/utils");
const verifier_1 = require("./verifier");
const read_only_broker_1 = require("./read-only-broker");
class FineTuningBroker extends read_only_broker_1.ReadOnlyFineTuningBroker {
    signer;
    fineTuningCA;
    ledger;
    datasetProcessor;
    verifier;
    serviceProvider;
    _gasPrice;
    _maxGasPrice;
    _step;
    constructor(signer, fineTuningCA, ledger, gasPrice, maxGasPrice, step) {
        super(signer, fineTuningCA);
        this.signer = signer;
        this.fineTuningCA = fineTuningCA;
        this.ledger = ledger;
        this._gasPrice = gasPrice;
        this._maxGasPrice = maxGasPrice;
        this._step = step;
    }
    async initialize() {
        let userAddress;
        try {
            userAddress = await this.signer.getAddress();
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
        const contract = new contract_1.FineTuningServingContract(this.signer, this.fineTuningCA, userAddress, this._gasPrice, this._maxGasPrice, this._step);
        this.serviceProvider = new provider_1.Provider(contract);
        this.modelProcessor = new model_1.ModelProcessor(contract, this.ledger, this.serviceProvider);
        this.datasetProcessor = new dataset_1.DatasetProcessor(contract, this.ledger, this.serviceProvider);
        this.serviceProcessor = new service_1.ServiceProcessor(contract, this.ledger, this.serviceProvider);
        this.verifier = new verifier_1.Verifier(contract, this.ledger, this.serviceProvider);
    }
    getLockedTime = async () => {
        try {
            return await this.serviceProcessor.getLockTime();
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    getAccount = async (providerAddress) => {
        try {
            return await this.serviceProcessor.getAccount(providerAddress);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    getAccountWithDetail = async (providerAddress) => {
        try {
            return await this.serviceProcessor.getAccountWithDetail(providerAddress);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    acknowledgeProviderSigner = async (providerAddress, gasPrice) => {
        try {
            return await this.serviceProcessor.acknowledgeProviderSigner(providerAddress, gasPrice);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    acknowledgeTEESignerByOwner = async (providerAddress, gasPrice) => {
        try {
            return await this.serviceProcessor.acknowledgeTEESignerByOwner(providerAddress, gasPrice);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    revokeTEESignerAcknowledgement = async (providerAddress, gasPrice) => {
        try {
            return await this.serviceProcessor.revokeTEESignerAcknowledgement(providerAddress, gasPrice);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    removeService = async (gasPrice) => {
        try {
            return await this.serviceProcessor.removeService(gasPrice);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    modelUsage = (providerAddress, preTrainedModelName, output) => {
        try {
            return this.serviceProcessor.modelUsage(providerAddress, preTrainedModelName, output);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    uploadDataset = async (dataPath, gasPrice, maxGasPrice) => {
        try {
            return await this.datasetProcessor.uploadDataset(this.signer.privateKey, dataPath, gasPrice || this._gasPrice, maxGasPrice || this._maxGasPrice);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    downloadDataset = async (dataPath, dataRoot) => {
        try {
            await this.datasetProcessor.downloadDataset(dataPath, dataRoot);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    calculateToken = async (datasetPath, preTrainedModelName, usePython, providerAddress) => {
        try {
            await this.datasetProcessor.calculateToken(datasetPath, usePython, preTrainedModelName, providerAddress);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    createTask = async (providerAddress, preTrainedModelName, datasetHash, trainingPath, gasPrice) => {
        try {
            return await this.serviceProcessor.createTask(providerAddress, preTrainedModelName, datasetHash, trainingPath, gasPrice);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    cancelTask = async (providerAddress, taskID) => {
        try {
            return await this.serviceProcessor.cancelTask(providerAddress, taskID);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    listTask = async (providerAddress) => {
        try {
            return await this.serviceProcessor.listTask(providerAddress);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    getTask = async (providerAddress, taskID) => {
        try {
            const task = await this.serviceProcessor.getTask(providerAddress, taskID);
            return task;
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    getLog = async (providerAddress, taskID) => {
        try {
            return await this.serviceProcessor.getLog(providerAddress, taskID);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    acknowledgeModel = async (providerAddress, taskId, dataPath, options) => {
        try {
            return await this.modelProcessor.acknowledgeModel(providerAddress, taskId, dataPath, options);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    /**
     * Download LoRA model directly from TEE
     */
    downloadLoRAFromTEE = async (providerAddress, taskId, outputPath) => {
        try {
            return await this.modelProcessor.downloadLoRAFromTEE(providerAddress, taskId, outputPath);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    /**
     * Upload dataset directly to TEE (broker)
     * Returns the dataset hash for use in task creation
     * This is the preferred method for testing
     */
    uploadDatasetToTEE = async (providerAddress, datasetPath) => {
        try {
            return await this.serviceProvider.uploadDatasetToTEE(providerAddress, datasetPath);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    /**
     * Download encrypted model from 0G Storage (for advanced use cases)
     */
    downloadModelFrom0GStorage = async (providerAddress, taskId, dataPath) => {
        try {
            return await this.modelProcessor.downloadModelFrom0GStorage(providerAddress, taskId, dataPath);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    decryptModel = async (providerAddress, taskId, encryptedModelPath, decryptedModelPath) => {
        try {
            return await this.modelProcessor.decryptModel(providerAddress, taskId, encryptedModelPath, decryptedModelPath);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    /**
     * Verify fine-tuning service TEE attestation (DStack only)
     *
     * Downloads and verifies the TEE attestation report to ensure the provider
     * is running in a trusted execution environment. This simplified version
     * only supports DStack (Intel TDX) verification.
     *
     * @param providerAddress - The provider address to verify
     * @param outputDir - Directory to save attestation reports (default: current directory)
     * @returns Promise resolving to verification results with report paths and status
     *
     * @example
     * ```typescript
     * const broker = await createFineTuningBroker(signer, contractAddress, ledger);
     *
     * // Verify a provider's TEE attestation
     * const result = await broker.verifyService(
     *   '0x1234...',
     *   './attestation-reports'
     * );
     *
     * if (result.success && result.reportsData) {
     *   console.log('Verification successful');
     *   console.log('Reports saved to:', result.outputDirectory);
     * }
     * ```
     *
     * @remarks
     * This method downloads the attestation report and performs automated checks:
     * 1. TEE Signer Address Verification - matches contract signer with report signer
     * 2. Docker Compose Verification - validates compose hash against event log
     *
     * After automated checks, users must manually verify:
     * 1. Docker images using sigstore (https://search.sigstore.dev/)
     * 2. Run dstack-verifier for complete quote verification
     *
     * @throws {Error} If provider doesn't exist or attestation report cannot be retrieved
     */
    verifyService = async (providerAddress, outputDir = '.', onLog) => {
        try {
            return await this.verifier.verifyService(providerAddress, outputDir, onLog);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
}
exports.FineTuningBroker = FineTuningBroker;
/**
 * createFineTuningBroker is used to initialize ZGServingUserBroker
 *
 * @param signer - Signer from ethers.js.
 * @param contractAddress - 0G Serving contract address, use default address if not provided.
 * @param ledger - Ledger broker instance.
 * @param gasPrice - Gas price for transactions. If not provided, the gas price will be calculated automatically.
 *
 * @returns broker instance.
 *
 * @throws An error if the broker cannot be initialized.
 */
async function createFineTuningBroker(signer, contractAddress, ledger, gasPrice, maxGasPrice, step) {
    const broker = new FineTuningBroker(signer, contractAddress, ledger, gasPrice, maxGasPrice, step);
    try {
        await broker.initialize();
        return broker;
    }
    catch (error) {
        throw error;
    }
}
