// @0glabs/0g-serving-broker/cli.commonjs/sdk/fine-tuning/broker/read-only-broker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReadOnlyFineTuningBroker = void 0;
exports.createReadOnlyFineTuningBroker = createReadOnlyFineTuningBroker;
const ethers_1 = require("ethers");
const contract_1 = require("../contract");
const read_only_service_1 = require("./read-only-service");
const read_only_model_1 = require("./read-only-model");
const constants_1 = require("../../constants");
/**
 * Read-only fine-tuning broker with operations that don't require authentication.
 * Can be used with a Provider (read-only) without wallet connection.
 *
 * Use this broker to:
 * - List available fine-tuning providers before connecting wallet
 * - List available models without authentication
 * - Browse services without authentication
 */
class ReadOnlyFineTuningBroker {
    serviceProcessor;
    modelProcessor;
    constructor(provider, contractAddress) {
        const contract = new contract_1.ReadOnlyFineTuningServingContract(provider, contractAddress);
        this.serviceProcessor = new read_only_service_1.ReadOnlyServiceProcessor(contract);
        this.modelProcessor = new read_only_model_1.ReadOnlyModelProcessor(contract);
    }
    /**
     * Retrieves a list of fine-tuning services from the contract.
     * This is a read-only operation that doesn't require authentication.
     *
     * @param {boolean} includeUnacknowledged - Whether to include providers whose TEE signer is not acknowledged (default: false).
     * @returns {Promise<ServiceStructOutput[]>} A promise that resolves to an array of ServiceStructOutput objects.
     * @throws An error if the service list cannot be retrieved.
     */
    async listService(includeUnacknowledged = false) {
        return this.serviceProcessor.listService(includeUnacknowledged);
    }
    /**
     * List all available models including both standard pre-trained models
     * and customized models from providers.
     * This is a read-only operation that doesn't require authentication.
     *
     * @returns A tuple containing two arrays:
     *   - [0]: Standard pre-trained models with their configurations
     *   - [1]: Customized models from providers with descriptions
     */
    async listModel() {
        return this.modelProcessor.listModel();
    }
}
exports.ReadOnlyFineTuningBroker = ReadOnlyFineTuningBroker;
/**
 * Factory function to create a read-only fine-tuning broker.
 * No authentication required - perfect for listing providers without wallet connection.
 *
 * @param rpcUrl - JSON-RPC endpoint URL (e.g., https://evmrpc-testnet.0g.ai)
 * @param chainId - Optional chain ID (auto-detected if not provided)
 * @returns Read-only fine-tuning broker instance
 */
async function createReadOnlyFineTuningBroker(rpcUrl, chainId) {
    const provider = new ethers_1.JsonRpcProvider(rpcUrl);
    const detectedChainId = chainId ?? Number((await provider.getNetwork()).chainId);
    let fineTuningCA;
    if (detectedChainId === Number(constants_1.MAINNET_CHAIN_ID)) {
        fineTuningCA = constants_1.CONTRACT_ADDRESSES.mainnet.fineTuning;
    }
    else if (detectedChainId === Number(constants_1.TESTNET_CHAIN_ID)) {
        fineTuningCA = (0, constants_1.isDevMode)()
            ? constants_1.CONTRACT_ADDRESSES.testnetDev.fineTuning
            : constants_1.CONTRACT_ADDRESSES.testnet.fineTuning;
    }
    else if (detectedChainId === Number(constants_1.HARDHAT_CHAIN_ID)) {
        fineTuningCA = constants_1.CONTRACT_ADDRESSES.hardhat.fineTuning;
    }
    else {
        fineTuningCA = constants_1.CONTRACT_ADDRESSES.testnet.fineTuning;
    }
    return new ReadOnlyFineTuningBroker(provider, fineTuningCA);
}
