// @uniswap/sdk-core/dist/esm/src/chains.js
// Supported chain IDs for Uniswap deployments
export var ChainId;
(function (ChainId) {
    ChainId[ChainId["MAINNET"] = 1] = "MAINNET";
    ChainId[ChainId["GOERLI"] = 5] = "GOERLI";
    ChainId[ChainId["SEPOLIA"] = 11155111] = "SEPOLIA";
    ChainId[ChainId["OPTIMISM"] = 10] = "OPTIMISM";
    ChainId[ChainId["OPTIMISM_GOERLI"] = 420] = "OPTIMISM_GOERLI";
    ChainId[ChainId["OPTIMISM_SEPOLIA"] = 11155420] = "OPTIMISM_SEPOLIA";
    ChainId[ChainId["ARBITRUM_ONE"] = 42161] = "ARBITRUM_ONE";
    ChainId[ChainId["ARBITRUM_GOERLI"] = 421613] = "ARBITRUM_GOERLI";
    ChainId[ChainId["ARBITRUM_SEPOLIA"] = 421614] = "ARBITRUM_SEPOLIA";
    ChainId[ChainId["POLYGON"] = 137] = "POLYGON";
    ChainId[ChainId["POLYGON_MUMBAI"] = 80001] = "POLYGON_MUMBAI";
    ChainId[ChainId["CELO"] = 42220] = "CELO";
    ChainId[ChainId["CELO_ALFAJORES"] = 44787] = "CELO_ALFAJORES";
    ChainId[ChainId["GNOSIS"] = 100] = "GNOSIS";
    ChainId[ChainId["MOONBEAM"] = 1284] = "MOONBEAM";
    ChainId[ChainId["BNB"] = 56] = "BNB";
    ChainId[ChainId["AVALANCHE"] = 43114] = "AVALANCHE";
    ChainId[ChainId["BASE_GOERLI"] = 84531] = "BASE_GOERLI";
    ChainId[ChainId["BASE_SEPOLIA"] = 84532] = "BASE_SEPOLIA";
    ChainId[ChainId["BASE"] = 8453] = "BASE";
    ChainId[ChainId["ZORA"] = 7777777] = "ZORA";
    ChainId[ChainId["ZORA_SEPOLIA"] = 999999999] = "ZORA_SEPOLIA";
    ChainId[ChainId["ROOTSTOCK"] = 30] = "ROOTSTOCK";
    ChainId[ChainId["BLAST"] = 81457] = "BLAST";
    ChainId[ChainId["ZKSYNC"] = 324] = "ZKSYNC";
    ChainId[ChainId["WORLDCHAIN"] = 480] = "WORLDCHAIN";
    ChainId[ChainId["UNICHAIN_SEPOLIA"] = 1301] = "UNICHAIN_SEPOLIA";
    ChainId[ChainId["UNICHAIN"] = 130] = "UNICHAIN";
    ChainId[ChainId["MONAD_TESTNET"] = 10143] = "MONAD_TESTNET";
    ChainId[ChainId["SONEIUM"] = 1868] = "SONEIUM";
    ChainId[ChainId["MONAD"] = 143] = "MONAD";
    ChainId[ChainId["XLAYER"] = 196] = "XLAYER";
    ChainId[ChainId["LINEA"] = 59144] = "LINEA";
    ChainId[ChainId["TEMPO"] = 4217] = "TEMPO";
})(ChainId || (ChainId = {}));
export const SUPPORTED_CHAINS = [
    ChainId.MAINNET,
    ChainId.OPTIMISM,
    ChainId.OPTIMISM_GOERLI,
    ChainId.OPTIMISM_SEPOLIA,
    ChainId.ARBITRUM_ONE,
    ChainId.ARBITRUM_GOERLI,
    ChainId.ARBITRUM_SEPOLIA,
    ChainId.POLYGON,
    ChainId.POLYGON_MUMBAI,
    ChainId.GOERLI,
    ChainId.SEPOLIA,
    ChainId.CELO_ALFAJORES,
    ChainId.CELO,
    ChainId.BNB,
    ChainId.AVALANCHE,
    ChainId.BASE,
    ChainId.BASE_GOERLI,
    ChainId.BASE_SEPOLIA,
    ChainId.ZORA,
    ChainId.ZORA_SEPOLIA,
    ChainId.ROOTSTOCK,
    ChainId.BLAST,
    ChainId.ZKSYNC,
    ChainId.WORLDCHAIN,
    ChainId.UNICHAIN_SEPOLIA,
    ChainId.UNICHAIN,
    ChainId.MONAD_TESTNET,
    ChainId.SONEIUM,
    ChainId.MONAD,
    ChainId.XLAYER,
    ChainId.LINEA,
    ChainId.TEMPO,
];
export var NativeCurrencyName;
(function (NativeCurrencyName) {
    // Strings match input for CLI
    NativeCurrencyName["ETHER"] = "ETH";
    NativeCurrencyName["MATIC"] = "MATIC";
    NativeCurrencyName["CELO"] = "CELO";
    NativeCurrencyName["GNOSIS"] = "XDAI";
    NativeCurrencyName["MOONBEAM"] = "GLMR";
    NativeCurrencyName["BNB"] = "BNB";
    NativeCurrencyName["AVAX"] = "AVAX";
    NativeCurrencyName["ROOTSTOCK"] = "RBTC";
})(NativeCurrencyName || (NativeCurrencyName = {}));
