// @uniswap/v3-sdk/dist/cjs/src/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TICK_SPACINGS = exports.FeeAmount = exports.poolInitCodeHash = exports.POOL_INIT_CODE_HASH = exports.ADDRESS_ZERO = exports.FACTORY_ADDRESS = void 0;
const sdk_core_1 = require("@uniswap/sdk-core");
exports.FACTORY_ADDRESS = '0x1F98431c8aD98523631AE4a59f267346ea31F984';
exports.ADDRESS_ZERO = '0x0000000000000000000000000000000000000000';
// @deprecated please use poolInitCodeHash(chainId: ChainId)
exports.POOL_INIT_CODE_HASH = '0xe34f199b19b2b4f47f68442619d555527d244f78a3297ea89325f843f87b8b54';
function poolInitCodeHash(chainId) {
    switch (chainId) {
        case sdk_core_1.ChainId.ZKSYNC:
            return '0x010013f177ea1fcbc4520f9a3ca7cd2d1d77959e05aa66484027cb38e712aeed';
        default:
            return exports.POOL_INIT_CODE_HASH;
    }
}
exports.poolInitCodeHash = poolInitCodeHash;
/**
 * The default factory enabled fee amounts, denominated in hundredths of bips.
 */
var FeeAmount;
(function (FeeAmount) {
    FeeAmount[FeeAmount["LOWEST"] = 100] = "LOWEST";
    FeeAmount[FeeAmount["LOW_200"] = 200] = "LOW_200";
    FeeAmount[FeeAmount["LOW_300"] = 300] = "LOW_300";
    FeeAmount[FeeAmount["LOW_400"] = 400] = "LOW_400";
    FeeAmount[FeeAmount["LOW"] = 500] = "LOW";
    FeeAmount[FeeAmount["MEDIUM"] = 3000] = "MEDIUM";
    FeeAmount[FeeAmount["HIGH"] = 10000] = "HIGH";
})(FeeAmount = exports.FeeAmount || (exports.FeeAmount = {}));
/**
 * The default factory tick spacings by fee amount.
 */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
exports.TICK_SPACINGS = {
    [FeeAmount.LOWEST]: 1,
    [FeeAmount.LOW_200]: 4,
    [FeeAmount.LOW_300]: 6,
    [FeeAmount.LOW_400]: 8,
    [FeeAmount.LOW]: 10,
    [FeeAmount.MEDIUM]: 60,
    [FeeAmount.HIGH]: 200,
};
