// @uniswap/v3-sdk/dist/cjs/src/nonfungiblePositionManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NonfungiblePositionManager = void 0;
const tslib_1 = require("tslib");
const sdk_core_1 = require("@uniswap/sdk-core");
const jsbi_1 = tslib_1.__importDefault(require("jsbi"));
const tiny_invariant_1 = tslib_1.__importDefault(require("tiny-invariant"));
const position_1 = require("./entities/position");
const internalConstants_1 = require("./internalConstants");
const calldata_1 = require("./utils/calldata");
const abi_1 = require("@ethersproject/abi");
const NonfungiblePositionManager_json_1 = tslib_1.__importDefault(require("@uniswap/v3-periphery/artifacts/contracts/NonfungiblePositionManager.sol/NonfungiblePositionManager.json"));
const selfPermit_1 = require("./selfPermit");
const constants_1 = require("./constants");
const multicall_1 = require("./multicall");
const payments_1 = require("./payments");
const MaxUint128 = (0, calldata_1.toHex)(jsbi_1.default.subtract(jsbi_1.default.exponentiate(jsbi_1.default.BigInt(2), jsbi_1.default.BigInt(128)), jsbi_1.default.BigInt(1)));
// type guard
function isMint(options) {
    return Object.keys(options).some((k) => k === 'recipient');
}
const NFT_PERMIT_TYPES = {
    Permit: [
        { name: 'spender', type: 'address' },
        { name: 'tokenId', type: 'uint256' },
        { name: 'nonce', type: 'uint256' },
        { name: 'deadline', type: 'uint256' },
    ],
};
class NonfungiblePositionManager {
    /**
     * Cannot be constructed.
     */
    constructor() { }
    static encodeCreate(pool) {
        return NonfungiblePositionManager.INTERFACE.encodeFunctionData('createAndInitializePoolIfNecessary', [
            pool.token0.address,
            pool.token1.address,
            pool.fee,
            (0, calldata_1.toHex)(pool.sqrtRatioX96),
        ]);
    }
    static createCallParameters(pool) {
        return {
            calldata: this.encodeCreate(pool),
            value: (0, calldata_1.toHex)(0),
        };
    }
    static addCallParameters(position, options) {
        (0, tiny_invariant_1.default)(jsbi_1.default.greaterThan(position.liquidity, internalConstants_1.ZERO), 'ZERO_LIQUIDITY');
        const calldatas = [];
        // get amounts
        const { amount0: amount0Desired, amount1: amount1Desired } = position.mintAmounts;
        // adjust for slippage
        const minimumAmounts = position.mintAmountsWithSlippage(options.slippageTolerance);
        const amount0Min = (0, calldata_1.toHex)(minimumAmounts.amount0);
        const amount1Min = (0, calldata_1.toHex)(minimumAmounts.amount1);
        const deadline = (0, calldata_1.toHex)(options.deadline);
        // create pool if needed
        if (isMint(options) && options.createPool) {
            calldatas.push(this.encodeCreate(position.pool));
        }
        // permits if necessary
        if (options.token0Permit) {
            calldatas.push(selfPermit_1.SelfPermit.encodePermit(position.pool.token0, options.token0Permit));
        }
        if (options.token1Permit) {
            calldatas.push(selfPermit_1.SelfPermit.encodePermit(position.pool.token1, options.token1Permit));
        }
        // mint
        if (isMint(options)) {
            const recipient = (0, sdk_core_1.validateAndParseAddress)(options.recipient);
            calldatas.push(NonfungiblePositionManager.INTERFACE.encodeFunctionData('mint', [
                {
                    token0: position.pool.token0.address,
                    token1: position.pool.token1.address,
                    fee: position.pool.fee,
                    tickLower: position.tickLower,
                    tickUpper: position.tickUpper,
                    amount0Desired: (0, calldata_1.toHex)(amount0Desired),
                    amount1Desired: (0, calldata_1.toHex)(amount1Desired),
                    amount0Min,
                    amount1Min,
                    recipient,
                    deadline,
                },
            ]));
        }
        else {
            // increase
            calldatas.push(NonfungiblePositionManager.INTERFACE.encodeFunctionData('increaseLiquidity', [
                {
                    tokenId: (0, calldata_1.toHex)(options.tokenId),
                    amount0Desired: (0, calldata_1.toHex)(amount0Desired),
                    amount1Desired: (0, calldata_1.toHex)(amount1Desired),
                    amount0Min,
                    amount1Min,
                    deadline,
                },
            ]));
        }
        let value = (0, calldata_1.toHex)(0);
        if (options.useNative) {
            const wrapped = options.useNative.wrapped;
            (0, tiny_invariant_1.default)(position.pool.token0.equals(wrapped) || position.pool.token1.equals(wrapped), 'NO_WETH');
            const wrappedValue = position.pool.token0.equals(wrapped) ? amount0Desired : amount1Desired;
            // we only need to refund if we're actually sending ETH
            if (jsbi_1.default.greaterThan(wrappedValue, internalConstants_1.ZERO)) {
                calldatas.push(payments_1.Payments.encodeRefundETH());
            }
            value = (0, calldata_1.toHex)(wrappedValue);
        }
        return {
            calldata: multicall_1.Multicall.encodeMulticall(calldatas),
            value,
        };
    }
    static encodeCollect(options) {
        const calldatas = [];
        const tokenId = (0, calldata_1.toHex)(options.tokenId);
        const involvesETH = options.expectedCurrencyOwed0.currency.isNative || options.expectedCurrencyOwed1.currency.isNative;
        const recipient = (0, sdk_core_1.validateAndParseAddress)(options.recipient);
        // collect
        calldatas.push(NonfungiblePositionManager.INTERFACE.encodeFunctionData('collect', [
            {
                tokenId,
                recipient: involvesETH ? constants_1.ADDRESS_ZERO : recipient,
                amount0Max: MaxUint128,
                amount1Max: MaxUint128,
            },
        ]));
        if (involvesETH) {
            const ethAmount = options.expectedCurrencyOwed0.currency.isNative
                ? options.expectedCurrencyOwed0.quotient
                : options.expectedCurrencyOwed1.quotient;
            const token = options.expectedCurrencyOwed0.currency.isNative
                ? options.expectedCurrencyOwed1.currency
                : options.expectedCurrencyOwed0.currency;
            const tokenAmount = options.expectedCurrencyOwed0.currency.isNative
                ? options.expectedCurrencyOwed1.quotient
                : options.expectedCurrencyOwed0.quotient;
            calldatas.push(payments_1.Payments.encodeUnwrapWETH9(ethAmount, recipient));
            calldatas.push(payments_1.Payments.encodeSweepToken(token, tokenAmount, recipient));
        }
        return calldatas;
    }
    static collectCallParameters(options) {
        const calldatas = NonfungiblePositionManager.encodeCollect(options);
        return {
            calldata: multicall_1.Multicall.encodeMulticall(calldatas),
            value: (0, calldata_1.toHex)(0),
        };
    }
    /**
     * Produces the calldata for completely or partially exiting a position
     * @param position The position to exit
     * @param options Additional information necessary for generating the calldata
     * @returns The call parameters
     */
    static removeCallParameters(position, options) {
        const calldatas = [];
        const deadline = (0, calldata_1.toHex)(options.deadline);
        const tokenId = (0, calldata_1.toHex)(options.tokenId);
        // construct a partial position with a percentage of liquidity
        const partialPosition = new position_1.Position({
            pool: position.pool,
            liquidity: options.liquidityPercentage.multiply(position.liquidity).quotient,
            tickLower: position.tickLower,
            tickUpper: position.tickUpper,
        });
        (0, tiny_invariant_1.default)(jsbi_1.default.greaterThan(partialPosition.liquidity, internalConstants_1.ZERO), 'ZERO_LIQUIDITY');
        // slippage-adjusted underlying amounts
        const { amount0: amount0Min, amount1: amount1Min } = partialPosition.burnAmountsWithSlippage(options.slippageTolerance);
        if (options.permit) {
            calldatas.push(NonfungiblePositionManager.INTERFACE.encodeFunctionData('permit', [
                (0, sdk_core_1.validateAndParseAddress)(options.permit.spender),
                tokenId,
                (0, calldata_1.toHex)(options.permit.deadline),
                options.permit.v,
                options.permit.r,
                options.permit.s,
            ]));
        }
        // remove liquidity
        calldatas.push(NonfungiblePositionManager.INTERFACE.encodeFunctionData('decreaseLiquidity', [
            {
                tokenId,
                liquidity: (0, calldata_1.toHex)(partialPosition.liquidity),
                amount0Min: (0, calldata_1.toHex)(amount0Min),
                amount1Min: (0, calldata_1.toHex)(amount1Min),
                deadline,
            },
        ]));
        const { expectedCurrencyOwed0, expectedCurrencyOwed1, ...rest } = options.collectOptions;
        calldatas.push(...NonfungiblePositionManager.encodeCollect({
            tokenId: (0, calldata_1.toHex)(options.tokenId),
            // add the underlying value to the expected currency already owed
            expectedCurrencyOwed0: expectedCurrencyOwed0.add(sdk_core_1.CurrencyAmount.fromRawAmount(expectedCurrencyOwed0.currency, amount0Min)),
            expectedCurrencyOwed1: expectedCurrencyOwed1.add(sdk_core_1.CurrencyAmount.fromRawAmount(expectedCurrencyOwed1.currency, amount1Min)),
            ...rest,
        }));
        if (options.liquidityPercentage.equalTo(internalConstants_1.ONE)) {
            if (options.burnToken) {
                calldatas.push(NonfungiblePositionManager.INTERFACE.encodeFunctionData('burn', [tokenId]));
            }
        }
        else {
            (0, tiny_invariant_1.default)(options.burnToken !== true, 'CANNOT_BURN');
        }
        return {
            calldata: multicall_1.Multicall.encodeMulticall(calldatas),
            value: (0, calldata_1.toHex)(0),
        };
    }
    static safeTransferFromParameters(options) {
        const recipient = (0, sdk_core_1.validateAndParseAddress)(options.recipient);
        const sender = (0, sdk_core_1.validateAndParseAddress)(options.sender);
        let calldata;
        if (options.data) {
            calldata = NonfungiblePositionManager.INTERFACE.encodeFunctionData('safeTransferFrom(address,address,uint256,bytes)', [sender, recipient, (0, calldata_1.toHex)(options.tokenId), options.data]);
        }
        else {
            calldata = NonfungiblePositionManager.INTERFACE.encodeFunctionData('safeTransferFrom(address,address,uint256)', [
                sender,
                recipient,
                (0, calldata_1.toHex)(options.tokenId),
            ]);
        }
        return {
            calldata: calldata,
            value: (0, calldata_1.toHex)(0),
        };
    }
    // Prepare the params for an EIP712 signTypedData request
    static getPermitData(permit, positionManagerAddress, chainId) {
        return {
            domain: {
                name: 'Uniswap V3 Positions NFT-V1',
                chainId,
                version: '1',
                verifyingContract: positionManagerAddress,
            },
            types: NFT_PERMIT_TYPES,
            values: permit,
        };
    }
}
exports.NonfungiblePositionManager = NonfungiblePositionManager;
NonfungiblePositionManager.INTERFACE = new abi_1.Interface(NonfungiblePositionManager_json_1.default.abi);
