// @uniswap/v3-sdk/dist/cjs/src/quoter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SwapQuoter = void 0;
const tslib_1 = require("tslib");
const abi_1 = require("@ethersproject/abi");
const sdk_core_1 = require("@uniswap/sdk-core");
const utils_1 = require("./utils");
const Quoter_json_1 = tslib_1.__importDefault(require("@uniswap/v3-periphery/artifacts/contracts/lens/Quoter.sol/Quoter.json"));
const QuoterV2_json_1 = tslib_1.__importDefault(require("@uniswap/swap-router-contracts/artifacts/contracts/lens/QuoterV2.sol/QuoterV2.json"));
const tiny_invariant_1 = tslib_1.__importDefault(require("tiny-invariant"));
/**
 * Represents the Uniswap V3 QuoterV1 contract with a method for returning the formatted
 * calldata needed to call the quoter contract.
 */
class SwapQuoter {
    /**
     * Produces the on-chain method name of the appropriate function within QuoterV2,
     * and the relevant hex encoded parameters.
     * @template TInput The input token, either Ether or an ERC-20
     * @template TOutput The output token, either Ether or an ERC-20
     * @param route The swap route, a list of pools through which a swap can occur
     * @param amount The amount of the quote, either an amount in, or an amount out
     * @param tradeType The trade type, either exact input or exact output
     * @param options The optional params including price limit and Quoter contract switch
     * @returns The formatted calldata
     */
    static quoteCallParameters(route, amount, tradeType, options = {}) {
        var _a;
        const singleHop = route.pools.length === 1;
        const quoteAmount = (0, utils_1.toHex)(amount.quotient);
        let calldata;
        const swapInterface = options.useQuoterV2 ? this.V2INTERFACE : this.V1INTERFACE;
        if (singleHop) {
            const baseQuoteParams = {
                tokenIn: route.tokenPath[0].address,
                tokenOut: route.tokenPath[1].address,
                fee: route.pools[0].fee,
                sqrtPriceLimitX96: (0, utils_1.toHex)((_a = options === null || options === void 0 ? void 0 : options.sqrtPriceLimitX96) !== null && _a !== void 0 ? _a : 0),
            };
            const v2QuoteParams = {
                ...baseQuoteParams,
                ...(tradeType === sdk_core_1.TradeType.EXACT_INPUT ? { amountIn: quoteAmount } : { amount: quoteAmount }),
            };
            const v1QuoteParams = [
                baseQuoteParams.tokenIn,
                baseQuoteParams.tokenOut,
                baseQuoteParams.fee,
                quoteAmount,
                baseQuoteParams.sqrtPriceLimitX96,
            ];
            const tradeTypeFunctionName = tradeType === sdk_core_1.TradeType.EXACT_INPUT ? 'quoteExactInputSingle' : 'quoteExactOutputSingle';
            calldata = swapInterface.encodeFunctionData(tradeTypeFunctionName, options.useQuoterV2 ? [v2QuoteParams] : v1QuoteParams);
        }
        else {
            (0, tiny_invariant_1.default)((options === null || options === void 0 ? void 0 : options.sqrtPriceLimitX96) === undefined, 'MULTIHOP_PRICE_LIMIT');
            const path = (0, utils_1.encodeRouteToPath)(route, tradeType === sdk_core_1.TradeType.EXACT_OUTPUT);
            const tradeTypeFunctionName = tradeType === sdk_core_1.TradeType.EXACT_INPUT ? 'quoteExactInput' : 'quoteExactOutput';
            calldata = swapInterface.encodeFunctionData(tradeTypeFunctionName, [path, quoteAmount]);
        }
        return {
            calldata,
            value: (0, utils_1.toHex)(0),
        };
    }
}
exports.SwapQuoter = SwapQuoter;
SwapQuoter.V1INTERFACE = new abi_1.Interface(Quoter_json_1.default.abi);
SwapQuoter.V2INTERFACE = new abi_1.Interface(QuoterV2_json_1.default.abi);
