// @uniswap/v3-sdk/dist/cjs/src/staker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Staker = void 0;
const tslib_1 = require("tslib");
const sdk_core_1 = require("@uniswap/sdk-core");
const calldata_1 = require("./utils/calldata");
const abi_1 = require("@ethersproject/abi");
const UniswapV3Staker_json_1 = tslib_1.__importDefault(require("@uniswap/v3-staker/artifacts/contracts/UniswapV3Staker.sol/UniswapV3Staker.json"));
const entities_1 = require("./entities");
const multicall_1 = require("./multicall");
class Staker {
    constructor() { }
    /**
     *  To claim rewards, must unstake and then claim.
     * @param incentiveKey The unique identifier of a staking program.
     * @param options Options for producing the calldata to claim. Can't claim unless you unstake.
     * @returns The calldatas for 'unstakeToken' and 'claimReward'.
     */
    static encodeClaim(incentiveKey, options) {
        var _a;
        const calldatas = [];
        calldatas.push(Staker.INTERFACE.encodeFunctionData('unstakeToken', [
            this._encodeIncentiveKey(incentiveKey),
            (0, calldata_1.toHex)(options.tokenId),
        ]));
        const recipient = (0, sdk_core_1.validateAndParseAddress)(options.recipient);
        const amount = (_a = options.amount) !== null && _a !== void 0 ? _a : 0;
        calldatas.push(Staker.INTERFACE.encodeFunctionData('claimReward', [incentiveKey.rewardToken.address, recipient, (0, calldata_1.toHex)(amount)]));
        return calldatas;
    }
    /**
     *
     * Note:  A `tokenId` can be staked in many programs but to claim rewards and continue the program you must unstake, claim, and then restake.
     * @param incentiveKeys An IncentiveKey or array of IncentiveKeys that `tokenId` is staked in.
     * Input an array of IncentiveKeys to claim rewards for each program.
     * @param options ClaimOptions to specify tokenId, recipient, and amount wanting to collect.
     * Note that you can only specify one amount and one recipient across the various programs if you are collecting from multiple programs at once.
     * @returns
     */
    static collectRewards(incentiveKeys, options) {
        incentiveKeys = Array.isArray(incentiveKeys) ? incentiveKeys : [incentiveKeys];
        let calldatas = [];
        for (let i = 0; i < incentiveKeys.length; i++) {
            // the unique program tokenId is staked in
            const incentiveKey = incentiveKeys[i];
            // unstakes and claims for the unique program
            calldatas = calldatas.concat(this.encodeClaim(incentiveKey, options));
            // re-stakes the position for the unique program
            calldatas.push(Staker.INTERFACE.encodeFunctionData('stakeToken', [
                this._encodeIncentiveKey(incentiveKey),
                (0, calldata_1.toHex)(options.tokenId),
            ]));
        }
        return {
            calldata: multicall_1.Multicall.encodeMulticall(calldatas),
            value: (0, calldata_1.toHex)(0),
        };
    }
    /**
     *
     * @param incentiveKeys A list of incentiveKeys to unstake from. Should include all incentiveKeys (unique staking programs) that `options.tokenId` is staked in.
     * @param withdrawOptions Options for producing claim calldata and withdraw calldata. Can't withdraw without unstaking all programs for `tokenId`.
     * @returns Calldata for unstaking, claiming, and withdrawing.
     */
    static withdrawToken(incentiveKeys, withdrawOptions) {
        let calldatas = [];
        incentiveKeys = Array.isArray(incentiveKeys) ? incentiveKeys : [incentiveKeys];
        const claimOptions = {
            tokenId: withdrawOptions.tokenId,
            recipient: withdrawOptions.recipient,
            amount: withdrawOptions.amount,
        };
        for (let i = 0; i < incentiveKeys.length; i++) {
            const incentiveKey = incentiveKeys[i];
            calldatas = calldatas.concat(this.encodeClaim(incentiveKey, claimOptions));
        }
        const owner = (0, sdk_core_1.validateAndParseAddress)(withdrawOptions.owner);
        calldatas.push(Staker.INTERFACE.encodeFunctionData('withdrawToken', [
            (0, calldata_1.toHex)(withdrawOptions.tokenId),
            owner,
            withdrawOptions.data ? withdrawOptions.data : (0, calldata_1.toHex)(0),
        ]));
        return {
            calldata: multicall_1.Multicall.encodeMulticall(calldatas),
            value: (0, calldata_1.toHex)(0),
        };
    }
    /**
     *
     * @param incentiveKeys A single IncentiveKey or array of IncentiveKeys to be encoded and used in the data parameter in `safeTransferFrom`
     * @returns An IncentiveKey as a string
     */
    static encodeDeposit(incentiveKeys) {
        incentiveKeys = Array.isArray(incentiveKeys) ? incentiveKeys : [incentiveKeys];
        let data;
        if (incentiveKeys.length > 1) {
            const keys = [];
            for (let i = 0; i < incentiveKeys.length; i++) {
                const incentiveKey = incentiveKeys[i];
                keys.push(this._encodeIncentiveKey(incentiveKey));
            }
            data = abi_1.defaultAbiCoder.encode([`${Staker.INCENTIVE_KEY_ABI}[]`], [keys]);
        }
        else {
            data = abi_1.defaultAbiCoder.encode([Staker.INCENTIVE_KEY_ABI], [this._encodeIncentiveKey(incentiveKeys[0])]);
        }
        return data;
    }
    /**
     *
     * @param incentiveKey An `IncentiveKey` which represents a unique staking program.
     * @returns An encoded IncentiveKey to be read by ethers
     */
    static _encodeIncentiveKey(incentiveKey) {
        const { token0, token1, fee } = incentiveKey.pool;
        const refundee = (0, sdk_core_1.validateAndParseAddress)(incentiveKey.refundee);
        return {
            rewardToken: incentiveKey.rewardToken.address,
            pool: entities_1.Pool.getAddress(token0, token1, fee),
            startTime: (0, calldata_1.toHex)(incentiveKey.startTime),
            endTime: (0, calldata_1.toHex)(incentiveKey.endTime),
            refundee,
        };
    }
}
exports.Staker = Staker;
Staker.INTERFACE = new abi_1.Interface(UniswapV3Staker_json_1.default.abi);
Staker.INCENTIVE_KEY_ABI = 'tuple(address rewardToken, address pool, uint256 startTime, uint256 endTime, address refundee)';
