// @uniswap/v3-sdk/dist/cjs/src/swapRouter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SwapRouter = void 0;
const tslib_1 = require("tslib");
const abi_1 = require("@ethersproject/abi");
const sdk_core_1 = require("@uniswap/sdk-core");
const tiny_invariant_1 = tslib_1.__importDefault(require("tiny-invariant"));
const constants_1 = require("./constants");
const selfPermit_1 = require("./selfPermit");
const utils_1 = require("./utils");
const calldata_1 = require("./utils/calldata");
const SwapRouter_json_1 = tslib_1.__importDefault(require("@uniswap/v3-periphery/artifacts/contracts/SwapRouter.sol/SwapRouter.json"));
const multicall_1 = require("./multicall");
const payments_1 = require("./payments");
/**
 * Represents the Uniswap V3 SwapRouter, and has static methods for helping execute trades.
 */
class SwapRouter {
    /**
     * Cannot be constructed.
     */
    constructor() { }
    /**
     * Produces the on-chain method name to call and the hex encoded parameters to pass as arguments for a given trade.
     * @param trade to produce call parameters for
     * @param options options for the call parameters
     */
    static swapCallParameters(trades, options) {
        var _a, _b;
        if (!Array.isArray(trades)) {
            trades = [trades];
        }
        const sampleTrade = trades[0];
        const tokenIn = sampleTrade.inputAmount.currency.wrapped;
        const tokenOut = sampleTrade.outputAmount.currency.wrapped;
        // All trades should have the same starting and ending token.
        (0, tiny_invariant_1.default)(trades.every((trade) => trade.inputAmount.currency.wrapped.equals(tokenIn)), 'TOKEN_IN_DIFF');
        (0, tiny_invariant_1.default)(trades.every((trade) => trade.outputAmount.currency.wrapped.equals(tokenOut)), 'TOKEN_OUT_DIFF');
        const calldatas = [];
        const ZERO_IN = sdk_core_1.CurrencyAmount.fromRawAmount(trades[0].inputAmount.currency, 0);
        const ZERO_OUT = sdk_core_1.CurrencyAmount.fromRawAmount(trades[0].outputAmount.currency, 0);
        const totalAmountOut = trades.reduce((sum, trade) => sum.add(trade.minimumAmountOut(options.slippageTolerance)), ZERO_OUT);
        // flag for whether a refund needs to happen
        const mustRefund = sampleTrade.inputAmount.currency.isNative && sampleTrade.tradeType === sdk_core_1.TradeType.EXACT_OUTPUT;
        const inputIsNative = sampleTrade.inputAmount.currency.isNative;
        // flags for whether funds should be send first to the router
        const outputIsNative = sampleTrade.outputAmount.currency.isNative;
        const routerMustCustody = outputIsNative || !!options.fee;
        const totalValue = inputIsNative
            ? trades.reduce((sum, trade) => sum.add(trade.maximumAmountIn(options.slippageTolerance)), ZERO_IN)
            : ZERO_IN;
        // encode permit if necessary
        if (options.inputTokenPermit) {
            (0, tiny_invariant_1.default)(sampleTrade.inputAmount.currency.isToken, 'NON_TOKEN_PERMIT');
            calldatas.push(selfPermit_1.SelfPermit.encodePermit(sampleTrade.inputAmount.currency, options.inputTokenPermit));
        }
        const recipient = (0, sdk_core_1.validateAndParseAddress)(options.recipient);
        const deadline = (0, calldata_1.toHex)(options.deadline);
        for (const trade of trades) {
            for (const { route, inputAmount, outputAmount } of trade.swaps) {
                const amountIn = (0, calldata_1.toHex)(trade.maximumAmountIn(options.slippageTolerance, inputAmount).quotient);
                const amountOut = (0, calldata_1.toHex)(trade.minimumAmountOut(options.slippageTolerance, outputAmount).quotient);
                // flag for whether the trade is single hop or not
                const singleHop = route.pools.length === 1;
                if (singleHop) {
                    if (trade.tradeType === sdk_core_1.TradeType.EXACT_INPUT) {
                        const exactInputSingleParams = {
                            tokenIn: route.tokenPath[0].address,
                            tokenOut: route.tokenPath[1].address,
                            fee: route.pools[0].fee,
                            recipient: routerMustCustody ? constants_1.ADDRESS_ZERO : recipient,
                            deadline,
                            amountIn,
                            amountOutMinimum: amountOut,
                            sqrtPriceLimitX96: (0, calldata_1.toHex)((_a = options.sqrtPriceLimitX96) !== null && _a !== void 0 ? _a : 0),
                        };
                        calldatas.push(SwapRouter.INTERFACE.encodeFunctionData('exactInputSingle', [exactInputSingleParams]));
                    }
                    else {
                        const exactOutputSingleParams = {
                            tokenIn: route.tokenPath[0].address,
                            tokenOut: route.tokenPath[1].address,
                            fee: route.pools[0].fee,
                            recipient: routerMustCustody ? constants_1.ADDRESS_ZERO : recipient,
                            deadline,
                            amountOut,
                            amountInMaximum: amountIn,
                            sqrtPriceLimitX96: (0, calldata_1.toHex)((_b = options.sqrtPriceLimitX96) !== null && _b !== void 0 ? _b : 0),
                        };
                        calldatas.push(SwapRouter.INTERFACE.encodeFunctionData('exactOutputSingle', [exactOutputSingleParams]));
                    }
                }
                else {
                    (0, tiny_invariant_1.default)(options.sqrtPriceLimitX96 === undefined, 'MULTIHOP_PRICE_LIMIT');
                    const path = (0, utils_1.encodeRouteToPath)(route, trade.tradeType === sdk_core_1.TradeType.EXACT_OUTPUT);
                    if (trade.tradeType === sdk_core_1.TradeType.EXACT_INPUT) {
                        const exactInputParams = {
                            path,
                            recipient: routerMustCustody ? constants_1.ADDRESS_ZERO : recipient,
                            deadline,
                            amountIn,
                            amountOutMinimum: amountOut,
                        };
                        calldatas.push(SwapRouter.INTERFACE.encodeFunctionData('exactInput', [exactInputParams]));
                    }
                    else {
                        const exactOutputParams = {
                            path,
                            recipient: routerMustCustody ? constants_1.ADDRESS_ZERO : recipient,
                            deadline,
                            amountOut,
                            amountInMaximum: amountIn,
                        };
                        calldatas.push(SwapRouter.INTERFACE.encodeFunctionData('exactOutput', [exactOutputParams]));
                    }
                }
            }
        }
        // unwrap
        if (routerMustCustody) {
            if (!!options.fee) {
                if (outputIsNative) {
                    calldatas.push(payments_1.Payments.encodeUnwrapWETH9(totalAmountOut.quotient, recipient, options.fee));
                }
                else {
                    calldatas.push(payments_1.Payments.encodeSweepToken(sampleTrade.outputAmount.currency.wrapped, totalAmountOut.quotient, recipient, options.fee));
                }
            }
            else {
                calldatas.push(payments_1.Payments.encodeUnwrapWETH9(totalAmountOut.quotient, recipient));
            }
        }
        // refund
        if (mustRefund) {
            calldatas.push(payments_1.Payments.encodeRefundETH());
        }
        return {
            calldata: multicall_1.Multicall.encodeMulticall(calldatas),
            value: (0, calldata_1.toHex)(totalValue.quotient),
        };
    }
}
exports.SwapRouter = SwapRouter;
SwapRouter.INTERFACE = new abi_1.Interface(SwapRouter_json_1.default.abi);
