// @uniswap/universal-router-sdk/dist/cjs/src/utils/commandParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseV3PathExactOut = exports.parseV3PathExactIn = exports.GenericCommandParser = exports.CommandParser = void 0;
const tslib_1 = require("tslib");
const ethers_1 = require("ethers");
const UniversalRouter_json_1 = tslib_1.__importDefault(require("@uniswap/universal-router/artifacts/contracts/UniversalRouter.sol/UniversalRouter.json"));
const abi_1 = require("@ethersproject/abi");
const v4_sdk_1 = require("@uniswap/v4-sdk");
const routerCommands_1 = require("../utils/routerCommands");
// Parses UniversalRouter V2 commands
class CommandParser {
    static parseCalldata(calldata) {
        const genericParser = new GenericCommandParser(routerCommands_1.COMMAND_DEFINITION);
        const txDescription = CommandParser.INTERFACE.parseTransaction({ data: calldata });
        const { commands, inputs } = txDescription.args;
        return genericParser.parse(commands, inputs);
    }
}
exports.CommandParser = CommandParser;
CommandParser.INTERFACE = new abi_1.Interface(UniversalRouter_json_1.default.abi);
// Parses commands based on given command definition
class GenericCommandParser {
    constructor(commandDefinition) {
        this.commandDefinition = commandDefinition;
    }
    parse(commands, inputs) {
        const commandTypes = GenericCommandParser.getCommands(commands);
        return {
            commands: commandTypes.map((commandType, i) => {
                const commandDef = this.commandDefinition[commandType];
                if (commandDef.parser === routerCommands_1.Parser.V4Actions) {
                    const { actions } = v4_sdk_1.V4BaseActionsParser.parseCalldata(inputs[i]);
                    return {
                        commandName: routerCommands_1.CommandType[commandType],
                        commandType,
                        params: v4RouterCallToParams(actions),
                    };
                }
                else if (commandDef.parser === routerCommands_1.Parser.Abi) {
                    const abiDef = commandDef.params;
                    const rawParams = ethers_1.ethers.utils.defaultAbiCoder.decode(abiDef.map((command) => command.type), inputs[i]);
                    const params = rawParams.map((param, j) => {
                        switch (abiDef[j].subparser) {
                            case routerCommands_1.Subparser.V3PathExactIn:
                                return {
                                    name: abiDef[j].name,
                                    value: parseV3PathExactIn(param),
                                };
                            case routerCommands_1.Subparser.V3PathExactOut:
                                return {
                                    name: abiDef[j].name,
                                    value: parseV3PathExactOut(param),
                                };
                            default:
                                return {
                                    name: abiDef[j].name,
                                    value: param,
                                };
                        }
                    });
                    return {
                        commandName: routerCommands_1.CommandType[commandType],
                        commandType,
                        params,
                    };
                }
                else if (commandDef.parser === routerCommands_1.Parser.V3Actions) {
                    // TODO: implement better parsing here
                    return {
                        commandName: routerCommands_1.CommandType[commandType],
                        commandType,
                        params: inputs.map((input) => ({ name: 'command', value: input })),
                    };
                }
                else {
                    throw new Error(`Unsupported parser: ${commandDef}`);
                }
            }),
        };
    }
    // parse command types from bytes string
    static getCommands(commands) {
        const commandTypes = [];
        for (let i = 2; i < commands.length; i += 2) {
            const byte = commands.substring(i, i + 2);
            commandTypes.push(parseInt(byte, 16));
        }
        return commandTypes;
    }
}
exports.GenericCommandParser = GenericCommandParser;
function parseV3PathExactIn(path) {
    const strippedPath = path.replace('0x', '');
    let tokenIn = ethers_1.ethers.utils.getAddress(strippedPath.substring(0, 40));
    let loc = 40;
    const res = [];
    while (loc < strippedPath.length) {
        const feeAndTokenOut = strippedPath.substring(loc, loc + 46);
        const fee = parseInt(feeAndTokenOut.substring(0, 6), 16);
        const tokenOut = ethers_1.ethers.utils.getAddress(feeAndTokenOut.substring(6, 46));
        res.push({
            tokenIn,
            tokenOut,
            fee,
        });
        tokenIn = tokenOut;
        loc += 46;
    }
    return res;
}
exports.parseV3PathExactIn = parseV3PathExactIn;
function parseV3PathExactOut(path) {
    const strippedPath = path.replace('0x', '');
    let tokenIn = ethers_1.ethers.utils.getAddress(strippedPath.substring(strippedPath.length - 40));
    let loc = strippedPath.length - 86; // 86 = (20 addr + 3 fee + 20 addr) * 2 (for hex characters)
    const res = [];
    while (loc >= 0) {
        const feeAndTokenOut = strippedPath.substring(loc, loc + 46);
        const tokenOut = ethers_1.ethers.utils.getAddress(feeAndTokenOut.substring(0, 40));
        const fee = parseInt(feeAndTokenOut.substring(40, 46), 16);
        res.push({
            tokenIn,
            tokenOut,
            fee,
        });
        tokenIn = tokenOut;
        loc -= 46;
    }
    return res;
}
exports.parseV3PathExactOut = parseV3PathExactOut;
function v4RouterCallToParams(actions) {
    return actions.map((action) => {
        return {
            name: action.actionName,
            value: action.params.map((param) => {
                return {
                    name: param.name,
                    value: param.value,
                };
            }),
        };
    });
}
