// @uniswap/universal-router-sdk/dist/cjs/src/utils/eip712.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NONCE_SKIP_CHECK = exports.generateNonce = exports.getUniversalRouterDomain = exports.EXECUTE_SIGNED_TYPES = exports.EIP712_DOMAIN_VERSION = exports.EIP712_DOMAIN_NAME = void 0;
const ethers_1 = require("ethers");
exports.EIP712_DOMAIN_NAME = 'UniversalRouter';
exports.EIP712_DOMAIN_VERSION = '2';
exports.EXECUTE_SIGNED_TYPES = {
    ExecuteSigned: [
        { name: 'commands', type: 'bytes' },
        { name: 'inputs', type: 'bytes[]' },
        { name: 'intent', type: 'bytes32' },
        { name: 'data', type: 'bytes32' },
        { name: 'sender', type: 'address' },
        { name: 'nonce', type: 'bytes32' },
        { name: 'deadline', type: 'uint256' },
    ],
};
/**
 * Generate EIP712 domain for Universal Router
 */
function getUniversalRouterDomain(chainId, verifyingContract) {
    return {
        name: exports.EIP712_DOMAIN_NAME,
        version: exports.EIP712_DOMAIN_VERSION,
        chainId,
        verifyingContract,
    };
}
exports.getUniversalRouterDomain = getUniversalRouterDomain;
/**
 * Generate a random nonce for signed execution
 * Uses ethers.utils.randomBytes for secure randomness
 */
function generateNonce() {
    const randomBytes = ethers_1.ethers.utils.randomBytes(32);
    return ethers_1.ethers.utils.hexlify(randomBytes);
}
exports.generateNonce = generateNonce;
/**
 * Sentinel value to skip nonce checking (allows signature replay)
 */
exports.NONCE_SKIP_CHECK = '0x' + 'f'.repeat(64); // bytes32(type(uint256).max)
