// @uniswap/universal-router-sdk/dist/cjs/src/utils/routerTradeAdapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RouterTradeAdapter = exports.isNativeCurrency = exports.PoolType = void 0;
const router_sdk_1 = require("@uniswap/router-sdk");
const sdk_core_1 = require("@uniswap/sdk-core");
const v2_sdk_1 = require("@uniswap/v2-sdk");
const v3_sdk_1 = require("@uniswap/v3-sdk");
const v4_sdk_1 = require("@uniswap/v4-sdk");
const ethers_1 = require("ethers");
const constants_1 = require("./constants");
var PoolType;
(function (PoolType) {
    PoolType["V2Pool"] = "v2-pool";
    PoolType["V3Pool"] = "v3-pool";
    PoolType["V4Pool"] = "v4-pool";
})(PoolType = exports.PoolType || (exports.PoolType = {}));
const isNativeCurrency = (address) => address.toLowerCase() === constants_1.ETH_ADDRESS.toLowerCase() || address.toLowerCase() === constants_1.E_ETH_ADDRESS.toLowerCase();
exports.isNativeCurrency = isNativeCurrency;
// Helper class to convert routing-specific quote entities to RouterTrade entities
// the returned RouterTrade can then be used to build the UniswapTrade entity in this package
class RouterTradeAdapter {
    // Generate a RouterTrade using fields from a classic quote response
    static fromClassicQuote(quote) {
        const { route, tokenIn, tokenOut } = quote;
        if (!route)
            throw new Error('Expected route to be present');
        if (!route.length)
            throw new Error('Expected there to be at least one route');
        if (route.some((r) => !r.length))
            throw new Error('Expected all routes to have at least one pool');
        const firstRoute = route[0];
        const tokenInData = firstRoute[0].tokenIn;
        const tokenOutData = firstRoute[firstRoute.length - 1].tokenOut;
        if (!tokenInData || !tokenOutData)
            throw new Error('Expected both tokenIn and tokenOut to be present');
        if (tokenInData.chainId !== tokenOutData.chainId)
            throw new Error('Expected tokenIn and tokenOut to be have same chainId');
        const parsedCurrencyIn = RouterTradeAdapter.toCurrency((0, exports.isNativeCurrency)(tokenIn), tokenInData);
        const parsedCurrencyOut = RouterTradeAdapter.toCurrency((0, exports.isNativeCurrency)(tokenOut), tokenOutData);
        const typedRoutes = route.map((subRoute) => {
            const rawAmountIn = subRoute[0].amountIn;
            const rawAmountOut = subRoute[subRoute.length - 1].amountOut;
            if (!rawAmountIn || !rawAmountOut) {
                throw new Error('Expected both raw amountIn and raw amountOut to be present');
            }
            const inputAmount = sdk_core_1.CurrencyAmount.fromRawAmount(parsedCurrencyIn, rawAmountIn);
            const outputAmount = sdk_core_1.CurrencyAmount.fromRawAmount(parsedCurrencyOut, rawAmountOut);
            const isOnlyV2 = RouterTradeAdapter.isVersionedRoute(PoolType.V2Pool, subRoute);
            const isOnlyV3 = RouterTradeAdapter.isVersionedRoute(PoolType.V3Pool, subRoute);
            const isOnlyV4 = RouterTradeAdapter.isVersionedRoute(PoolType.V4Pool, subRoute);
            return {
                routev4: isOnlyV4
                    ? new v4_sdk_1.Route(subRoute.map(RouterTradeAdapter.toV4Pool), parsedCurrencyIn, parsedCurrencyOut)
                    : null,
                routev3: isOnlyV3
                    ? new v3_sdk_1.Route(subRoute.map(RouterTradeAdapter.toV3Pool), parsedCurrencyIn, parsedCurrencyOut)
                    : null,
                routev2: isOnlyV2
                    ? new v2_sdk_1.Route(subRoute.map(RouterTradeAdapter.toPair), parsedCurrencyIn, parsedCurrencyOut)
                    : null,
                mixedRoute: !isOnlyV4 && !isOnlyV3 && !isOnlyV2
                    ? new router_sdk_1.MixedRouteSDK(subRoute.map(RouterTradeAdapter.toPoolOrPair), parsedCurrencyIn, parsedCurrencyOut)
                    : null,
                inputAmount,
                outputAmount,
            };
        });
        return new router_sdk_1.Trade({
            v2Routes: typedRoutes
                .filter((route) => route.routev2)
                .map((route) => ({
                routev2: route.routev2,
                inputAmount: route.inputAmount,
                outputAmount: route.outputAmount,
            })),
            v3Routes: typedRoutes
                .filter((route) => route.routev3)
                .map((route) => ({
                routev3: route.routev3,
                inputAmount: route.inputAmount,
                outputAmount: route.outputAmount,
            })),
            v4Routes: typedRoutes
                .filter((route) => route.routev4)
                .map((route) => ({
                routev4: route.routev4,
                inputAmount: route.inputAmount,
                outputAmount: route.outputAmount,
            })),
            mixedRoutes: typedRoutes
                .filter((route) => route.mixedRoute)
                .map((route) => ({
                mixedRoute: route.mixedRoute,
                inputAmount: route.inputAmount,
                outputAmount: route.outputAmount,
            })),
            tradeType: quote.tradeType,
        });
    }
    static toCurrency(isNative, token) {
        if (isNative) {
            return sdk_core_1.Ether.onChain(token.chainId);
        }
        return this.toToken(token);
    }
    static toToken(token) {
        const { chainId, address, decimals, symbol, buyFeeBps, sellFeeBps } = token;
        return new sdk_core_1.Token(chainId, address, parseInt(decimals.toString()), symbol, 
        /* name */ undefined, false, buyFeeBps ? ethers_1.BigNumber.from(buyFeeBps) : undefined, sellFeeBps ? ethers_1.BigNumber.from(sellFeeBps) : undefined);
    }
    static toV3Pool({ fee, sqrtRatioX96, liquidity, tickCurrent, tokenIn, tokenOut }) {
        return new v3_sdk_1.Pool(RouterTradeAdapter.toToken(tokenIn), RouterTradeAdapter.toToken(tokenOut), parseInt(fee), sqrtRatioX96, liquidity, parseInt(tickCurrent));
    }
    static toV4Pool(pool) {
        const parsedCurrencyIn = RouterTradeAdapter.toCurrency((0, exports.isNativeCurrency)(pool.tokenIn.address), pool.tokenIn);
        const parsedCurrencyOut = RouterTradeAdapter.toCurrency((0, exports.isNativeCurrency)(pool.tokenOut.address), pool.tokenOut);
        return new v4_sdk_1.Pool(parsedCurrencyIn, parsedCurrencyOut, parseInt(pool.fee), parseInt(pool.tickSpacing), pool.hooks, pool.sqrtRatioX96, pool.liquidity, parseInt(pool.tickCurrent));
    }
    static isVersionedRoute(type, route) {
        return route.every((pool) => pool.type === type);
    }
}
exports.RouterTradeAdapter = RouterTradeAdapter;
RouterTradeAdapter.toPoolOrPair = (pool) => {
    switch (pool.type) {
        case PoolType.V4Pool:
            return RouterTradeAdapter.toV4Pool(pool);
        case PoolType.V3Pool:
            return RouterTradeAdapter.toV3Pool(pool);
        case PoolType.V2Pool:
            return RouterTradeAdapter.toPair(pool);
        default:
            throw new Error('Invalid pool type');
    }
};
RouterTradeAdapter.toPair = ({ reserve0, reserve1 }) => {
    return new v2_sdk_1.Pair(sdk_core_1.CurrencyAmount.fromRawAmount(RouterTradeAdapter.toToken(reserve0.token), reserve0.quotient), sdk_core_1.CurrencyAmount.fromRawAmount(RouterTradeAdapter.toToken(reserve1.token), reserve1.quotient));
};
