// @uniswap/smart-order-router/build/main/providers/caching/route/model/cached-route.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CachedRoute = void 0;
const router_sdk_1 = require("@uniswap/router-sdk");
const v2_sdk_1 = require("@uniswap/v2-sdk");
const v3_sdk_1 = require("@uniswap/v3-sdk");
const v4_sdk_1 = require("@uniswap/v4-sdk");
const util_1 = require("../../../../util");
/**
 * Class defining the route to cache
 *
 * @export
 * @class CachedRoute
 */
class CachedRoute {
    /**
     * @param route
     * @param percent
     */
    constructor({ route, percent }) {
        // Hashing function copying the same implementation as Java's `hashCode`
        // Sourced from: https://gist.github.com/hyamamoto/fd435505d29ebfa3d9716fd2be8d42f0?permalink_comment_id=4613539#gistcomment-4613539
        this.hashCode = (str) => [...str].reduce((s, c) => (Math.imul(31, s) + c.charCodeAt(0)) | 0, 0);
        this.route = route;
        this.percent = percent;
    }
    get protocol() {
        return this.route.protocol;
    }
    get currencyIn() {
        return this.route.input;
    }
    get currencyOut() {
        return this.route.output;
    }
    get routePath() {
        switch (this.protocol) {
            case router_sdk_1.Protocol.V4:
                return this.route.pools
                    .map((pool) => `[V4]${(0, util_1.getAddress)(pool.token0)}/${(0, util_1.getAddress)(pool.token1)}/${pool.fee}/${pool.hooks}/${pool.tickSpacing}`)
                    .join('->');
            case router_sdk_1.Protocol.V3:
                return this.route.pools
                    .map((pool) => `[V3]${pool.token0.address}/${pool.token1.address}/${pool.fee}`)
                    .join('->');
            case router_sdk_1.Protocol.V2:
                return this.route.pairs
                    .map((pair) => `[V2]${pair.token0.address}/${pair.token1.address}`)
                    .join('->');
            case router_sdk_1.Protocol.MIXED:
                return this.route.pools
                    .map((pool) => {
                    if (pool instanceof v4_sdk_1.Pool) {
                        // TODO: ROUTE-217 - Support native currency routing in V4
                        return `[V4]${(0, util_1.getAddress)(pool.token0)}/${(0, util_1.getAddress)(pool.token1)}/${pool.fee}/${pool.hooks}/${pool.tickSpacing}`;
                    }
                    else if (pool instanceof v3_sdk_1.Pool) {
                        return `[V3]${pool.token0.address}/${pool.token1.address}/${pool.fee}`;
                    }
                    else if (pool instanceof v2_sdk_1.Pair) {
                        return `[V2]${pool.token0.address}/${pool.token1.address}`;
                    }
                    else {
                        throw new Error(`Unsupported pool type ${JSON.stringify(pool)}`);
                    }
                })
                    .join('->');
            default:
                throw new Error(`Unsupported protocol ${this.protocol}`);
        }
    }
    get routeId() {
        return this.hashCode(this.routePath);
    }
}
exports.CachedRoute = CachedRoute;
