// @0glabs/0g-serving-broker/cli.commonjs/sdk/inference/broker/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZGServingUserBrokerBase = exports.SessionMode = exports.EPHEMERAL_TOKEN_MAX_DURATION = exports.EPHEMERAL_TOKEN_ID = void 0;
const tslib_1 = require("tslib");
const extractor_1 = require("../extractor");
const utils_1 = require("../../common/utils");
const fs = tslib_1.__importStar(require("fs/promises"));
const storage_1 = require("../../common/storage");
const ethers_1 = require("ethers");
const logger_1 = require("../../common/logger");
const textToImage_1 = require("../extractor/textToImage");
const speech_to_text_1 = require("../extractor/speech-to-text");
const imageEditing_1 = require("../extractor/imageEditing");
/**
 * Special token ID reserved for ephemeral tokens.
 * Ephemeral tokens (tokenId=255) are not checked against the revoked bitmap,
 * only generation check applies. This allows unlimited ephemeral tokens without
 * consuming the 0-254 tokenId quota.
 */
exports.EPHEMERAL_TOKEN_ID = 255;
/**
 * Maximum duration for ephemeral tokens (24 hours in milliseconds).
 * Ephemeral tokens must have an expiration time and cannot exceed this duration.
 */
exports.EPHEMERAL_TOKEN_MAX_DURATION = 24 * 60 * 60 * 1000; // 24 hours
/**
 * Session mode for token generation
 */
var SessionMode;
(function (SessionMode) {
    /** Ephemeral token: uses tokenId=255, not individually revocable, no quota consumption */
    SessionMode["Ephemeral"] = "ephemeral";
    /** Persistent token: uses tokenId 0-254, individually revocable, consumes quota */
    SessionMode["Persistent"] = "persistent";
})(SessionMode || (exports.SessionMode = SessionMode = {}));
class ZGServingUserBrokerBase {
    contract;
    metadata;
    cache;
    // Minimum locked balance required by provider broker proxy (1 0G in neuron).
    // Matches MinimumLockedBalance in api/inference/const/const.go.
    // Provider requires: lockBalance >= unsettledFee + currentFee + MIN_LOCKED_BALANCE
    static MIN_LOCKED_BALANCE = BigInt(1) * BigInt(10 ** 18);
    ledger;
    constructor(contract, ledger, metadata, cache) {
        this.contract = contract;
        this.ledger = ledger;
        this.metadata = metadata;
        this.cache = cache;
    }
    async getService(providerAddress, useCache = true) {
        const key = storage_1.CacheKeyHelpers.getServiceKey(providerAddress);
        const cachedSvc = await this.cache.getItem(key);
        if (cachedSvc && useCache) {
            return cachedSvc;
        }
        try {
            const svc = await this.contract.getService(providerAddress);
            logger_1.logger.debug('Fetched service info from contract:', svc);
            await this.cache.setItem(key, svc, 10 * 60 * 1000, storage_1.CacheValueTypeEnum.Service);
            return svc;
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    async getQuote(providerAddress) {
        try {
            const service = await this.getService(providerAddress);
            const url = service.url;
            const endpoint = `${url}/v1/quote`;
            const rawReport = await this.fetchText(endpoint, {
                method: 'GET',
            });
            return {
                rawReport,
                signingAddress: '',
            };
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    async downloadQuoteReport(providerAddress, outputPath) {
        try {
            const service = await this.getService(providerAddress);
            const url = service.url;
            const endpoint = `${url}/v1/quote`;
            const quoteString = await this.fetchText(endpoint, {
                method: 'GET',
            });
            await fs.writeFile(outputPath, quoteString);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    async userAcknowledged(providerAddress) {
        const userAddress = this.contract.getUserAddress();
        const key = storage_1.CacheKeyHelpers.getUserAckKey(userAddress, providerAddress);
        const cachedSvc = await this.cache.getItem(key);
        if (cachedSvc) {
            return true;
        }
        try {
            const account = await this.contract.getAccount(providerAddress);
            if (account.acknowledged) {
                await this.cache.setItem(key, '', 10 * 60 * 1000, storage_1.CacheValueTypeEnum.Other);
                return true;
            }
            else {
                return false;
            }
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    async fetchText(endpoint, options) {
        try {
            const response = await fetch(endpoint, options);
            if (!response.ok) {
                throw new Error(`HTTP error! Status: ${response.status}`);
            }
            const buffer = await response.arrayBuffer();
            return Buffer.from(buffer).toString('utf-8');
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    async getExtractor(providerAddress, useCache = true) {
        try {
            const svc = await this.getService(providerAddress, useCache);
            const extractor = this.createExtractor(svc);
            return extractor;
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    createExtractor(svc) {
        switch (svc.serviceType) {
            case 'chatbot':
                return new extractor_1.ChatBot(svc);
            case 'text-to-image':
                return new textToImage_1.TextToImage(svc);
            case 'image-editing':
                return new imageEditing_1.ImageEditing(svc);
            case 'speech-to-text':
                return new speech_to_text_1.SpeechToText(svc);
            default:
                throw new Error('Unknown service type');
        }
    }
    a0giToNeuron(value) {
        const valueStr = value.toFixed(18);
        const parts = valueStr.split('.');
        // Handle integer part
        const integerPart = parts[0];
        let integerPartAsBigInt = BigInt(integerPart) * BigInt(10 ** 18);
        // Handle fractional part if it exists
        if (parts.length > 1) {
            let fractionalPart = parts[1];
            while (fractionalPart.length < 18) {
                fractionalPart += '0';
            }
            if (fractionalPart.length > 18) {
                fractionalPart = fractionalPart.slice(0, 18); // Truncate to avoid overflow
            }
            const fractionalPartAsBigInt = BigInt(fractionalPart);
            integerPartAsBigInt += fractionalPartAsBigInt;
        }
        return integerPartAsBigInt;
    }
    neuronToA0gi(value) {
        const divisor = BigInt(10 ** 18);
        const integerPart = value / divisor;
        const remainder = value % divisor;
        const decimalPart = Number(remainder) / Number(divisor);
        return Number(integerPart) + decimalPart;
    }
    generateNonce() {
        if (typeof window !== 'undefined' && window.crypto) {
            // Browser environment - use Web Crypto API
            const array = new Uint8Array(16);
            window.crypto.getRandomValues(array);
            return Array.from(array, (byte) => byte.toString(16).padStart(2, '0')).join('');
        }
        else {
            // Node.js or other environment - use timestamp-based nonce
            const timestamp = Date.now();
            const random = Math.random();
            const randomStr = random.toString(36).substring(2, 15);
            return `${timestamp}-${randomStr}`.padEnd(32, '0');
        }
    }
    /**
     * Get account info from cache or contract.
     * @param providerAddress - The provider address
     */
    async getAccountInfo(providerAddress) {
        const userAddress = this.contract.getUserAddress();
        const cacheKey = `account_info_${userAddress}_${providerAddress}`;
        // Try cache first
        const cached = (await this.cache.getItem(cacheKey));
        if (cached) {
            return {
                generation: cached.generation,
                revokedBitmap: BigInt(cached.revokedBitmap),
            };
        }
        // Fetch from contract
        try {
            const account = await this.contract.getAccount(providerAddress);
            // Handle case where account exists but fields don't exist (pre-upgrade accounts)
            const info = {
                generation: account.generation != null ? Number(account.generation) : 0,
                revokedBitmap: account.revokedBitmap ?? BigInt(0),
            };
            // Cache for 5 minutes
            await this.cache.setItem(cacheKey, {
                generation: info.generation,
                revokedBitmap: info.revokedBitmap.toString(),
            }, 5 * 60 * 1000, storage_1.CacheValueTypeEnum.Other);
            return info;
        }
        catch {
            // Account may not exist yet
            return {
                generation: 0,
                revokedBitmap: BigInt(0),
            };
        }
    }
    /**
     * Generate a new session token with generation and tokenId for revocation support
     * @param providerAddress - The provider address
     * @param options - Optional configuration for token generation
     * @returns The cached session with token, signature, and raw message
     */
    async generateSessionToken(providerAddress, options) {
        const userAddress = this.contract.getUserAddress();
        const timestamp = Date.now();
        const mode = options?.mode ?? SessionMode.Ephemeral;
        const nonce = this.generateNonce();
        // Determine duration and expiresAt based on mode
        let duration;
        let expiresAt;
        if (mode === SessionMode.Ephemeral) {
            // Ephemeral tokens MUST have an expiration time and cannot exceed 24 hours
            duration = options?.duration ?? exports.EPHEMERAL_TOKEN_MAX_DURATION;
            if (duration <= 0) {
                // Force ephemeral tokens to have expiration
                duration = exports.EPHEMERAL_TOKEN_MAX_DURATION;
            }
            if (duration > exports.EPHEMERAL_TOKEN_MAX_DURATION) {
                throw new Error(`Ephemeral token duration cannot exceed 24 hours (${exports.EPHEMERAL_TOKEN_MAX_DURATION}ms)`);
            }
            expiresAt = timestamp + duration;
        }
        else {
            // Persistent tokens can have any duration, including never expires (0)
            duration = options?.duration ?? 0;
            expiresAt = duration > 0 ? timestamp + duration : 0;
        }
        // Determine tokenId based on mode
        let tokenId;
        let generation;
        if (mode === SessionMode.Ephemeral) {
            // Ephemeral tokens always use tokenId=255
            tokenId = exports.EPHEMERAL_TOKEN_ID;
            const accountInfo = await this.getAccountInfo(providerAddress);
            generation = accountInfo.generation;
        }
        else {
            // Persistent tokens: use provided tokenId or find available one from bitmap
            const accountInfo = await this.getAccountInfo(providerAddress);
            generation = accountInfo.generation;
            if (options?.tokenId !== undefined) {
                // Use the specified tokenId
                tokenId = options.tokenId;
                if (tokenId < 0 || tokenId >= exports.EPHEMERAL_TOKEN_ID) {
                    throw new Error(`Invalid tokenId: ${tokenId}. Must be between 0 and ${exports.EPHEMERAL_TOKEN_ID - 1}`);
                }
                // Check if this tokenId is already revoked
                const bit = BigInt(1) << BigInt(tokenId);
                if ((accountInfo.revokedBitmap & bit) !== BigInt(0)) {
                    throw new Error(`TokenId ${tokenId} is already revoked. Use a different tokenId or call revokeAllTokens() to reset.`);
                }
            }
            else {
                // Find available tokenId from bitmap (only checks revoked, not occupied)
                // Note: This may return a tokenId that's already in use but not revoked yet.
                // UI layer should track occupied tokenIds and provide a specific tokenId.
                tokenId = this.findAvailableTokenId(accountInfo.revokedBitmap);
            }
        }
        const token = {
            address: userAddress,
            provider: providerAddress,
            timestamp,
            expiresAt,
            nonce,
            generation,
            tokenId,
        };
        // Create message to be signed
        const message = JSON.stringify(token);
        // Create hash using the same method as signRequest in encrypt.ts
        const messageHash = (0, ethers_1.keccak256)((0, ethers_1.toUtf8Bytes)(message));
        // Sign using the same pattern as signRequest: signMessage with toBeArray
        const signature = await this.contract.signer.signMessage(Buffer.from(messageHash.slice(2), 'hex'));
        const session = {
            token,
            signature,
            rawMessage: message,
        };
        // Only cache ephemeral sessions
        if (mode === SessionMode.Ephemeral) {
            const cacheKey = storage_1.CacheKeyHelpers.getSessionTokenKey(userAddress, providerAddress);
            await this.cache.setItem(cacheKey, session, duration, storage_1.CacheValueTypeEnum.Session);
        }
        return session;
    }
    /**
     * Find the smallest available tokenId from the revoked bitmap.
     * @param revokedBitmap - The bitmap of revoked tokenIds
     * @returns The smallest available tokenId (0-254)
     */
    findAvailableTokenId(revokedBitmap) {
        // Find the smallest available tokenId (0-254)
        // tokenId 255 is reserved for ephemeral tokens
        for (let tokenId = 0; tokenId < exports.EPHEMERAL_TOKEN_ID; tokenId++) {
            const bit = BigInt(1) << BigInt(tokenId);
            if ((revokedBitmap & bit) === BigInt(0)) {
                // This tokenId is not revoked, it's available
                return tokenId;
            }
        }
        // All 255 tokenIds are revoked
        throw new Error('API Key limit reached (255). Call revokeAllTokens() to reset.');
    }
    /**
     * Get or create an ephemeral session token for the provider.
     * Ephemeral tokens use tokenId=255 and don't consume the API key quota.
     * @param providerAddress - The provider address
     * @returns The cached or newly generated session
     */
    async getOrCreateSession(providerAddress) {
        const userAddress = this.contract.getUserAddress();
        const cacheKey = storage_1.CacheKeyHelpers.getSessionTokenKey(userAddress, providerAddress);
        const cached = (await this.cache.getItem(cacheKey));
        if (cached) {
            // Ephemeral tokens always have expiration time
            // Check if token has enough time remaining (at least 1 hour)
            const hasTimeRemaining = cached.token.expiresAt > Date.now() + 60 * 60 * 1000;
            if (hasTimeRemaining) {
                return cached;
            }
        }
        // Generate new ephemeral session
        return await this.generateSessionToken(providerAddress, {
            mode: SessionMode.Ephemeral,
        });
    }
    /**
     * Get request headers with an ephemeral session token.
     * This is the default method for SDK usage - it uses ephemeral tokens
     * that don't consume the API key quota.
     * @param providerAddress - The provider address
     * @returns Headers with Authorization
     */
    async getHeader(providerAddress) {
        // Check if provider is acknowledged - this is still necessary
        if (!(await this.userAcknowledged(providerAddress))) {
            throw new Error('Provider signer is not acknowledged');
        }
        // Get or create ephemeral session token
        const session = await this.getOrCreateSession(providerAddress);
        return {
            Authorization: `Bearer app-sk-${Buffer.from(session.rawMessage + '|' + session.signature).toString('base64')}`,
        };
    }
    // ==================== API Key Management ====================
    /**
     * Create a new API Key (persistent token).
     * API Keys consume tokenId quota (0-254) and can be individually revoked.
     * The tokenId is determined by finding the smallest available ID from the contract's bitmap.
     * @param providerAddress - The provider address
     * @param options - Optional configuration
     * @returns The API key information including the raw token
     */
    async createApiKey(providerAddress, options) {
        const session = await this.generateSessionToken(providerAddress, {
            mode: SessionMode.Persistent,
            duration: options?.expiresIn ?? 0, // Default: never expires
            tokenId: options?.tokenId,
        });
        const rawToken = `app-sk-${Buffer.from(session.rawMessage + '|' + session.signature).toString('base64')}`;
        return {
            tokenId: session.token.tokenId,
            createdAt: session.token.timestamp,
            expiresAt: session.token.expiresAt,
            rawToken,
        };
    }
    /**
     * Revoke an API Key by its tokenId.
     * This calls the contract to revoke the token.
     * @param providerAddress - The provider address
     * @param tokenId - The token ID to revoke (0-254)
     * @param gasPrice - Optional gas price
     */
    async revokeApiKey(providerAddress, tokenId, gasPrice) {
        if (tokenId === exports.EPHEMERAL_TOKEN_ID) {
            throw new Error('Cannot revoke ephemeral token individually. Use revokeAllTokens() instead.');
        }
        // Revoke on contract
        await this.contract.revokeToken(providerAddress, tokenId, gasPrice);
    }
    /**
     * Revoke all tokens (both ephemeral and persistent).
     * This increments the generation, invalidating all existing tokens.
     * @param providerAddress - The provider address
     * @param gasPrice - Optional gas price
     */
    async revokeAllTokens(providerAddress, gasPrice) {
        // Revoke on contract
        await this.contract.revokeAllTokens(providerAddress, gasPrice);
        // Clear ephemeral session cache
        await this.clearEphemeralSession(providerAddress);
        // Also clear account info cache to ensure fresh generation number is fetched
        // When generation increments, cached account info with old generation becomes stale
        const userAddress = this.contract.getUserAddress();
        const accountInfoKey = `account_info_${userAddress}_${providerAddress}`;
        this.cache.setItem(accountInfoKey, null, 1, storage_1.CacheValueTypeEnum.Other);
    }
    /**
     * Clear ephemeral session cache
     */
    async clearEphemeralSession(providerAddress) {
        const userAddress = this.contract.getUserAddress();
        const cacheKey = storage_1.CacheKeyHelpers.getSessionTokenKey(userAddress, providerAddress);
        // Remove by setting to null with short TTL
        await this.cache.setItem(cacheKey, null, 1, storage_1.CacheValueTypeEnum.Other);
    }
    async calculateFee(extractor, content) {
        const svc = await extractor.getSvcInfo();
        const outputCount = await extractor.getOutputCount(content);
        const inputCount = await extractor.getInputCount(content);
        return (BigInt(outputCount) * BigInt(svc.outputPrice) +
            BigInt(inputCount) * BigInt(svc.inputPrice));
    }
    // Long TTL for fee caches — these track unsettled fees across the lifetime
    // of a long-running process (e.g. inference server). They should only be
    // cleared explicitly after a successful transfer, not by expiration.
    static FEE_CACHE_TTL = 24 * 60 * 60 * 1000; // 24 hours
    /**
     * Accumulate fee into the cachedFee counter.
     * Called by processResponse() to track usage between auto-funding cycles.
     * This value is used as a fallback by fetchUnsettledFee() when the provider
     * does not support the /unsettledfee API endpoint.
     */
    async updateCachedFee(provider, fee) {
        try {
            const cacheFundKey = storage_1.CacheKeyHelpers.getCachedFeeKey(provider);
            const curFee = (await this.cache.getItem(cacheFundKey)) || BigInt(0);
            await this.cache.setItem(cacheFundKey, BigInt(curFee) + fee, ZGServingUserBrokerBase.FEE_CACHE_TTL, storage_1.CacheValueTypeEnum.BigInt);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    async clearCacheFee(provider) {
        try {
            const key = storage_1.CacheKeyHelpers.getCachedFeeKey(provider);
            await this.cache.setItem(key, BigInt(0), ZGServingUserBrokerBase.FEE_CACHE_TTL, storage_1.CacheValueTypeEnum.BigInt);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    // ==================== Background Auto-Funding ====================
    autoFundingTimers = new Map();
    /**
     * Start background auto-funding for a provider.
     *
     * Runs an immediate balance check, then periodically checks the provider
     * sub-account balance and tops up if needed. This is completely decoupled
     * from the request path, so getRequestHeaders() has zero extra latency.
     *
     * @param provider - The provider address to auto-fund.
     * @param config - Optional auto-funding configuration.
     * @param config.interval - Polling interval in ms (default: 30000 = 30s).
     * @param config.bufferMultiplier - Multiplier for MIN_LOCKED_BALANCE buffer (default: 2).
     *   requiredBalance = unsettledFee + bufferMultiplier * MIN_LOCKED_BALANCE
     * @param gasPrice - Optional gas price for transactions.
     */
    async startAutoFunding(provider, config, gasPrice) {
        // Stop existing timer for this provider if any
        this.stopAutoFunding(provider);
        const interval = config?.interval ?? 30_000;
        const bufferMultiplier = config?.bufferMultiplier ?? 2;
        // Run immediately on start
        await this.checkAndFund(provider, bufferMultiplier, gasPrice);
        // Then run on interval
        const timer = setInterval(async () => {
            await this.checkAndFund(provider, bufferMultiplier, gasPrice);
        }, interval);
        this.autoFundingTimers.set(provider, timer);
        logger_1.logger.debug(`[Auto-funding] Started for provider ${provider} ` +
            `(interval=${interval}ms, bufferMultiplier=${bufferMultiplier})`);
    }
    /**
     * Stop background auto-funding for a provider.
     *
     * @param provider - The provider address. If omitted, stops all auto-funding timers.
     */
    stopAutoFunding(provider) {
        if (provider) {
            const timer = this.autoFundingTimers.get(provider);
            if (timer) {
                clearInterval(timer);
                this.autoFundingTimers.delete(provider);
                logger_1.logger.debug(`[Auto-funding] Stopped for provider ${provider}`);
            }
        }
        else {
            // Stop all
            for (const [addr, timer] of this.autoFundingTimers) {
                clearInterval(timer);
                logger_1.logger.debug(`[Auto-funding] Stopped for provider ${addr}`);
            }
            this.autoFundingTimers.clear();
        }
    }
    /**
     * Check if a background auto-funding timer is active for the given provider.
     */
    hasAutoFunding(provider) {
        return this.autoFundingTimers.has(provider);
    }
    /**
     * Single check-and-fund cycle. Queries the provider for the real unsettled
     * fee, computes the required balance with buffer, and transfers the deficit.
     *
     * Called by the background timer, and also inline from getRequestHeaders()
     * when no background timer is active (to preserve backward compatibility).
     */
    async checkAndFund(provider, bufferMultiplier, gasPrice) {
        try {
            // Skip auto-funding in browser environments.
            // Browser signers (e.g. MetaMask) require user confirmation for each
            // transaction, so silent auto-funding would cause unexpected wallet popups.
            if (typeof window !== 'undefined' &&
                typeof window.document !== 'undefined') {
                return;
            }
            const minLocked = ZGServingUserBrokerBase.MIN_LOCKED_BALANCE;
            const unsettledFee = await this.fetchUnsettledFee(provider);
            const requiredBalance = unsettledFee + BigInt(bufferMultiplier) * minLocked;
            logger_1.logger.debug(`[Auto-funding] Check: unsettledFee=${this.neuronToA0gi(unsettledFee).toFixed(6)} 0G, ` +
                `requiredBalance=${this.neuronToA0gi(requiredBalance).toFixed(6)} 0G`);
            const deficit = await this.getTransferDeficit(provider, requiredBalance);
            if (deficit > BigInt(0)) {
                await this.doTransfer(provider, deficit, gasPrice);
            }
        }
        catch (error) {
            logger_1.logger.warn(`[Auto-funding] Check-and-fund failed: ${error?.message || error}`);
        }
    }
    /**
     * Fetch unsettled fee from the provider's API.
     * Requires a valid session token (Authorization header).
     * Falls back to cached fee if the provider doesn't support this endpoint.
     */
    async fetchUnsettledFee(provider) {
        try {
            const svc = await this.getService(provider);
            const headers = await this.getHeader(provider);
            const userAddress = this.contract.getUserAddress();
            const response = await fetch(`${svc.url}/v1/user/${userAddress}/unsettledfee`, {
                method: 'GET',
                headers: {
                    ...headers,
                },
            });
            if (!response.ok) {
                logger_1.logger.debug(`[Auto-funding] Provider does not support unsettled fee query (${response.status}), using cachedFee fallback`);
                return await this.getCachedFee(provider);
            }
            const data = await response.json();
            const fee = BigInt(data.unsettledFee || '0');
            logger_1.logger.debug(`[Auto-funding] Provider unsettled fee: ${this.neuronToA0gi(fee).toFixed(6)} 0G`);
            // Snapshot the real unsettled fee into cachedFee as fallback
            const cacheFundKey = storage_1.CacheKeyHelpers.getCachedFeeKey(provider);
            await this.cache.setItem(cacheFundKey, fee, ZGServingUserBrokerBase.FEE_CACHE_TTL, storage_1.CacheValueTypeEnum.BigInt);
            return fee;
        }
        catch (error) {
            logger_1.logger.debug(`[Auto-funding] Failed to fetch unsettled fee: ${error?.message || error}, using cachedFee fallback`);
            return await this.getCachedFee(provider);
        }
    }
    async getCachedFee(provider) {
        try {
            const key = storage_1.CacheKeyHelpers.getCachedFeeKey(provider);
            return BigInt((await this.cache.getItem(key)) || BigInt(0));
        }
        catch {
            return BigInt(0);
        }
    }
    /**
     * Calculate how much additional fund is needed to meet the required balance.
     * Returns 0 if the current balance is already sufficient.
     * Returns requiredBalance if the account doesn't exist yet.
     */
    async getTransferDeficit(provider, requiredBalance) {
        try {
            const acc = await this.contract.getAccount(provider);
            const lockedFund = acc.balance - acc.pendingRefund;
            logger_1.logger.debug(`Locked fund for provider ${provider}: ${lockedFund.toString()}, required: ${requiredBalance.toString()}`);
            if (lockedFund >= requiredBalance) {
                return BigInt(0);
            }
            return requiredBalance - lockedFund;
        }
        catch {
            // Account doesn't exist, need to create it by transferring full amount
            return requiredBalance;
        }
    }
    async doTransfer(provider, amount, gasPrice) {
        // Ensure minimum transfer of MIN_LOCKED_BALANCE to avoid frequent
        // tiny transfers that trigger warnings and get rejected by provider.
        const minTransfer = ZGServingUserBrokerBase.MIN_LOCKED_BALANCE;
        const transferAmount = amount < minTransfer ? minTransfer : amount;
        try {
            await this.ledger.transferFund(provider, 'inference', transferAmount, gasPrice);
            // Clear fee cache after successful transfer
            await this.clearCacheFee(provider);
        }
        catch (error) {
            const amountInOG = this.neuronToA0gi(transferAmount);
            const errorMessage = error?.message?.toLowerCase() || '';
            if (errorMessage.includes('insufficient')) {
                logger_1.logger.warn(`[Auto-funding] Requires ${amountInOG.toFixed(4)} 0G in your ledger to transfer to the ` +
                    `provider sub-account, but your ledger available balance is insufficient. ` +
                    `Please deposit more funds:\n` +
                    `  SDK: broker.ledger.depositFund(${Math.ceil(amountInOG)})\n` +
                    `  CLI: 0g-compute-cli deposit --amount ${Math.ceil(amountInOG)}`);
                return;
            }
            logger_1.logger.warn(`[Auto-funding] Failed to transfer ${amountInOG.toFixed(4)} 0G to provider sub-account: ${error?.message || error}`);
        }
    }
}
exports.ZGServingUserBrokerBase = ZGServingUserBrokerBase;
