// @uniswap/smart-order-router/build/module/routers/alpha-router/functions/compute-all-routes.js
import { ADDRESS_ZERO } from '@uniswap/router-sdk';
import { Pair } from '@uniswap/v2-sdk';
import { Pool as V3Pool } from '@uniswap/v3-sdk';
import { Pool as V4Pool } from '@uniswap/v4-sdk';
import { getAddressLowerCase, nativeOnChain, V4_ETH_WETH_FAKE_POOL, } from '../../../util';
import { HooksOptions } from '../../../util/hooksOptions';
import { log } from '../../../util/log';
import { poolToString, routeToString } from '../../../util/routes';
import { MixedRoute, V2Route, V3Route, V4Route, } from '../../router';
export function computeAllV4Routes(currencyIn, currencyOut, pools, maxHops, hooksOptions) {
    let filteredPools = pools;
    if (hooksOptions === HooksOptions.HOOKS_ONLY) {
        filteredPools = pools.filter((pool) => pool.hooks !== ADDRESS_ZERO);
    }
    if (hooksOptions === HooksOptions.NO_HOOKS) {
        filteredPools = pools.filter((pool) => pool.hooks === ADDRESS_ZERO);
    }
    return computeAllRoutes(currencyIn, currencyOut, (route, currencyIn, currencyOut) => {
        return new V4Route(route, currencyIn, currencyOut);
    }, (pool, currency) => pool.involvesToken(currency), filteredPools, maxHops);
}
export function computeAllV3Routes(tokenIn, tokenOut, pools, maxHops) {
    return computeAllRoutes(tokenIn, tokenOut, (route, tokenIn, tokenOut) => {
        return new V3Route(route, tokenIn, tokenOut);
    }, (pool, token) => pool.involvesToken(token), pools, maxHops);
}
export function computeAllV2Routes(tokenIn, tokenOut, pools, maxHops) {
    return computeAllRoutes(tokenIn, tokenOut, (route, tokenIn, tokenOut) => {
        return new V2Route(route, tokenIn, tokenOut);
    }, (pool, token) => pool.involvesToken(token), pools, maxHops);
}
export function computeAllMixedRoutes(currencyIn, currencyOut, parts, maxHops, shouldEnableMixedRouteEthWeth, hooksOptions) {
    // first we need to filter non v4-pools
    const filteredPools = !hooksOptions || hooksOptions === HooksOptions.HOOKS_INCLUSIVE
        ? parts
        : parts.filter((pool) => !(pool instanceof V4Pool));
    if (hooksOptions === HooksOptions.HOOKS_ONLY) {
        // we need to filter out v4-pools with hooks
        // then concat the v4-pools with hooks
        const v4HookslessPools = parts.filter((pool) => pool instanceof V4Pool && pool.hooks !== ADDRESS_ZERO);
        parts = filteredPools.concat(v4HookslessPools);
    }
    if (hooksOptions === HooksOptions.NO_HOOKS) {
        // we need to filter out v4-pools without hooks
        // then concat the v4-pools without hooks
        const v4HookfulPools = parts.filter((pool) => pool instanceof V4Pool && pool.hooks === ADDRESS_ZERO);
        parts = filteredPools.concat(v4HookfulPools);
    }
    // only add fake v4 pool, if we see there's a native v4 pool in the candidate pool
    const containsV4NativePools = parts.filter((pool) => pool instanceof V4Pool &&
        pool.v4InvolvesToken(nativeOnChain(currencyIn.chainId))).length > 0;
    const amendedPools = containsV4NativePools && shouldEnableMixedRouteEthWeth
        ? parts.concat(V4_ETH_WETH_FAKE_POOL[currencyIn.chainId])
        : parts;
    // NOTE: we added a fake v4 pool, in order for mixed route to connect the v3 weth pool with v4 eth pool
    const routesRaw = computeAllRoutes(currencyIn, currencyOut, (route, currencyIn, currencyOut) => {
        // we only retake the fake v4 pool if the route contains a native v4 pool
        return new MixedRoute(route, currencyIn, currencyOut, containsV4NativePools);
    }, (pool, currency) => currency.isNative
        ? pool.involvesToken(currency)
        : pool.involvesToken(currency), amendedPools, maxHops);
    /// filter out pure v4 and v3 and v2 routes
    return routesRaw.filter((route) => {
        return (!route.pools.every((pool) => pool instanceof V4Pool) &&
            !route.pools.every((pool) => pool instanceof V3Pool) &&
            !route.pools.every((pool) => pool instanceof Pair));
    });
}
export function computeAllRoutes(tokenIn, tokenOut, buildRoute, involvesToken, pools, maxHops) {
    var _a;
    const poolsUsed = Array(pools.length).fill(false);
    const routes = [];
    const computeRoutes = (tokenIn, tokenOut, currentRoute, poolsUsed, tokensVisited, _previousTokenOut) => {
        const currentRouteContainsFakeV4Pool = currentRoute.filter((pool) => pool instanceof V4Pool &&
            pool.tickSpacing ===
                V4_ETH_WETH_FAKE_POOL[tokenIn.chainId].tickSpacing).length > 0;
        const amendedMaxHops = currentRouteContainsFakeV4Pool
            ? maxHops + 1
            : maxHops;
        // amendedMaxHops is the maxHops + 1 if the current route contains a fake v4 pool
        // b/c we want to allow the route to go through the fake v4 pool
        // also gas wise, if a route goes through the fake v4 pool, mixed quoter will add the wrap/unwrap gas cost:
        // https://github.com/Uniswap/mixed-quoter/pull/41/files#diff-a4d1289f264d1da22aac20cc55a9d01c8ba9cccd76ce1af8f952ec9034e7e1aaR189
        // and SOR will use the gas cost from the mixed quoter:
        // https://github.com/Uniswap/smart-order-router/blob/17da523f1af050e6430afb866d96681346c8fb8b/src/routers/alpha-router/gas-models/mixedRoute/mixed-route-heuristic-gas-model.ts#L222
        if (currentRoute.length > amendedMaxHops) {
            return;
        }
        if (currentRoute.length > 0 &&
            involvesToken(currentRoute[currentRoute.length - 1], tokenOut)) {
            routes.push(buildRoute([...currentRoute], tokenIn, tokenOut));
            return;
        }
        for (let i = 0; i < pools.length; i++) {
            if (poolsUsed[i]) {
                continue;
            }
            const curPool = pools[i];
            const previousTokenOut = _previousTokenOut ? _previousTokenOut : tokenIn;
            if (!involvesToken(curPool, previousTokenOut)) {
                continue;
            }
            const currentTokenOut = curPool.token0.equals(previousTokenOut)
                ? curPool.token1
                : curPool.token0;
            if (tokensVisited.has(getAddressLowerCase(currentTokenOut))) {
                continue;
            }
            tokensVisited.add(getAddressLowerCase(currentTokenOut));
            currentRoute.push(curPool);
            poolsUsed[i] = true;
            computeRoutes(tokenIn, tokenOut, currentRoute, poolsUsed, tokensVisited, currentTokenOut);
            poolsUsed[i] = false;
            currentRoute.pop();
            tokensVisited.delete(getAddressLowerCase(currentTokenOut));
        }
    };
    computeRoutes(tokenIn, tokenOut, [], poolsUsed, new Set([getAddressLowerCase(tokenIn)]));
    log.info({
        routes: routes.map(routeToString),
        pools: pools.map(poolToString),
    }, `Computed ${routes.length} possible routes for type ${(_a = routes[0]) === null || _a === void 0 ? void 0 : _a.protocol}.`);
    return routes;
}
