// @uniswap/universal-router-sdk/dist/esm/src/utils/commandParser.js
import { ethers } from 'ethers';
import UniversalRouter from '@uniswap/universal-router/artifacts/contracts/UniversalRouter.sol/UniversalRouter.json';
import { Interface } from '@ethersproject/abi';
import { V4BaseActionsParser } from '@uniswap/v4-sdk';
import { CommandType, COMMAND_DEFINITION, Subparser, Parser } from '../utils/routerCommands';
// Parses UniversalRouter V2 commands
export class CommandParser {
    static parseCalldata(calldata) {
        const genericParser = new GenericCommandParser(COMMAND_DEFINITION);
        const txDescription = CommandParser.INTERFACE.parseTransaction({ data: calldata });
        const { commands, inputs } = txDescription.args;
        return genericParser.parse(commands, inputs);
    }
}
CommandParser.INTERFACE = new Interface(UniversalRouter.abi);
// Parses commands based on given command definition
export class GenericCommandParser {
    constructor(commandDefinition) {
        this.commandDefinition = commandDefinition;
    }
    parse(commands, inputs) {
        const commandTypes = GenericCommandParser.getCommands(commands);
        return {
            commands: commandTypes.map((commandType, i) => {
                const commandDef = this.commandDefinition[commandType];
                if (commandDef.parser === Parser.V4Actions) {
                    const { actions } = V4BaseActionsParser.parseCalldata(inputs[i]);
                    return {
                        commandName: CommandType[commandType],
                        commandType,
                        params: v4RouterCallToParams(actions),
                    };
                }
                else if (commandDef.parser === Parser.Abi) {
                    const abiDef = commandDef.params;
                    const rawParams = ethers.utils.defaultAbiCoder.decode(abiDef.map((command) => command.type), inputs[i]);
                    const params = rawParams.map((param, j) => {
                        switch (abiDef[j].subparser) {
                            case Subparser.V3PathExactIn:
                                return {
                                    name: abiDef[j].name,
                                    value: parseV3PathExactIn(param),
                                };
                            case Subparser.V3PathExactOut:
                                return {
                                    name: abiDef[j].name,
                                    value: parseV3PathExactOut(param),
                                };
                            default:
                                return {
                                    name: abiDef[j].name,
                                    value: param,
                                };
                        }
                    });
                    return {
                        commandName: CommandType[commandType],
                        commandType,
                        params,
                    };
                }
                else if (commandDef.parser === Parser.V3Actions) {
                    // TODO: implement better parsing here
                    return {
                        commandName: CommandType[commandType],
                        commandType,
                        params: inputs.map((input) => ({ name: 'command', value: input })),
                    };
                }
                else {
                    throw new Error(`Unsupported parser: ${commandDef}`);
                }
            }),
        };
    }
    // parse command types from bytes string
    static getCommands(commands) {
        const commandTypes = [];
        for (let i = 2; i < commands.length; i += 2) {
            const byte = commands.substring(i, i + 2);
            commandTypes.push(parseInt(byte, 16));
        }
        return commandTypes;
    }
}
export function parseV3PathExactIn(path) {
    const strippedPath = path.replace('0x', '');
    let tokenIn = ethers.utils.getAddress(strippedPath.substring(0, 40));
    let loc = 40;
    const res = [];
    while (loc < strippedPath.length) {
        const feeAndTokenOut = strippedPath.substring(loc, loc + 46);
        const fee = parseInt(feeAndTokenOut.substring(0, 6), 16);
        const tokenOut = ethers.utils.getAddress(feeAndTokenOut.substring(6, 46));
        res.push({
            tokenIn,
            tokenOut,
            fee,
        });
        tokenIn = tokenOut;
        loc += 46;
    }
    return res;
}
export function parseV3PathExactOut(path) {
    const strippedPath = path.replace('0x', '');
    let tokenIn = ethers.utils.getAddress(strippedPath.substring(strippedPath.length - 40));
    let loc = strippedPath.length - 86; // 86 = (20 addr + 3 fee + 20 addr) * 2 (for hex characters)
    const res = [];
    while (loc >= 0) {
        const feeAndTokenOut = strippedPath.substring(loc, loc + 46);
        const tokenOut = ethers.utils.getAddress(feeAndTokenOut.substring(0, 40));
        const fee = parseInt(feeAndTokenOut.substring(40, 46), 16);
        res.push({
            tokenIn,
            tokenOut,
            fee,
        });
        tokenIn = tokenOut;
        loc -= 46;
    }
    return res;
}
function v4RouterCallToParams(actions) {
    return actions.map((action) => {
        return {
            name: action.actionName,
            value: action.params.map((param) => {
                return {
                    name: param.name,
                    value: param.value,
                };
            }),
        };
    });
}
