// @uniswap/universal-router-sdk/dist/esm/src/utils/routerCommands.js
import { defaultAbiCoder } from 'ethers/lib/utils';
import { isAtLeastV2_1_1 } from '@uniswap/v4-sdk';
/**
 * CommandTypes
 * @description Flags that modify a command's execution
 * @enum {number}
 */
export var CommandType;
(function (CommandType) {
    CommandType[CommandType["V3_SWAP_EXACT_IN"] = 0] = "V3_SWAP_EXACT_IN";
    CommandType[CommandType["V3_SWAP_EXACT_OUT"] = 1] = "V3_SWAP_EXACT_OUT";
    CommandType[CommandType["PERMIT2_TRANSFER_FROM"] = 2] = "PERMIT2_TRANSFER_FROM";
    CommandType[CommandType["PERMIT2_PERMIT_BATCH"] = 3] = "PERMIT2_PERMIT_BATCH";
    CommandType[CommandType["SWEEP"] = 4] = "SWEEP";
    CommandType[CommandType["TRANSFER"] = 5] = "TRANSFER";
    CommandType[CommandType["PAY_PORTION"] = 6] = "PAY_PORTION";
    CommandType[CommandType["PAY_PORTION_FULL_PRECISION"] = 7] = "PAY_PORTION_FULL_PRECISION";
    CommandType[CommandType["V2_SWAP_EXACT_IN"] = 8] = "V2_SWAP_EXACT_IN";
    CommandType[CommandType["V2_SWAP_EXACT_OUT"] = 9] = "V2_SWAP_EXACT_OUT";
    CommandType[CommandType["PERMIT2_PERMIT"] = 10] = "PERMIT2_PERMIT";
    CommandType[CommandType["WRAP_ETH"] = 11] = "WRAP_ETH";
    CommandType[CommandType["UNWRAP_WETH"] = 12] = "UNWRAP_WETH";
    CommandType[CommandType["PERMIT2_TRANSFER_FROM_BATCH"] = 13] = "PERMIT2_TRANSFER_FROM_BATCH";
    CommandType[CommandType["BALANCE_CHECK_ERC20"] = 14] = "BALANCE_CHECK_ERC20";
    CommandType[CommandType["V4_SWAP"] = 16] = "V4_SWAP";
    CommandType[CommandType["V3_POSITION_MANAGER_PERMIT"] = 17] = "V3_POSITION_MANAGER_PERMIT";
    CommandType[CommandType["V3_POSITION_MANAGER_CALL"] = 18] = "V3_POSITION_MANAGER_CALL";
    CommandType[CommandType["V4_INITIALIZE_POOL"] = 19] = "V4_INITIALIZE_POOL";
    CommandType[CommandType["V4_POSITION_MANAGER_CALL"] = 20] = "V4_POSITION_MANAGER_CALL";
    CommandType[CommandType["EXECUTE_SUB_PLAN"] = 33] = "EXECUTE_SUB_PLAN";
    // 3rd party integrations (0x40-0x5f range)
    CommandType[CommandType["ACROSS_V4_DEPOSIT_V3"] = 64] = "ACROSS_V4_DEPOSIT_V3";
})(CommandType || (CommandType = {}));
export var Subparser;
(function (Subparser) {
    Subparser[Subparser["V3PathExactIn"] = 0] = "V3PathExactIn";
    Subparser[Subparser["V3PathExactOut"] = 1] = "V3PathExactOut";
})(Subparser || (Subparser = {}));
export var Parser;
(function (Parser) {
    Parser[Parser["Abi"] = 0] = "Abi";
    Parser[Parser["V4Actions"] = 1] = "V4Actions";
    Parser[Parser["V3Actions"] = 2] = "V3Actions";
})(Parser || (Parser = {}));
const ALLOW_REVERT_FLAG = 0x80;
const REVERTIBLE_COMMANDS = new Set([CommandType.EXECUTE_SUB_PLAN]);
const PERMIT_STRUCT = '((address token,uint160 amount,uint48 expiration,uint48 nonce) details,address spender,uint256 sigDeadline)';
const PERMIT_BATCH_STRUCT = '((address token,uint160 amount,uint48 expiration,uint48 nonce)[] details,address spender,uint256 sigDeadline)';
const POOL_KEY_STRUCT = '(address currency0,address currency1,uint24 fee,int24 tickSpacing,address hooks)';
const PERMIT2_TRANSFER_FROM_STRUCT = '(address from,address to,uint160 amount,address token)';
const PERMIT2_TRANSFER_FROM_BATCH_STRUCT = PERMIT2_TRANSFER_FROM_STRUCT + '[]';
export const COMMAND_DEFINITION = {
    // Batch Reverts
    [CommandType.EXECUTE_SUB_PLAN]: {
        parser: Parser.Abi,
        params: [
            { name: 'commands', type: 'bytes' },
            { name: 'inputs', type: 'bytes[]' },
        ],
    },
    // Permit2 Actions
    [CommandType.PERMIT2_PERMIT]: {
        parser: Parser.Abi,
        params: [
            { name: 'permit', type: PERMIT_STRUCT },
            { name: 'signature', type: 'bytes' },
        ],
    },
    [CommandType.PERMIT2_PERMIT_BATCH]: {
        parser: Parser.Abi,
        params: [
            { name: 'permit', type: PERMIT_BATCH_STRUCT },
            { name: 'signature', type: 'bytes' },
        ],
    },
    [CommandType.PERMIT2_TRANSFER_FROM]: {
        parser: Parser.Abi,
        params: [
            { name: 'token', type: 'address' },
            { name: 'recipient', type: 'address' },
            { name: 'amount', type: 'uint160' },
        ],
    },
    [CommandType.PERMIT2_TRANSFER_FROM_BATCH]: {
        parser: Parser.Abi,
        params: [
            {
                name: 'transferFrom',
                type: PERMIT2_TRANSFER_FROM_BATCH_STRUCT,
            },
        ],
    },
    // Uniswap Actions
    [CommandType.V3_SWAP_EXACT_IN]: {
        parser: Parser.Abi,
        params: [
            { name: 'recipient', type: 'address' },
            { name: 'amountIn', type: 'uint256' },
            { name: 'amountOutMin', type: 'uint256' },
            { name: 'path', subparser: Subparser.V3PathExactIn, type: 'bytes' },
            { name: 'payerIsUser', type: 'bool' },
        ],
    },
    [CommandType.V3_SWAP_EXACT_OUT]: {
        parser: Parser.Abi,
        params: [
            { name: 'recipient', type: 'address' },
            { name: 'amountOut', type: 'uint256' },
            { name: 'amountInMax', type: 'uint256' },
            { name: 'path', subparser: Subparser.V3PathExactOut, type: 'bytes' },
            { name: 'payerIsUser', type: 'bool' },
        ],
    },
    [CommandType.V2_SWAP_EXACT_IN]: {
        parser: Parser.Abi,
        params: [
            { name: 'recipient', type: 'address' },
            { name: 'amountIn', type: 'uint256' },
            { name: 'amountOutMin', type: 'uint256' },
            { name: 'path', type: 'address[]' },
            { name: 'payerIsUser', type: 'bool' },
        ],
    },
    [CommandType.V2_SWAP_EXACT_OUT]: {
        parser: Parser.Abi,
        params: [
            { name: 'recipient', type: 'address' },
            { name: 'amountOut', type: 'uint256' },
            { name: 'amountInMax', type: 'uint256' },
            { name: 'path', type: 'address[]' },
            { name: 'payerIsUser', type: 'bool' },
        ],
    },
    [CommandType.V4_SWAP]: { parser: Parser.V4Actions },
    // Token Actions and Checks
    [CommandType.WRAP_ETH]: {
        parser: Parser.Abi,
        params: [
            { name: 'recipient', type: 'address' },
            { name: 'amount', type: 'uint256' },
        ],
    },
    [CommandType.UNWRAP_WETH]: {
        parser: Parser.Abi,
        params: [
            { name: 'recipient', type: 'address' },
            { name: 'amountMin', type: 'uint256' },
        ],
    },
    [CommandType.SWEEP]: {
        parser: Parser.Abi,
        params: [
            { name: 'token', type: 'address' },
            { name: 'recipient', type: 'address' },
            { name: 'amountMin', type: 'uint256' },
        ],
    },
    [CommandType.TRANSFER]: {
        parser: Parser.Abi,
        params: [
            { name: 'token', type: 'address' },
            { name: 'recipient', type: 'address' },
            { name: 'value', type: 'uint256' },
        ],
    },
    [CommandType.PAY_PORTION]: {
        parser: Parser.Abi,
        params: [
            { name: 'token', type: 'address' },
            { name: 'recipient', type: 'address' },
            { name: 'bips', type: 'uint256' },
        ],
    },
    [CommandType.PAY_PORTION_FULL_PRECISION]: {
        parser: Parser.Abi,
        params: [
            { name: 'token', type: 'address' },
            { name: 'recipient', type: 'address' },
            { name: 'portion', type: 'uint256' },
        ],
    },
    [CommandType.BALANCE_CHECK_ERC20]: {
        parser: Parser.Abi,
        params: [
            { name: 'owner', type: 'address' },
            { name: 'token', type: 'address' },
            { name: 'minBalance', type: 'uint256' },
        ],
    },
    [CommandType.V4_INITIALIZE_POOL]: {
        parser: Parser.Abi,
        params: [
            { name: 'poolKey', type: POOL_KEY_STRUCT },
            { name: 'sqrtPriceX96', type: 'uint160' },
        ],
    },
    // Position Actions
    [CommandType.V3_POSITION_MANAGER_PERMIT]: { parser: Parser.V3Actions },
    [CommandType.V3_POSITION_MANAGER_CALL]: { parser: Parser.V3Actions },
    [CommandType.V4_POSITION_MANAGER_CALL]: { parser: Parser.V4Actions },
    // 3rd Party Integrations
    [CommandType.ACROSS_V4_DEPOSIT_V3]: {
        parser: Parser.Abi,
        params: [
            { name: 'depositor', type: 'address' },
            { name: 'recipient', type: 'address' },
            { name: 'inputToken', type: 'address' },
            { name: 'outputToken', type: 'address' },
            { name: 'inputAmount', type: 'uint256' },
            { name: 'outputAmount', type: 'uint256' },
            { name: 'destinationChainId', type: 'uint256' },
            { name: 'exclusiveRelayer', type: 'address' },
            { name: 'quoteTimestamp', type: 'uint32' },
            { name: 'fillDeadline', type: 'uint32' },
            { name: 'exclusivityDeadline', type: 'uint32' },
            { name: 'message', type: 'bytes' },
            { name: 'useNative', type: 'bool' },
        ],
    },
};
// V2.1.1 ABI definitions for V2/V3 swap commands (extended with maxHopSlippage)
export const V2V3_SWAP_COMMANDS_V2_1_1 = {
    [CommandType.V3_SWAP_EXACT_IN]: {
        parser: Parser.Abi,
        params: [
            { name: 'recipient', type: 'address' },
            { name: 'amountIn', type: 'uint256' },
            { name: 'amountOutMin', type: 'uint256' },
            { name: 'path', subparser: Subparser.V3PathExactIn, type: 'bytes' },
            { name: 'payerIsUser', type: 'bool' },
            { name: 'maxHopSlippage', type: 'uint256[]' },
        ],
    },
    [CommandType.V3_SWAP_EXACT_OUT]: {
        parser: Parser.Abi,
        params: [
            { name: 'recipient', type: 'address' },
            { name: 'amountOut', type: 'uint256' },
            { name: 'amountInMax', type: 'uint256' },
            { name: 'path', subparser: Subparser.V3PathExactOut, type: 'bytes' },
            { name: 'payerIsUser', type: 'bool' },
            { name: 'maxHopSlippage', type: 'uint256[]' },
        ],
    },
    [CommandType.V2_SWAP_EXACT_IN]: {
        parser: Parser.Abi,
        params: [
            { name: 'recipient', type: 'address' },
            { name: 'amountIn', type: 'uint256' },
            { name: 'amountOutMin', type: 'uint256' },
            { name: 'path', type: 'address[]' },
            { name: 'payerIsUser', type: 'bool' },
            { name: 'maxHopSlippage', type: 'uint256[]' },
        ],
    },
    [CommandType.V2_SWAP_EXACT_OUT]: {
        parser: Parser.Abi,
        params: [
            { name: 'recipient', type: 'address' },
            { name: 'amountOut', type: 'uint256' },
            { name: 'amountInMax', type: 'uint256' },
            { name: 'path', type: 'address[]' },
            { name: 'payerIsUser', type: 'bool' },
            { name: 'maxHopSlippage', type: 'uint256[]' },
        ],
    },
};
export class RoutePlanner {
    constructor() {
        this.commands = '0x';
        this.inputs = [];
    }
    addSubPlan(subplan) {
        this.addCommand(CommandType.EXECUTE_SUB_PLAN, [subplan.commands, subplan.inputs], true);
        return this;
    }
    addCommand(type, parameters, allowRevert = false, urVersion) {
        let command = createCommand(type, parameters, urVersion);
        this.inputs.push(command.encodedInput);
        if (allowRevert) {
            if (!REVERTIBLE_COMMANDS.has(command.type)) {
                throw new Error(`command type: ${command.type} cannot be allowed to revert`);
            }
            command.type = command.type | ALLOW_REVERT_FLAG;
        }
        this.commands = this.commands.concat(command.type.toString(16).padStart(2, '0'));
        return this;
    }
    /**
     * Add Across bridge deposit command for cross-chain bridging
     * @param params AcrossV4DepositV3Params containing bridge parameters
     * @returns RoutePlanner instance for chaining
     */
    addAcrossBridge(params) {
        this.addCommand(CommandType.ACROSS_V4_DEPOSIT_V3, [
            params.depositor,
            params.recipient,
            params.inputToken,
            params.outputToken,
            params.inputAmount,
            params.outputAmount,
            params.destinationChainId,
            params.exclusiveRelayer,
            params.quoteTimestamp,
            params.fillDeadline,
            params.exclusivityDeadline,
            params.message,
            params.useNative,
        ]);
        return this;
    }
}
export function createCommand(type, parameters, urVersion) {
    const commandDef = isAtLeastV2_1_1(urVersion) && type in V2V3_SWAP_COMMANDS_V2_1_1
        ? V2V3_SWAP_COMMANDS_V2_1_1[type]
        : COMMAND_DEFINITION[type];
    switch (commandDef.parser) {
        case Parser.Abi:
            const encodedInput = defaultAbiCoder.encode(commandDef.params.map((abi) => abi.type), parameters);
            return { type, encodedInput };
        case Parser.V4Actions:
            // v4 swap data comes pre-encoded at index 0
            return { type, encodedInput: parameters[0] };
        case Parser.V3Actions:
            // v3 swap data comes pre-encoded at index 0
            return { type, encodedInput: parameters[0] };
    }
}
