// @uniswap/universal-router-sdk/dist/esm/src/utils/routerTradeAdapter.js
import { MixedRouteSDK, Trade as RouterTrade } from '@uniswap/router-sdk';
import { CurrencyAmount, Ether, Token } from '@uniswap/sdk-core';
import { Pair, Route as V2Route } from '@uniswap/v2-sdk';
import { Pool as V3Pool, Route as V3Route } from '@uniswap/v3-sdk';
import { Pool as V4Pool, Route as V4Route } from '@uniswap/v4-sdk';
import { BigNumber } from 'ethers';
import { ETH_ADDRESS, E_ETH_ADDRESS } from './constants';
export var PoolType;
(function (PoolType) {
    PoolType["V2Pool"] = "v2-pool";
    PoolType["V3Pool"] = "v3-pool";
    PoolType["V4Pool"] = "v4-pool";
})(PoolType || (PoolType = {}));
export const isNativeCurrency = (address) => address.toLowerCase() === ETH_ADDRESS.toLowerCase() || address.toLowerCase() === E_ETH_ADDRESS.toLowerCase();
// Helper class to convert routing-specific quote entities to RouterTrade entities
// the returned RouterTrade can then be used to build the UniswapTrade entity in this package
export class RouterTradeAdapter {
    // Generate a RouterTrade using fields from a classic quote response
    static fromClassicQuote(quote) {
        const { route, tokenIn, tokenOut } = quote;
        if (!route)
            throw new Error('Expected route to be present');
        if (!route.length)
            throw new Error('Expected there to be at least one route');
        if (route.some((r) => !r.length))
            throw new Error('Expected all routes to have at least one pool');
        const firstRoute = route[0];
        const tokenInData = firstRoute[0].tokenIn;
        const tokenOutData = firstRoute[firstRoute.length - 1].tokenOut;
        if (!tokenInData || !tokenOutData)
            throw new Error('Expected both tokenIn and tokenOut to be present');
        if (tokenInData.chainId !== tokenOutData.chainId)
            throw new Error('Expected tokenIn and tokenOut to be have same chainId');
        const parsedCurrencyIn = RouterTradeAdapter.toCurrency(isNativeCurrency(tokenIn), tokenInData);
        const parsedCurrencyOut = RouterTradeAdapter.toCurrency(isNativeCurrency(tokenOut), tokenOutData);
        const typedRoutes = route.map((subRoute) => {
            const rawAmountIn = subRoute[0].amountIn;
            const rawAmountOut = subRoute[subRoute.length - 1].amountOut;
            if (!rawAmountIn || !rawAmountOut) {
                throw new Error('Expected both raw amountIn and raw amountOut to be present');
            }
            const inputAmount = CurrencyAmount.fromRawAmount(parsedCurrencyIn, rawAmountIn);
            const outputAmount = CurrencyAmount.fromRawAmount(parsedCurrencyOut, rawAmountOut);
            const isOnlyV2 = RouterTradeAdapter.isVersionedRoute(PoolType.V2Pool, subRoute);
            const isOnlyV3 = RouterTradeAdapter.isVersionedRoute(PoolType.V3Pool, subRoute);
            const isOnlyV4 = RouterTradeAdapter.isVersionedRoute(PoolType.V4Pool, subRoute);
            return {
                routev4: isOnlyV4
                    ? new V4Route(subRoute.map(RouterTradeAdapter.toV4Pool), parsedCurrencyIn, parsedCurrencyOut)
                    : null,
                routev3: isOnlyV3
                    ? new V3Route(subRoute.map(RouterTradeAdapter.toV3Pool), parsedCurrencyIn, parsedCurrencyOut)
                    : null,
                routev2: isOnlyV2
                    ? new V2Route(subRoute.map(RouterTradeAdapter.toPair), parsedCurrencyIn, parsedCurrencyOut)
                    : null,
                mixedRoute: !isOnlyV4 && !isOnlyV3 && !isOnlyV2
                    ? new MixedRouteSDK(subRoute.map(RouterTradeAdapter.toPoolOrPair), parsedCurrencyIn, parsedCurrencyOut)
                    : null,
                inputAmount,
                outputAmount,
            };
        });
        return new RouterTrade({
            v2Routes: typedRoutes
                .filter((route) => route.routev2)
                .map((route) => ({
                routev2: route.routev2,
                inputAmount: route.inputAmount,
                outputAmount: route.outputAmount,
            })),
            v3Routes: typedRoutes
                .filter((route) => route.routev3)
                .map((route) => ({
                routev3: route.routev3,
                inputAmount: route.inputAmount,
                outputAmount: route.outputAmount,
            })),
            v4Routes: typedRoutes
                .filter((route) => route.routev4)
                .map((route) => ({
                routev4: route.routev4,
                inputAmount: route.inputAmount,
                outputAmount: route.outputAmount,
            })),
            mixedRoutes: typedRoutes
                .filter((route) => route.mixedRoute)
                .map((route) => ({
                mixedRoute: route.mixedRoute,
                inputAmount: route.inputAmount,
                outputAmount: route.outputAmount,
            })),
            tradeType: quote.tradeType,
        });
    }
    static toCurrency(isNative, token) {
        if (isNative) {
            return Ether.onChain(token.chainId);
        }
        return this.toToken(token);
    }
    static toToken(token) {
        const { chainId, address, decimals, symbol, buyFeeBps, sellFeeBps } = token;
        return new Token(chainId, address, parseInt(decimals.toString()), symbol, 
        /* name */ undefined, false, buyFeeBps ? BigNumber.from(buyFeeBps) : undefined, sellFeeBps ? BigNumber.from(sellFeeBps) : undefined);
    }
    static toV3Pool({ fee, sqrtRatioX96, liquidity, tickCurrent, tokenIn, tokenOut }) {
        return new V3Pool(RouterTradeAdapter.toToken(tokenIn), RouterTradeAdapter.toToken(tokenOut), parseInt(fee), sqrtRatioX96, liquidity, parseInt(tickCurrent));
    }
    static toV4Pool(pool) {
        const parsedCurrencyIn = RouterTradeAdapter.toCurrency(isNativeCurrency(pool.tokenIn.address), pool.tokenIn);
        const parsedCurrencyOut = RouterTradeAdapter.toCurrency(isNativeCurrency(pool.tokenOut.address), pool.tokenOut);
        return new V4Pool(parsedCurrencyIn, parsedCurrencyOut, parseInt(pool.fee), parseInt(pool.tickSpacing), pool.hooks, pool.sqrtRatioX96, pool.liquidity, parseInt(pool.tickCurrent));
    }
    static isVersionedRoute(type, route) {
        return route.every((pool) => pool.type === type);
    }
}
RouterTradeAdapter.toPoolOrPair = (pool) => {
    switch (pool.type) {
        case PoolType.V4Pool:
            return RouterTradeAdapter.toV4Pool(pool);
        case PoolType.V3Pool:
            return RouterTradeAdapter.toV3Pool(pool);
        case PoolType.V2Pool:
            return RouterTradeAdapter.toPair(pool);
        default:
            throw new Error('Invalid pool type');
    }
};
RouterTradeAdapter.toPair = ({ reserve0, reserve1 }) => {
    return new Pair(CurrencyAmount.fromRawAmount(RouterTradeAdapter.toToken(reserve0.token), reserve0.quotient), CurrencyAmount.fromRawAmount(RouterTradeAdapter.toToken(reserve1.token), reserve1.quotient));
};
