// @uniswap/v3-sdk/dist/cjs/src/entities/pool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pool = void 0;
const tslib_1 = require("tslib");
const sdk_core_1 = require("@uniswap/sdk-core");
const jsbi_1 = tslib_1.__importDefault(require("jsbi"));
const tiny_invariant_1 = tslib_1.__importDefault(require("tiny-invariant"));
const constants_1 = require("../constants");
const internalConstants_1 = require("../internalConstants");
const computePoolAddress_1 = require("../utils/computePoolAddress");
const v3swap_1 = require("../utils/v3swap");
const tickMath_1 = require("../utils/tickMath");
const tickDataProvider_1 = require("./tickDataProvider");
const tickListDataProvider_1 = require("./tickListDataProvider");
/**
 * By default, pools will not allow operations that require ticks.
 */
const NO_TICK_DATA_PROVIDER_DEFAULT = new tickDataProvider_1.NoTickDataProvider();
/**
 * Represents a V3 pool
 */
class Pool {
    static getAddress(tokenA, tokenB, fee, initCodeHashManualOverride, factoryAddressOverride) {
        return (0, computePoolAddress_1.computePoolAddress)({
            factoryAddress: factoryAddressOverride !== null && factoryAddressOverride !== void 0 ? factoryAddressOverride : constants_1.FACTORY_ADDRESS,
            fee,
            tokenA,
            tokenB,
            initCodeHashManualOverride,
        });
    }
    /**
     * Construct a pool
     * @param tokenA One of the tokens in the pool
     * @param tokenB The other token in the pool
     * @param fee The fee in hundredths of a bips of the input amount of every swap that is collected by the pool
     * @param sqrtRatioX96 The sqrt of the current ratio of amounts of token1 to token0
     * @param liquidity The current value of in range liquidity
     * @param tickCurrent The current tick of the pool
     * @param ticks The current state of the pool ticks or a data provider that can return tick data
     */
    constructor(tokenA, tokenB, fee, sqrtRatioX96, liquidity, tickCurrent, ticks = NO_TICK_DATA_PROVIDER_DEFAULT) {
        (0, tiny_invariant_1.default)(Number.isInteger(fee) && fee < 1000000, 'FEE');
        const tickCurrentSqrtRatioX96 = tickMath_1.TickMath.getSqrtRatioAtTick(tickCurrent);
        const nextTickSqrtRatioX96 = tickMath_1.TickMath.getSqrtRatioAtTick(tickCurrent + 1);
        (0, tiny_invariant_1.default)(jsbi_1.default.greaterThanOrEqual(jsbi_1.default.BigInt(sqrtRatioX96), tickCurrentSqrtRatioX96) &&
            jsbi_1.default.lessThanOrEqual(jsbi_1.default.BigInt(sqrtRatioX96), nextTickSqrtRatioX96), 'PRICE_BOUNDS');
        [this.token0, this.token1] = tokenA.sortsBefore(tokenB) ? [tokenA, tokenB] : [tokenB, tokenA];
        this.fee = fee;
        this.sqrtRatioX96 = jsbi_1.default.BigInt(sqrtRatioX96);
        this.liquidity = jsbi_1.default.BigInt(liquidity);
        this.tickCurrent = tickCurrent;
        this.tickDataProvider = Array.isArray(ticks) ? new tickListDataProvider_1.TickListDataProvider(ticks, constants_1.TICK_SPACINGS[fee]) : ticks;
    }
    /**
     * Returns true if the token is either token0 or token1
     * @param token The token to check
     * @returns True if token is either token0 or token
     */
    involvesToken(token) {
        return token.equals(this.token0) || token.equals(this.token1);
    }
    /**
     * Returns the current mid price of the pool in terms of token0, i.e. the ratio of token1 over token0
     */
    get token0Price() {
        var _a;
        return ((_a = this._token0Price) !== null && _a !== void 0 ? _a : (this._token0Price = new sdk_core_1.Price(this.token0, this.token1, internalConstants_1.Q192, jsbi_1.default.multiply(this.sqrtRatioX96, this.sqrtRatioX96))));
    }
    /**
     * Returns the current mid price of the pool in terms of token1, i.e. the ratio of token0 over token1
     */
    get token1Price() {
        var _a;
        return ((_a = this._token1Price) !== null && _a !== void 0 ? _a : (this._token1Price = new sdk_core_1.Price(this.token1, this.token0, jsbi_1.default.multiply(this.sqrtRatioX96, this.sqrtRatioX96), internalConstants_1.Q192)));
    }
    /**
     * Return the price of the given token in terms of the other token in the pool.
     * @param token The token to return price of
     * @returns The price of the given token, in terms of the other.
     */
    priceOf(token) {
        (0, tiny_invariant_1.default)(this.involvesToken(token), 'TOKEN');
        return token.equals(this.token0) ? this.token0Price : this.token1Price;
    }
    /**
     * Returns the chain ID of the tokens in the pool.
     */
    get chainId() {
        return this.token0.chainId;
    }
    /**
     * Given an input amount of a token, return the computed output amount, and a pool with state updated after the trade
     * @param inputAmount The input amount for which to quote the output amount
     * @param sqrtPriceLimitX96 The Q64.96 sqrt price limit
     * @returns The output amount and the pool with updated state
     */
    async getOutputAmount(inputAmount, sqrtPriceLimitX96) {
        (0, tiny_invariant_1.default)(this.involvesToken(inputAmount.currency), 'TOKEN');
        const zeroForOne = inputAmount.currency.equals(this.token0);
        const { amountCalculated: outputAmount, sqrtRatioX96, liquidity, tickCurrent, } = await this.swap(zeroForOne, inputAmount.quotient, sqrtPriceLimitX96);
        const outputToken = zeroForOne ? this.token1 : this.token0;
        return [
            sdk_core_1.CurrencyAmount.fromRawAmount(outputToken, jsbi_1.default.multiply(outputAmount, internalConstants_1.NEGATIVE_ONE)),
            new Pool(this.token0, this.token1, this.fee, sqrtRatioX96, liquidity, tickCurrent, this.tickDataProvider),
        ];
    }
    /**
     * Given a desired output amount of a token, return the computed input amount and a pool with state updated after the trade
     * @param outputAmount the output amount for which to quote the input amount
     * @param sqrtPriceLimitX96 The Q64.96 sqrt price limit. If zero for one, the price cannot be less than this value after the swap. If one for zero, the price cannot be greater than this value after the swap
     * @returns The input amount and the pool with updated state
     */
    async getInputAmount(outputAmount, sqrtPriceLimitX96) {
        (0, tiny_invariant_1.default)(outputAmount.currency.isToken && this.involvesToken(outputAmount.currency), 'TOKEN');
        const zeroForOne = outputAmount.currency.equals(this.token1);
        const { amountCalculated: inputAmount, sqrtRatioX96, liquidity, tickCurrent, } = await this.swap(zeroForOne, jsbi_1.default.multiply(outputAmount.quotient, internalConstants_1.NEGATIVE_ONE), sqrtPriceLimitX96);
        const inputToken = zeroForOne ? this.token0 : this.token1;
        return [
            sdk_core_1.CurrencyAmount.fromRawAmount(inputToken, inputAmount),
            new Pool(this.token0, this.token1, this.fee, sqrtRatioX96, liquidity, tickCurrent, this.tickDataProvider),
        ];
    }
    /**
     * Executes a swap
     * @param zeroForOne Whether the amount in is token0 or token1
     * @param amountSpecified The amount of the swap, which implicitly configures the swap as exact input (positive), or exact output (negative)
     * @param sqrtPriceLimitX96 The Q64.96 sqrt price limit. If zero for one, the price cannot be less than this value after the swap. If one for zero, the price cannot be greater than this value after the swap
     * @returns amountCalculated
     * @returns sqrtRatioX96
     * @returns liquidity
     * @returns tickCurrent
     */
    async swap(zeroForOne, amountSpecified, sqrtPriceLimitX96) {
        return (0, v3swap_1.v3Swap)(jsbi_1.default.BigInt(this.fee), this.sqrtRatioX96, this.tickCurrent, this.liquidity, this.tickSpacing, this.tickDataProvider, zeroForOne, amountSpecified, sqrtPriceLimitX96);
    }
    get tickSpacing() {
        return constants_1.TICK_SPACINGS[this.fee];
    }
}
exports.Pool = Pool;
